<!DOCTYPE html>












  


<html class="theme-next gemini use-motion" lang="en">
<head><meta name="generator" content="Hexo 3.8.0">
  <meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">












<meta http-equiv="Cache-Control" content="no-transform">
<meta http-equiv="Cache-Control" content="no-siteapp">



  <meta name="google-site-verification" content="3dnqtuxhGQK5nK7aqRtU2c3DzZlaVtBMN6M39UoGcgY">




















<link href="/blog/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css">

<link href="/blog/css/main.css?v=6.3.0" rel="stylesheet" type="text/css">


  <link rel="apple-touch-icon" sizes="180x180" href="/blog/images/apple-touch-icon-next.png?v=6.3.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/blog/images/favicon-32x32-next.png?v=6.3.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/blog/images/favicon-16x16-next.png?v=6.3.0">


  <link rel="mask-icon" href="/blog/images/logo.svg?v=6.3.0" color="#222">









<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/blog/',
    scheme: 'Gemini',
    version: '6.3.0',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":true,"onmobile":false},
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="The content were compiled from multiple resources on the internet (forum, papers, workshop etc.). I could not indicate all the sources, but I want to thank them for sharing experiences/knowledge/code.">
<meta name="keywords" content="peak,ATAC-seq,MACS,bowtie2">
<meta property="og:type" content="article">
<meta property="og:title" content="ATAC-seq data analysis: from FASTQ to peaks">
<meta property="og:url" content="https://yiweiniu.github.io/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/index.html">
<meta property="og:site_name" content="Yiwei Niu&#39;s Note">
<meta property="og:description" content="The content were compiled from multiple resources on the internet (forum, papers, workshop etc.). I could not indicate all the sources, but I want to thank them for sharing experiences/knowledge/code.">
<meta property="og:locale" content="en">
<meta property="og:image" content="https://yiweiniu.github.io/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/20190320181006767_5393.png">
<meta property="og:image" content="https://yiweiniu.github.io/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/20190320180433696_25889.png">
<meta property="og:image" content="https://yiweiniu.github.io/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/20190318163447907_19139.png">
<meta property="og:image" content="https://yiweiniu.github.io/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/20190111215229195_23911.png">
<meta property="og:image" content="https://yiweiniu.github.io/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/20190103170441947_28541.png">
<meta property="og:image" content="https://yiweiniu.github.io/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/20190103170526610_9297.png">
<meta property="og:image" content="https://yiweiniu.github.io/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/20190318163604339_31625.png">
<meta property="og:updated_time" content="2019-04-01T08:06:36.315Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="ATAC-seq data analysis: from FASTQ to peaks">
<meta name="twitter:description" content="The content were compiled from multiple resources on the internet (forum, papers, workshop etc.). I could not indicate all the sources, but I want to thank them for sharing experiences/knowledge/code.">
<meta name="twitter:image" content="https://yiweiniu.github.io/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/20190320181006767_5393.png">



  <link rel="alternate" href="/blog/atom.xml" title="Yiwei Niu's Note" type="application/atom+xml">




  <link rel="canonical" href="https://yiweiniu.github.io/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/">



<script type="text/javascript" id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>ATAC-seq data analysis: from FASTQ to peaks | Yiwei Niu's Note</title>
  




<script async src="https://www.googletagmanager.com/gtag/js?id=UA-122074944-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-122074944-1');
</script>






  <noscript>
  <style type="text/css">
    .use-motion .motion-element,
    .use-motion .brand,
    .use-motion .menu-item,
    .sidebar-inner,
    .use-motion .post-block,
    .use-motion .pagination,
    .use-motion .comments,
    .use-motion .post-header,
    .use-motion .post-body,
    .use-motion .collection-title { opacity: initial; }

    .use-motion .logo,
    .use-motion .site-title,
    .use-motion .site-subtitle {
      opacity: initial;
      top: initial;
    }

    .use-motion {
      .logo-line-before i { left: initial; }
      .logo-line-after i { right: initial; }
    }
  </style>
</noscript>

</head>

<body itemscope="" itemtype="http://schema.org/WebPage" lang="en">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope="" itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/blog/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Yiwei Niu's Note</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
      
        <p class="site-subtitle">to share, to learn</p>
      
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="Toggle navigation bar">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home">
    <a href="/blog/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-home"></i> <br>Home</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-categories">
    <a href="/blog/categories/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-th"></i> <br>Categories</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-tags">
    <a href="/blog/tags" rel="section">
      <i class="menu-item-icon fa fa-fw fa-tags"></i> <br>Tags</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">
    <a href="/blog/archives" rel="section">
      <i class="menu-item-icon fa fa-fw fa-archive"></i> <br>Archives</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-about">
    <a href="https:/yiweiniu.github.io/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-user"></i> <br>About</a>
  </li>

      
      
    </ul>
  

  

  
</nav>



  



</div>
    </header>

    


    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
            

          
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  

  <article class="post post-type-normal" itemscope="" itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="https://yiweiniu.github.io/blog/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/">

    <span hidden itemprop="author" itemscope="" itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Yiwei Niu">
      <meta itemprop="description" content="a beginner, a learner">
      <meta itemprop="image" content="/blog/images/portrait.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope="" itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Yiwei Niu's Note">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">ATAC-seq data analysis: from FASTQ to peaks
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2019-03-20 18:15:57" itemprop="dateCreated datePublished" datetime="2019-03-20T18:15:57+08:00">2019-03-20</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope="" itemtype="http://schema.org/Thing"><a href="/blog/categories/Peak/" itemprop="url" rel="index"><span itemprop="name">Peak</span></a></span>

                
                
                  , 
                
              
                <span itemprop="about" itemscope="" itemtype="http://schema.org/Thing"><a href="/blog/categories/Peak/ATAC-seq/" itemprop="url" rel="index"><span itemprop="name">ATAC-seq</span></a></span>

                
                
              
            </span>
          

          
            
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
                <a href="/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/#comments" itemprop="discussionUrl">
                
                  <span class="post-comments-count disqus-comment-count" data-disqus-identifier="2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
             <span id="/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/" class="leancloud_visitors" data-flag-title="ATAC-seq data analysis: from FASTQ to peaks">
               <span class="post-meta-divider">|</span>
               <span class="post-meta-item-icon">
                 <i class="fa fa-eye"></i>
               </span>
               
                 <span class="post-meta-item-text">Views: </span>
               
                 <span class="leancloud-visitors-count"></span>
             </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                <span title="Symbols count in article">29k</span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                <span title="Reading time">0:26</span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <p>The content were compiled from multiple resources on the internet (forum, papers, workshop etc.). I could not indicate all the sources, but I want to thank them for sharing experiences/knowledge/code.</p>
<h2 id="atac-seq-overview"><a class="markdownIt-Anchor" href="#atac-seq-overview"></a> ATAC-seq overview</h2>
<p>ATAC-seq (Assay for Transposase-Accessible Chromatin with high-throughput sequencing) is a method for determining chromatin accessibility across the genome. It utilizes a hyperactive Tn5 transposase to insert sequencing adapters into open chromatin regions.</p>
<img src="/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/20190320181006767_5393.png">
<p>ATAC-seq overview (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4374986/" target="_blank" rel="noopener">Buenrostro <em>et al.</em>, 2015</a>).</p>
<p>And the peaks look like the following figure.</p>
<img src="/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/20190320180433696_25889.png">
<p>ATAC-seq peaks <a href="https://doi.org/10.1186/1756-8935-7-33" target="_blank" rel="noopener">(Tsompana and Buck, 2014)</a></p>
<p>And ATAC-seq can be used to:</p>
<ul>
<li>generate epigenomic profiles</li>
<li>map accessible chromatin across tissues or conditions</li>
<li>retrieve nucleosome positions</li>
<li>identify important transcription factors</li>
<li>generate occupancy profiles of TFs (footprinting)</li>
</ul>
<h2 id="experimental-design"><a class="markdownIt-Anchor" href="#experimental-design"></a> Experimental design</h2>
<p>See <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4374986/" target="_blank" rel="noopener">Buenrostro <em>et al.</em>, 2015</a>, <a href="https://www.encodeproject.org/atac-seq/" target="_blank" rel="noopener">ENCODE - ATAC-seq Data Standards and Prototype Processing Pipeline</a>, and <a href="https://github.com/harvardinformatics/ATAC-seq" target="_blank" rel="noopener">Harvard FAS Informatics - ATAC-seq Guidelines</a> for details.</p>
<ul>
<li>two or more biological replicates</li>
<li>each replicate has 25 million non-duplicate, non-mitochondrial aligned reads for single-end sequencing and 50 million for paired-ended sequencing</li>
<li>typically, no need for “input”</li>
<li>use as few PCR cycles as possible when constructing the library</li>
<li>paired-end sequencing is preferred</li>
</ul>
<h2 id="data-analysis"><a class="markdownIt-Anchor" href="#data-analysis"></a> Data analysis</h2>
<p>Several useful pipelines.</p>
<ul>
<li><a href="https://github.com/harvardinformatics/ATAC-seq" target="_blank" rel="noopener">Harvard FAS Informatics - ATAC-seq Guidelines</a> – clear and up-to-date.</li>
<li><a href="https://www.encodeproject.org/atac-seq/" target="_blank" rel="noopener">ENCODE - ATAC-seq Data Standards and Prototype Processing Pipeline</a></li>
<li><a href="https://github.com/tobiasrausch/ATACseq" target="_blank" rel="noopener">Tobias Rausch - ATAC-seq analysis pipeline</a></li>
<li><a href="https://github.com/ENCODE-DCC/atac-seq-pipeline" target="_blank" rel="noopener">ENCODE ATAC-seq pipeline</a></li>
<li><a href="https://github.com/ParkerLab/bioinf525" target="_blank" rel="noopener">Parker Lab - ATAC-seq lab for BIOINF525</a></li>
<li><a href="https://github.com/ay-lab/ATACProc" target="_blank" rel="noopener">Ferhat Ay Lab - ATAC-seq processing pipeline</a></li>
<li><a href="http://qiubio.com/new/book/chapter-06/#%E7%AC%AC%E4%BA%94%E7%AB%A0-atac-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90chapter-5-atac-seq-data-analysis" target="_blank" rel="noopener">生物信息学生 R 入门教程 - 第五章 ATAC-seq数据分析</a> – in R.</li>
</ul>
<p>The following pipeline includes several common analysis in ATAC-seq setting, from data trimming to peak calling. Some steps are optional, like merging BAMs.</p>
<h3 id="quality-control"><a class="markdownIt-Anchor" href="#quality-control"></a> Quality control</h3>
<p>Just like analyzing other NGS data, quality control is needed for raw <code>FASTQ</code>, and there are many programs avaliable, such as <a href="http://www.usadellab.org/cms/?page=trimmomatic" target="_blank" rel="noopener">Trimmomatic</a> and <a href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/" target="_blank" rel="noopener">Trim Galore</a>.</p>
<h3 id="alignment-and-filter"><a class="markdownIt-Anchor" href="#alignment-and-filter"></a> Alignment and filter</h3>
<p>The next step is to align reads to a reference genome. Two popular aligners are <a href="http://bio-bwa.sourceforge.net/bwa.shtml" target="_blank" rel="noopener">BWA</a> and <a href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml" target="_blank" rel="noopener">Bowtie2</a>. I will use <code>Bowtie2</code> (since it was used in many tutorals and papers).</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> better alignment results are frequently achieved with --very-sensitive</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> use -X 2000 to allow larger fragment size (default is 500)</span></span><br><span class="line">bowtie2 --very-sensitive -X 2000 -x $Bowtie2Index -1 $&#123;sample&#125;*_1.fq.gz -2 $&#123;sample&#125;*_2.fq.gz \</span><br><span class="line">  -p $PPN 2&gt; $$&#123;sample&#125;.bowtie2.log | $path2samtools sort -@ $PPN -O bam -o $&#123;sample&#125;.sorted.bam</span><br><span class="line"><span class="meta">$</span><span class="bash">path2samtools index -@ <span class="variable">$PPN</span> <span class="variable">$WORKDIR</span>/bowtie2/<span class="variable">$&#123;sample&#125;</span>.sorted.bam</span></span><br></pre></td></tr></table></figure>
<p>Then, the alignment results should be filtered.</p>
<h4 id="mitochondrial-reads"><a class="markdownIt-Anchor" href="#mitochondrial-reads"></a> Mitochondrial reads</h4>
<p>Ref: <a href="https://github.com/harvardinformatics/ATAC-seq" target="_blank" rel="noopener">Harvard FAS Informatics - ATAC-seq Guidelines</a></p>
<blockquote>
<p>Since there are no ATAC-seq peaks of interest in the mitochondrial genome, these reads will only complicate the subsequent steps. Therefore, we recommend that they be removed from further analysis, via one of the following methods:</p>
<ol>
<li>Remove the mitochondrial genome from the reference genome before aligning the reads. In human/mouse genome builds, the mitochondrial genome is labeled ‘chrM’. That sequence can be deleted from the reference prior to building the genome indexes. The downside of this approach is that the alignment numbers will look much worse; all of the mitochondrial reads will count as unaligned.</li>
<li>Remove the mitochondrial reads after alignment. A python script, creatively named removeChrom, is available in the ATAC-seq module to accomplish this.</li>
</ol>
</blockquote>
<p>Since the percentage of mtDNA-reads is a indicator of library quality, we usually remove mitochondrial reads after alignment. It is run as follows:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools view -@ $PPN -h $&#123;sample&#125;.bam | grep -v chrM | samtools sort -@ $PPN -O bam -o $&#123;sample&#125;.rmChrM.bam</span><br></pre></td></tr></table></figure>
<h4 id="pcr-duplicates"><a class="markdownIt-Anchor" href="#pcr-duplicates"></a> PCR duplicates</h4>
<p>Ref: <a href="https://github.com/harvardinformatics/ATAC-seq" target="_blank" rel="noopener">Harvard FAS Informatics - ATAC-seq Guidelines</a></p>
<blockquote>
<p>PCR duplicates are exact copies of DNA fragments that arise during PCR. Since they are artifacts of the library preparation procedure, they may interfere with the biological signal of interest. Therefore, they should be removed as part of the analysis pipeline.</p>
<p>One commonly used program for removing PCR duplicates is Picard’s <code>MarkDuplicates</code>.</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">$</span><span class="bash">&#123;path2java&#125; -XX:ParallelGCThreads=<span class="variable">$&#123;PPN&#125;</span> -Djava.io.tmpdir=/tmp -jar <span class="variable">$&#123;path2picard&#125;</span> MarkDuplicates \</span></span><br><span class="line">  QUIET=true INPUT=$&#123;sample&#125;.bam OUTPUT=$&#123;sample&#125;.marked.bam METRICS_FILE=$&#123;sample&#125;.sorted.metrics \</span><br><span class="line">  REMOVE_DUPLICATES=false CREATE_INDEX=true VALIDATION_STRINGENCY=LENIENT TMP_DIR=/tmp</span><br><span class="line"><span class="meta">#</span><span class="bash"> REMOVE_DUPLICATES=<span class="literal">false</span>: mark duplicate reads, not remove.</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> Change it to <span class="literal">true</span> to remove duplicate reads.</span></span><br></pre></td></tr></table></figure>
<p><code>MarkDuplicates</code> will add a FALG <code>1024</code> to duplicate reads, we can remove them using <code>samtools</code>:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools view -h -b -F 1024 $&#123;sample&#125;.bam &gt; $&#123;sample&#125;.rmDup.bam</span><br></pre></td></tr></table></figure>
<h4 id="non-unique-alignments"><a class="markdownIt-Anchor" href="#non-unique-alignments"></a> Non-unique alignments</h4>
<p>Ref: <a href="https://github.com/harvardinformatics/ATAC-seq" target="_blank" rel="noopener">Harvard FAS Informatics - ATAC-seq Guidelines</a></p>
<blockquote>
<p>Some researchers choose to remove non-uniquely aligned reads, using the <code>-q</code> parameter of <code>samtools view</code>.</p>
</blockquote>
<p>Different genome aligners have varied implementation of mapping quality (MAPQ). See <a href="https://www.acgt.me/?offset=1426809676847" target="_blank" rel="noopener">More madness with MAPQ scores (a.k.a. why bioinformaticians hate poor and incomplete software documentation)</a>. So, when using MAPQ to filter non-unique alignments, do check the MAPQ values of the aligner using.</p>
<p>For Bowtie2, people usually use MAPQ &gt; 30.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> Remove multi-mapped reads (i.e. those with MAPQ &lt; 30, using -q <span class="keyword">in</span> SAMtools)</span></span><br><span class="line">samtools view -h -q 30 $&#123;sample&#125;.bam &gt; $&#123;sample&#125;.rmMulti.bam</span><br></pre></td></tr></table></figure>
<h4 id="others"><a class="markdownIt-Anchor" href="#others"></a> Others</h4>
<p>In the pipeline by ENCODE or some papers, the following reads were also removed  (samtoolf flag 1796 or 1804).</p>
<ul>
<li>reads unmapped,</li>
<li>not primary alignment</li>
<li>reads failing platform</li>
<li>duplicates</li>
</ul>
<p>The remaining reads are so-called “properly mapped reads”.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> Remove reads unmapped, mate unmapped, not primary alignment, reads failing platform, duplicates (-F 1804)</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> Retain properly paired reads -f 2</span></span><br><span class="line">samtools view -h -b -F 1804 -f 2 $&#123;sample&#125;.bam &gt; $&#123;sample&#125;.filtered.bam</span><br></pre></td></tr></table></figure>
<h4 id="merging-bams-optional"><a class="markdownIt-Anchor" href="#merging-bams-optional"></a> Merging BAMs (optional)</h4>
<p>When several libraries were constructed for one experimental condition (aka. one experimental condition inlcuded several biological and/or technical replicates), one may want to merge different <code>BAM</code> files before calling peaks (e.g. merge BAM files from technical replicates, merge BAM files to get bigger read depth).</p>
<p>Considerations about “merge BAMs” or “merge peaks” have been discussed:</p>
<ul>
<li>
<p><a href="https://www.biostars.org/p/112778/" target="_blank" rel="noopener">ChIP-Seq: Calling peaks with replicates</a></p>
<blockquote>
<p>First, are these A) technical or B) biological replicates? That is, the same biological sample run several times with the same antibody (same lot also if polyclonal) protocol, or different biological samples run the same way with the same protocol?</p>
<p>If it is A it may be reasonable to merge them for some analyses, such as just annotating peaks. I would merge the bam alignment files and then do the calls versus merging the calls.</p>
<p>However, first you have analyze your replicates to check they they all perform the same. We did a lot of performance comparisons here: <a href="https://epigeneticsandchromatin.biomedcentral.com/articles/10.1186/s13072-016-0100-6" target="_blank" rel="noopener">https://epigeneticsandchromatin.biomedcentral.com/articles/10.1186/s13072-016-0100-6</a></p>
<p>You can steal those ideas, especially using the ENCODE segmentation tracks if it’s human and they have tracks for something like your cell type. But just counting the reads in bins and then doing a correlation is pretty informative.</p>
<p>But even in our data, and we used a robot and do it a lot, one of our technical replicates behaved strangely. See supplemental figure S6.</p>
<p>If it is B, biological replicates, you almost certainly don’t want to merge them. You will lose your information about biological variance is present. If you are looking at something like differential peaks between conditions DESeq and really all reputable programs will want some sort of replicates, almost always biological. In general, if you want to compute a p value on anything you need separate replicates (not merged).</p>
<p>If you are just annotating peaks you don’t need a p value.</p>
</blockquote>
</li>
<li>
<p><a href="https://www.biostars.org/p/191474/" target="_blank" rel="noopener">how to pool together biological replicates?</a></p>
</li>
<li>
<p><a href="https://www.biostars.org/p/210564/" target="_blank" rel="noopener">ChipSeq: merge bam file before peak calling</a></p>
</li>
<li>
<p><a href="https://www.biostars.org/p/230055/" target="_blank" rel="noopener">Chip-Seq merging peak files</a></p>
</li>
</ul>
<p>And merging BAM can be done using <code>samtools</code>.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">samtools merge -@ $PPN condition1.merged.bam sample1.bam sample2.bam sample3.bam</span><br><span class="line">samtools index -@ $PPN condition1.merged.bam</span><br></pre></td></tr></table></figure>
<p>Also, <code>multiBamSummary</code> in <code>deepTools</code> can be used to check the correlations between <code>BAM</code> files before merging.</p>
<h3 id="shifting-reads"><a class="markdownIt-Anchor" href="#shifting-reads"></a> Shifting reads</h3>
<p>In the first ATAC-seq paper (<a href="https://www.nature.com/articles/nmeth.2688" target="_blank" rel="noopener">Buenrostro et al., 2013</a>), all reads aligning to the + strand were offset by +4 bp, and all reads aligning to the – strand were offset −5 bp, since Tn5 transposase has been shown to bind as a dimer and insert two adaptors separated by 9 bp <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-12-r119" target="_blank" rel="noopener">(Adey et al., 2010)</a>.</p>
<p>Ref: <a href="https://github.com/GreenleafLab/NucleoATAC/issues/58" target="_blank" rel="noopener">shifting reads bam for NucleoATAC?</a></p>
<blockquote>
<p>However, for peak calling, shifting of reads is not likely very important, as it is a pretty minor adjustment and peaks are 100s of basepairs. The shifting is only crucial when doing things where the exact position of the insertion matters at single base resolution, e.g. TF motif footprinting.</p>
</blockquote>
<p>Also, remember that not all TF footprinting tools need shifted reads. Some of them may do this internally, e.g. <code>NucleoATAC</code>.</p>
<p>But, <strong>how to adjust the reads alignment?</strong></p>
<p>First, we could do this using <code>bedtools</code> and <code>awk</code>.</p>
<p>Ref: <a href="https://www.biostars.org/p/187204/#187206" target="_blank" rel="noopener">Shifting reads for ATAC-seq alignments</a></p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> the BAM file should be sorted by <span class="built_in">read</span> name beforehand</span></span><br><span class="line">samtools sort -n -T aln.sorted -o aln.sorted.bam aln.bam</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> The bedtools <span class="built_in">command</span> should extract the paired-end alignments as bedpe format, <span class="keyword">then</span> the awk <span class="built_in">command</span> should <span class="built_in">shift</span> the fragments as needed</span></span><br><span class="line">bedtools bamtobed -i reads.bam -bedpe | awk -v OFS="\t" '&#123;($9=="+")&#123;print $1,$2+4,$6+4&#125; \</span><br><span class="line"><span class="meta">  ($</span><span class="bash">9==<span class="string">"-"</span>)&#123;<span class="built_in">print</span> <span class="variable">$1</span>,<span class="variable">$2</span>-5,<span class="variable">$6</span>-5&#125;&#125;<span class="string">' &gt; fragments.bed</span></span></span><br></pre></td></tr></table></figure>
<p>Or, we could do this using <a href="https://deeptools.readthedocs.io/en/develop/content/tools/alignmentSieve.html" target="_blank" rel="noopener">alignmentSieve</a>.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> use --ATACshift</span></span><br><span class="line">alignmentSieve --numberOfProcessors 8 --ATACshift --bam sample1.bam -o sample1.tmp.bam</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> the bam file needs to be sorted again</span></span><br><span class="line">samtools sort -@ 8 -O bam -o sample1.shifted.bam sample1.tmp.bam</span><br><span class="line">samtools index -@ 8 sample1.shifted.bam</span><br><span class="line">rm sample1.tmp.bam</span><br></pre></td></tr></table></figure>
<img src="/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/20190318163447907_19139.png">
<p>We could also do this in <code>R</code> using <a href="https://bioconductor.org/packages/release/bioc/html/ATACseqQC.html" target="_blank" rel="noopener">ATACseqQC</a></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## load the library</span></span><br><span class="line"><span class="keyword">library</span>(ATACseqQC)</span><br><span class="line"></span><br><span class="line"><span class="comment">## input is bamFile</span></span><br><span class="line">bamfile &lt;- system.file(<span class="string">"extdata"</span>, <span class="string">"GL1.bam"</span>, package=<span class="string">"ATACseqQC"</span>, mustWork=<span class="literal">TRUE</span>)</span><br><span class="line">bamfile.labels &lt;- gsub(<span class="string">".bam"</span>, <span class="string">""</span>, basename(bamfile))</span><br><span class="line"></span><br><span class="line"><span class="comment">## bamfile tags</span></span><br><span class="line">tags &lt;- c(<span class="string">"AS"</span>, <span class="string">"XN"</span>, <span class="string">"XM"</span>, <span class="string">"XO"</span>, <span class="string">"XG"</span>, <span class="string">"NM"</span>, <span class="string">"MD"</span>, <span class="string">"YS"</span>, <span class="string">"YT"</span>)</span><br><span class="line"><span class="comment">## files will be output into outPath</span></span><br><span class="line">outPath &lt;- <span class="string">"splited"</span></span><br><span class="line">dir.create(outPath)</span><br><span class="line"><span class="comment">## shift the bam file by the 5'ends</span></span><br><span class="line"><span class="keyword">library</span>(BSgenome.Hsapiens.UCSC.hg19)</span><br><span class="line">seqlev &lt;- <span class="string">"chr1"</span> <span class="comment">## subsample data for quick run</span></span><br><span class="line">which &lt;- as(seqinfo(Hsapiens)[seqlev], <span class="string">"GRanges"</span>)</span><br><span class="line">gal &lt;- readBamFile(bamfile, tag=tags, which=which, asMates=<span class="literal">TRUE</span>)</span><br><span class="line">gal1 &lt;- shiftGAlignmentsList(gal)</span><br><span class="line">shiftedBamfile &lt;- file.path(outPath, <span class="string">"shifted.bam"</span>)</span><br><span class="line">export(gal1, shiftedBamfile)</span><br></pre></td></tr></table></figure>
<h3 id="peak-calling-using-macs2"><a class="markdownIt-Anchor" href="#peak-calling-using-macs2"></a> Peak calling using MACS2</h3>
<p>Ref: <a href="https://www.biostars.org/p/265061/#265063" target="_blank" rel="noopener">Biostars - ATACseq with STAR and macs2</a></p>
<blockquote>
<p>You typically use the <code>--nomodel</code> option, as the shifting model of MACS does not really make sense for open chromation data. As you have probably paired-end data, also use the <code>-f BAMPE</code> option. It forces MACS to pileup the real fragment length instead of an estimate, which maked sense imho, due to the quiet different fragment sizes that the library prep creates, also if you have paired-end why not make full use of it. Check reproducibility of the peaks between replicate samples, then rerun MACS with the merged bam file and feed the count matrix into DESeq2. Reference e.g. Corces et al 2016 Nat Genetics.</p>
</blockquote>
<p>Ref: <a href="https://github.com/taoliu/MACS/issues/145" target="_blank" rel="noopener">ATAC-seq settings · Issue #145 · taoliu/MACS</a></p>
<blockquote>
<p>Liu Tao: If you followed original protocol for ATAC-Seq, you should get Paired-End reads. If so, I would suggest you just use <code>--format BAMPE</code> to let MACS2 pileup the whole fragments in general. But if you want to focus on looking for where the ‘cutting sites’ are, then <code>--nomodel --shift -100 --extsize 200</code> should work.</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> -f BAMPE</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> --keep-dup all, keep all duplicate reads.</span></span><br><span class="line">macs2 callpeak -f BAMPE -g hs --keep-dup all --cutoff-analysis -n sample1 \</span><br><span class="line">  -t sample1.shifted.bam --outdir macs2/sample1 2&gt; macs2.log</span><br></pre></td></tr></table></figure>
<h3 id="creating-browser-tracks"><a class="markdownIt-Anchor" href="#creating-browser-tracks"></a> Creating browser tracks</h3>
<p>If <code>-B</code> parameter was used when running <code>macs2 callpeak</code>, you would get bedGraph files together with narrowPeak files. Someone would use these bedGraph files to create browser tracks (e.g. <a href="https://github.com/ParkerLab/bioinf545" target="_blank" rel="noopener">ParkerLab - ATAC-seq lab for BIOINF545</a>), while others say they look kind of weird (<a href="https://www.biostars.org/p/325946/#325951" target="_blank" rel="noopener">Biostars - How to compare bigwig tracks of two ATAC libraries</a>). I do not know yet.</p>
<p>I would like to create bigWig files for visualizing using <code>bamCoverage</code> in <code>deepTools</code>. It provides several different ways to normalize the signal.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> bam to bigwig, normalize using 1x effective genome size</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> effective genome size: https://deeptools.readthedocs.io/en/latest/content/feature/effectiveGenomeSize.html</span></span><br><span class="line">bamCoverage --numberOfProcessors 8 --binSize 10 --normalizeUsing RPGC \</span><br><span class="line">  --effectiveGenomeSize $effect_genome_size --bam sample1.shifted.bam -o sample1.shifted.bw</span><br></pre></td></tr></table></figure>
<h3 id="quality-check"><a class="markdownIt-Anchor" href="#quality-check"></a> Quality check</h3>
<p>In processing ATAC-seq data, we would get several metrics to check the quality of data/libraries.</p>
<p>Ref:</p>
<ul>
<li><a href="https://www.encodeproject.org/data-standards/terms/#enrichment" target="_blank" rel="noopener">ENCODE - Terms and Definitions</a></li>
<li><a href="https://www.encodeproject.org/atac-seq/" target="_blank" rel="noopener">ENCODE - ATAC-seq Data Standards and Prototype Processing Pipeline</a></li>
</ul>
<p>Here is the standards used by ENCODE, and the detailed description of each term will be explained below.</p>
<img src="/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/20190111215229195_23911.png" title="ENCODE ATAC-seq standards (accessed: 20190111)">
<h4 id="fragmentinsert-size"><a class="markdownIt-Anchor" href="#fragmentinsert-size"></a> Fragment/Insert size</h4>
<p>Ref: <a href="https://dbrg77.wordpress.com/2017/02/10/atac-seq-insert-size-plotting/" target="_blank" rel="noopener">Not A Rocket Scientist - ATAC-seq insert size plotting</a></p>
<blockquote>
<p>One common QC for the data is to plot the fragment size density of your libraries. The successful construction of a ATAC library requires a proper pair of Tn5 transposase cutting events at the ends of DNA. In the nucleosome-free open chromatin regions, many molecules of Tn5 can kick in and chop the DNA into small pieces; around nucleosome-occupied regions, Tn5 can only access the linker regions. Therefore, in a normal ATAC-seq library, you should expect to see a sharp peak at the &lt;100 bp region (open chromatin), and a peak at ~200bp region (mono-nucleosome), and other larger peaks (multi-nucleosomes). Examples from one of my data:</p>
<p>Regular scale:</p>
<img src="/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/20190103170441947_28541.png">
<p>Log scale:</p>
<img src="/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/20190103170526610_9297.png">
<p>This clear nucleosome phasing pattern indicates a good quality of the experiment.</p>
<p>Different people probably have different way of plotting this using different codes, but making this plot can be simply achieved by a combination of one line of code and Excel. You don’t need some complicated scripts to do it at all.</p>
<p><code>samtools view ATAC_f2q30_sorted.bam | awk '$9&gt;0' | cut -f 9 | sort | uniq -c | sort -b -k2,2n | sed -e 's/^[ \t]*//' &gt; fragment_length_count.txt</code></p>
</blockquote>
<p>See discussions here: <a href="https://www.biostars.org/p/332440/" target="_blank" rel="noopener">Biostars - ATAC-seq fragment length distribution</a></p>
<h4 id="mitochondrial-reads-2"><a class="markdownIt-Anchor" href="#mitochondrial-reads-2"></a> %mitochondrial reads</h4>
<p>High mitochondrial reads was a well-known problem, but in the latest ATAC-seq protocol (Omni-ATAC) this problem has been well addressed (<a href="https://www.nature.com/articles/nmeth.4396" target="_blank" rel="noopener">Corces et al., 2017</a>).</p>
<p>I guess in the future people will not care about this any more, and I keep this section in case.</p>
<p>Ref: <a href="https://github.com/harvardinformatics/ATAC-seq" target="_blank" rel="noopener">Harvard FAS Informatics - ATAC-seq Guidelines</a></p>
<blockquote>
<p>It is a well-known problem that ATAC-seq datasets usually contain a large percentage of reads that is derived from mitochondrial DNA (for example, see <a href="http://seqanswers.com/forums/showthread.php?t=35318" target="_blank" rel="noopener">this discussion</a>). Some have gone as far as <a href="https://www.nature.com/articles/s41598-017-02547-w" target="_blank" rel="noopener">using CRISPR to reduce mitochondrial contamination</a>. The recently published <a href="https://www.nature.com/articles/nmeth.4396" target="_blank" rel="noopener">Omni-ATAC method</a> uses detergents to remove mitochondria and is likely to be more accessible for most researchers (but, <a href="https://www.biorxiv.org/content/early/2018/12/17/496521" target="_blank" rel="noopener">do <strong>not</strong> follow their computational workflow</a>).</p>
</blockquote>
<p>The following code can be used to compute the percentage of mitochondrial reads.</p>
<p>Ref: <a href="https://www.biostars.org/p/170294/#196173" target="_blank" rel="noopener">Biostars - ATACseq alignment issues</a></p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash">!/bin/bash</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">### Calculate percentage of reads mapped to mitochondrial genome (mtDNA) using SAMtools idxstats</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">### Can be useful for ATAC-seq data. Requires an indexed BAM file:</span></span></span><br><span class="line"></span><br><span class="line">if [[ $# -eq 0 ]] ; then</span><br><span class="line">  echo '[ERROR]: No input file given!'</span><br><span class="line">  exit 1</span><br><span class="line">fi</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Check if index is present. If not, create it:</span></span></span><br><span class="line">if [[ ! -e $&#123;1&#125;.bai ]];</span><br><span class="line">  then</span><br><span class="line">  echo '[INFO]: File does not seem to be indexed. Indexing now:'</span><br><span class="line">  samtools index $i</span><br><span class="line">  fi</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Calculate %mtDNA:</span></span></span><br><span class="line">mtReads=$(samtools idxstats $1 | grep 'chrM' | cut -f 3)</span><br><span class="line">totalReads=$(samtools idxstats $1 | awk '&#123;SUM += $3&#125; END &#123;print SUM&#125;')</span><br><span class="line"></span><br><span class="line">echo '==&gt; mtDNA Content:' $(bc &lt;&lt;&lt; "scale=2;100*$mtReads/$totalReads")'%'</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage: ./script.sh atacseq.bam</span></span></span><br></pre></td></tr></table></figure>
<h4 id="library-complexity"><a class="markdownIt-Anchor" href="#library-complexity"></a> Library complexity</h4>
<p>Ref: <a href="https://www.encodeproject.org/atac-seq/" target="_blank" rel="noopener">ENCODE - ATAC-seq Data Standards and Prototype Processing Pipeline</a></p>
<blockquote>
<p>Library complexity is measured using the <a href="https://www.encodeproject.org/data-standards/terms/#library" target="_blank" rel="noopener">Non-Redundant Fraction (NRF) and PCR Bottlenecking Coefficients 1 and 2, or PBC1 and PBC2</a>. The preferred values are as follows: NRF&gt;0.9, PBC1&gt;0.9, and PBC2&gt;3.</p>
</blockquote>
<img src="/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/20190318163604339_31625.png">
<p>The following code was from <a href="https://github.com/ENCODE-DCC/atac-seq-pipeline" target="_blank" rel="noopener">ENCODE ATAC-seq pipeline</a>.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> the bam used here is sorted bam after duplicates marking</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> sort bam by names</span></span><br><span class="line">samtools sort -@ 10 -n -O BAM -o tmp.bam $&#123;sample&#125;.bam</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> calculate PBC metrics</span></span><br><span class="line">bedtools bamtobed -bedpe -i tmp.bam | awk 'BEGIN&#123;OFS="\t"&#125;&#123;print $1,$2,$4,$6,$9,$10&#125;' \</span><br><span class="line">  | grep -v 'chrM' | sort | uniq -c | awk 'BEGIN&#123;mt=0;m0=0;m1=0;m2=0&#125;($1==1)&#123;m1=m1+1&#125; \</span><br><span class="line"><span class="meta">  ($</span><span class="bash">1==2)&#123;m2=m2+1&#125; &#123;m0=m0+1&#125; &#123;mt=mt+<span class="variable">$1</span>&#125; \</span></span><br><span class="line">  END&#123;printf "%d\t%d\t%d\t%d\t%f\t%f\t%f\n", mt,m0,m1,m2,m0/mt,m1/m0,m1/m2&#125;' &gt; $&#123;sample&#125;.pbc.qc</span><br><span class="line">rm tmp.bam</span><br></pre></td></tr></table></figure>
<h4 id="fraction-of-reads-in-peaks-frip"><a class="markdownIt-Anchor" href="#fraction-of-reads-in-peaks-frip"></a> Fraction of reads in peaks (FRiP)</h4>
<p>Ref: <a href="https://www.encodeproject.org/data-standards/terms/#enrichment" target="_blank" rel="noopener">ENCODE - Terms and Definitions</a></p>
<blockquote>
<p>Fraction of reads in peaks (FRiP) - Fraction of all mapped reads that fall into the called peak regions, i.e. usable reads in significantly enriched peaks divided by all usable reads. In general, FRiP scores correlate positively with the number of regions. (Landt et al, Genome Research Sept. 2012, 22(9): 1813–1831)</p>
</blockquote>
<p>Ref: <a href="https://www.biostars.org/p/337872/#356888" target="_blank" rel="noopener">Biostars - FRIP score ATAC-seq</a></p>
<blockquote>
<p>In paired-end sequencing, we use the word fragment because the two reads that are produced always originate from the same DNA fragment and are therefore not independent of each other as reads from single-end sequencing would be. As FRiP comes from single-end ChIP-seq data, this is why they probably termed it reads. ATAC-seq is most commonly paired-end. You can use BEDtools for paired-end data but it requires more pre-processing of your data, that is why I use featureCounts, being faster and more convinient with plenty of customizable options. Choice is still yours. FRiP is probably not a very objective measure anyway, as it highly depends on how you prefilter your data, e.g. in terms of mapping quality, the definition of a properly-paired reads and the stringency of your peak calling (last sentence thinking aloud).</p>
</blockquote>
<p>Ref: <a href="https://www.encodeproject.org/atac-seq/" target="_blank" rel="noopener">ENCODE - ATAC-seq Data Standards and Prototype Processing Pipeline</a></p>
<blockquote>
<p>The fraction of reads in called peak regions (<a href="https://www.encodeproject.org/data-standards/terms/#enrichment" target="_blank" rel="noopener">FRiP score</a>) should be &gt;0.3, though values greater than 0.2 are acceptable. For EN-TEx tissues (ENCODE GTEx tissue sample), FRiP scores will not be enforced as QC metric. TSS enrichment remains in place as a key signal to noise measure.</p>
</blockquote>
<p>FRiP score can be calculated by <code>samtools</code> and <code>bedtools</code>:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> total reads</span></span><br><span class="line">total_reads=$(samtools view -c $&#123;sample&#125;.bam)</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> reads <span class="keyword">in</span> peaks</span></span><br><span class="line">reads_in_peaks=$(bedtools sort -i $&#123;sample&#125;_peaks.narrowPeak \</span><br><span class="line">  | bedtools merge -i stdin | bedtools intersect -u -nonamecheck \</span><br><span class="line">  -a $&#123;sample&#125;.bam -b stdin -ubam | samtools view -c)</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> FRiP score</span></span><br><span class="line">FRiP=$(awk "BEGIN &#123;print "$&#123;reads_in_peaks&#125;"/"$&#123;total_reads&#125;"&#125;")</span><br></pre></td></tr></table></figure>
<p>While someone recommended to use <code>featureCounts</code> (<a href="https://www.biostars.org/p/337872/#337890" target="_blank" rel="noopener">https://www.biostars.org/p/337872/#337890</a>), I found the results of <code>featureCounts</code> and <code>intersectBed</code> were close. So, I would like to use <code>bedtools</code>.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">samtools view -c $&#123;sample&#125;.bam</span><br><span class="line">38439210</span><br><span class="line"></span><br><span class="line">bedtools sort -i $&#123;sample&#125;_peaks.narrowPeak \</span><br><span class="line">  | bedtools merge -i stdin | bedtools intersect -u -nonamecheck \</span><br><span class="line">  -a $&#123;sample&#125;.bam -b stdin -ubam | samtools view -c</span><br><span class="line">428359</span><br><span class="line"></span><br><span class="line">awk 'BEGIN&#123;FS=OFS="\t"; print "GeneID\tChr\tStart\tEnd\tStrand"&#125; \</span><br><span class="line">  &#123;print $4, $1, $2+1, $3, "."&#125;' $&#123;sample&#125;_peaks.narrowPeak &gt; $&#123;sample&#125;_peaks.saf</span><br><span class="line">featureCounts -p -a M0.A.005_peaks.saf -F SAF -o readCountInPeaks.txt $&#123;sample&#125;.bam</span><br><span class="line">||    Total alignments : 19219605                                             ||</span><br><span class="line">||    Successfully assigned alignments : 230826 (1.2%)                        ||</span><br><span class="line"></span><br><span class="line"><span class="meta">&gt;</span><span class="bash"> 428359/38439210</span></span><br><span class="line">[1] 0.0111438</span><br><span class="line"><span class="meta">&gt;</span><span class="bash"> 230826/19219605</span></span><br><span class="line">[1] 0.01200992</span><br></pre></td></tr></table></figure>
<p>Find more details in note: <a href="/blog/2019/03/Calculate-FRiP-score/" title="Calculate FRiP score">Calculate FRiP score</a>.</p>
<h4 id="transcription-start-site-tss-enrichment"><a class="markdownIt-Anchor" href="#transcription-start-site-tss-enrichment"></a> Transcription start site (TSS) enrichment</h4>
<p>Ref: <a href="https://www.encodeproject.org/atac-seq/" target="_blank" rel="noopener">ENCODE - ATAC-seq Data Standards and Prototype Processing Pipeline</a></p>
<blockquote>
<p><a href="https://www.encodeproject.org/data-standards/terms/#enrichment" target="_blank" rel="noopener">Transcription start site (TSS) enrichment</a> values are dependent on the reference files used; cutoff values for high quality data are listed in the table below.</p>
</blockquote>
<p>Ref: <a href="https://www.encodeproject.org/data-standards/terms/#enrichment" target="_blank" rel="noopener">ENCODE - Terms and Definitions</a></p>
<blockquote>
<p><strong>Transcription Start Site (TSS) Enrichment Score</strong> - The TSS enrichment calculation is a signal to noise calculation. The reads around a reference set of TSSs are collected to form an aggregate distribution of reads centered on the TSSs and extending to 1000 bp in either direction (for a total of 2000bp). This distribution is then normalized by taking the average read depth in the 100 bps at each of the end flanks of the distribution (for a total of 200bp of averaged data) and calculating a fold change at each position over that average read depth. This means that the flanks should start at 1, and if there is high read signal at transcription start sites (highly open regions of the genome) there should be an increase in signal up to a peak in the middle. We take the signal value at the center of the distribution after this normalization as our TSS enrichment metric. <strong>Used to evaluate ATAC-seq.</strong></p>
</blockquote>
<p>The following code was from <a href="https://github.com/kundajelab/ataqc" target="_blank" rel="noopener">ATAqC</a>.</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> matplotlib</span><br><span class="line">matplotlib.use(<span class="string">'Agg'</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> pybedtools</span><br><span class="line"><span class="keyword">import</span> metaseq</span><br><span class="line"><span class="keyword">import</span> logging</span><br><span class="line"></span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> mlab</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">make_tss_plot</span><span class="params">(bam_file, tss, prefix, chromsizes, read_len, bins=<span class="number">400</span>, bp_edge=<span class="number">2000</span>,</span></span></span><br><span class="line"><span class="function"><span class="params">                  processes=<span class="number">8</span>, greenleaf_norm=True)</span>:</span></span><br><span class="line">    <span class="string">'''</span></span><br><span class="line"><span class="string">    Take bootstraps, generate tss plots, and get a mean and</span></span><br><span class="line"><span class="string">    standard deviation on the plot. Produces 2 plots. One is the</span></span><br><span class="line"><span class="string">    aggregation plot alone, while the other also shows the signal</span></span><br><span class="line"><span class="string">    at each TSS ordered by strength.</span></span><br><span class="line"><span class="string">    '''</span></span><br><span class="line">    logging.info(<span class="string">'Generating tss plot...'</span>)</span><br><span class="line">    tss_plot_file = <span class="string">'&#123;0&#125;_tss-enrich.pdf'</span>.format(prefix)</span><br><span class="line">    tss_plot_large_file = <span class="string">'&#123;0&#125;_large_tss-enrich.pdf'</span>.format(prefix)</span><br><span class="line"></span><br><span class="line">    <span class="comment"># Load the TSS file</span></span><br><span class="line">    tss = pybedtools.BedTool(tss)</span><br><span class="line">    tss_ext = tss.slop(b=bp_edge, g=chromsizes)</span><br><span class="line"></span><br><span class="line">    <span class="comment"># Load the bam file</span></span><br><span class="line">    bam = metaseq.genomic_signal(bam_file, <span class="string">'bam'</span>) <span class="comment"># Need to shift reads and just get ends, just load bed file?</span></span><br><span class="line">    bam_array = bam.array(tss_ext, bins=bins, shift_width = -read_len/<span class="number">2</span>, <span class="comment"># Shift to center the read on the cut site</span></span><br><span class="line">                          processes=processes, stranded=<span class="keyword">True</span>)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">if</span> greenleaf_norm:</span><br><span class="line">        <span class="comment"># Use enough bins to cover 100 bp on either end</span></span><br><span class="line">        num_edge_bins = int(<span class="number">100</span>/(<span class="number">2</span>*bp_edge/bins))</span><br><span class="line">        bin_means = bam_array.mean(axis=<span class="number">0</span>)</span><br><span class="line">        avg_noise = (sum(bin_means[:num_edge_bins]) +</span><br><span class="line">                     sum(bin_means[-num_edge_bins:]))/(<span class="number">2</span>*num_edge_bins)</span><br><span class="line">        bam_array /= avg_noise</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        bam_array /= bam.mapped_read_count() / <span class="number">1e6</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># Generate a line plot</span></span><br><span class="line">    fig = plt.figure()</span><br><span class="line">    ax = fig.add_subplot(<span class="number">111</span>)</span><br><span class="line">    x = np.linspace(-bp_edge, bp_edge, bins)</span><br><span class="line"></span><br><span class="line">    ax.plot(x, bam_array.mean(axis=<span class="number">0</span>), color=<span class="string">'r'</span>, label=<span class="string">'Mean'</span>)</span><br><span class="line">    ax.axvline(<span class="number">0</span>, linestyle=<span class="string">':'</span>, color=<span class="string">'k'</span>)</span><br><span class="line"></span><br><span class="line">    <span class="comment"># Note the middle high point (TSS)</span></span><br><span class="line">    tss_point_val = max(bam_array.mean(axis=<span class="number">0</span>))</span><br><span class="line"></span><br><span class="line">    ax.set_xlabel(<span class="string">'Distance from TSS (bp)'</span>)</span><br><span class="line">    ax.set_ylabel(<span class="string">'Average read coverage (per million mapped reads)'</span>)</span><br><span class="line">    ax.legend(loc=<span class="string">'best'</span>)</span><br><span class="line"></span><br><span class="line">    fig.savefig(tss_plot_file)</span><br><span class="line"></span><br><span class="line">    <span class="comment"># Print a more complicated plot with lots of info</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># Find a safe upper percentile - we can't use X if the Xth percentile is 0</span></span><br><span class="line">    upper_prct = <span class="number">99</span></span><br><span class="line">    <span class="keyword">if</span> mlab.prctile(bam_array.ravel(), upper_prct) == <span class="number">0.0</span>:</span><br><span class="line">        upper_prct = <span class="number">100.0</span></span><br><span class="line"></span><br><span class="line">    plt.rcParams[<span class="string">'font.size'</span>] = <span class="number">8</span></span><br><span class="line">    fig = metaseq.plotutils.imshow(bam_array,</span><br><span class="line">                                   x=x,</span><br><span class="line">                                   figsize=(<span class="number">5</span>, <span class="number">10</span>),</span><br><span class="line">                                   vmin=<span class="number">5</span>, vmax=upper_prct, percentile=<span class="keyword">True</span>,</span><br><span class="line">                                   line_kwargs=dict(color=<span class="string">'k'</span>, label=<span class="string">'All'</span>),</span><br><span class="line">                                   fill_kwargs=dict(color=<span class="string">'k'</span>, alpha=<span class="number">0.3</span>),</span><br><span class="line">                                   sort_by=bam_array.mean(axis=<span class="number">1</span>))</span><br><span class="line"></span><br><span class="line">    <span class="comment"># And save the file</span></span><br><span class="line">    fig.savefig(tss_plot_large_file)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">return</span> tss_plot_file, tss_plot_large_file, tss_point_val</span><br></pre></td></tr></table></figure>
<h3 id="blacklist-filtering-for-peaks"><a class="markdownIt-Anchor" href="#blacklist-filtering-for-peaks"></a> Blacklist filtering for peaks</h3>
<p>One may want to filter peaks using <a href="https://www.encodeproject.org/annotations/ENCSR636HFF/" target="_blank" rel="noopener">DAC Blacklisted Regions</a>.</p>
<p>The following code was from <a href="https://www.encodeproject.org/atac-seq/" target="_blank" rel="noopener">ENCODE - ATAC-seq Data Standards and Prototype Processing Pipeline</a></p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">PEAK="$&#123;PREFIX&#125;.narrowPeak"</span><br><span class="line">FILTERED_PEAK="$&#123;PREFIX&#125;.narrowPeak.filt.gz"</span><br><span class="line">bedtools intersect -v -a $&#123;PEAK&#125; -b $&#123;BLACKLIST&#125; \</span><br><span class="line">  | awk 'BEGIN&#123;OFS="\t"&#125; &#123;if ($5&gt;1000) $5=1000; print $0&#125;' \</span><br><span class="line">  | grep -P 'chr[0-9XY]+(?!_)' | gzip -nc &gt; $&#123;FILTERED_PEAK&#125;</span><br></pre></td></tr></table></figure>
<p>The 5th column of narroPeak file is integer score for display calculated as int(-10*log10qvalue). Since currently this value might be out of the [0-1000] range defined in <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format12" target="_blank" rel="noopener">UCSC Encode narrowPeak format</a>, the <code>awk</code> code is to assign values greater than 1000 to 1000.</p>
<h3 id="merging-peaks-optional"><a class="markdownIt-Anchor" href="#merging-peaks-optional"></a> Merging peaks (optional)</h3>
<p>One may want to merge peaks from different libraries or different samples.</p>
<p><a href="https://www.nature.com/articles/ng.3646" target="_blank" rel="noopener">Corces et al., 2016</a> used the following method to merge peaks, and I like their way:</p>
<blockquote>
<p>To generate a non-redundant list of hematopoiesis- and cancer-related peaks, we first extended summits to 500-bp windows (±250 bp). We then ranked the 500-bp peaks by summit significance value (defined by MACS2) and chose a list of non-overlapping, maximally significant peaks.</p>
</blockquote>
<p>We can do this by using <code>BEDOPS</code> (ref: <a href="https://bedops.readthedocs.io/en/latest/content/usage-examples/master-list.html" target="_blank" rel="noopener">Collapsing multiple BED files into a master list by signal</a>)</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash">!/bin/bash</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> modified from https://bedops.readthedocs.io/en/latest/content/usage-examples/master-list.html</span></span><br><span class="line"></span><br><span class="line">summit_bed=(sample1_summits.bed sample2_summits.bed sample3_summits.bed)</span><br><span class="line"></span><br><span class="line">out=fAdrenal.master.merge.bed</span><br><span class="line"></span><br><span class="line">tmpd=/tmp/tmp$$</span><br><span class="line">mkdir -p $tmpd</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># First, union all the peaks together into a single file.</span></span></span><br><span class="line">bedlist=""</span><br><span class="line">for bed in $&#123;beds[*]&#125;</span><br><span class="line">do</span><br><span class="line">    bedlist="$bedlist $summit_bed"</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> extended summits to 500-bp windows (±250 bp)</span></span><br><span class="line">bedops --range 250 -u $bedlist &gt; $tmpd/tmp.bed</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># The master list is constructed iteratively.  For each pass through</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># the loop, elements not yet in the master list are merged into</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># non-overlapping intervals that span the union (this is just bedops</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># -m).  Then for each merged interval, an original element of highest</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># score within the interval is selected to go in the master list.</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Anything that overlaps the selected element is thrown out, and the</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># process then repeats.</span></span></span><br><span class="line">iters=1</span><br><span class="line">solns=""</span><br><span class="line">stop=0</span><br><span class="line">while [ $stop == 0 ]</span><br><span class="line">do</span><br><span class="line">    echo "merge steps..."</span><br><span class="line"></span><br><span class="line">    ## Condense the union into merged intervals. This klugey bit</span><br><span class="line">    ## before and after the merging is because we don't want to merge</span><br><span class="line">    ## regions that are simply adjacent but not overlapping</span><br><span class="line">    bedops -m --range 0:-1 $tmpd/tmp.bed \</span><br><span class="line">        | bedops -u --range 0:1 - \</span><br><span class="line">        &gt; $tmpd/tmpm.bed</span><br><span class="line"></span><br><span class="line">    ## Grab the element with the highest score among all elements forming each interval.</span><br><span class="line">    ## If multiple elements tie for the highest score, just grab one of them.</span><br><span class="line">    ## Result is the current master list.  Probably don't need to sort, but do it anyway</span><br><span class="line">    ## to be safe since we're not using --echo with bedmap call.</span><br><span class="line">    bedmap --max-element $tmpd/tmpm.bed $tmpd/tmp.bed \</span><br><span class="line">        | sort-bed - \</span><br><span class="line">        &gt; $tmpd/$iters.bed</span><br><span class="line">    solns="$solns $tmpd/$iters.bed"</span><br><span class="line">    echo "Adding `awk 'END &#123; print NR &#125;' $tmpd/$iters.bed` elements"</span><br><span class="line"></span><br><span class="line">    ## Are there any elements that don't overlap the current master</span><br><span class="line">    ## list?  If so, add those in, and repeat.  If not, we're done.</span><br><span class="line">    bedops -n 1 $tmpd/tmp.bed $tmpd/$iters.bed \</span><br><span class="line">       &gt; $tmpd/tmp2.bed</span><br><span class="line"></span><br><span class="line">    mv $tmpd/tmp2.bed $tmpd/tmp.bed</span><br><span class="line"></span><br><span class="line">    if [ ! -s $tmpd/tmp.bed ]</span><br><span class="line">    then</span><br><span class="line">        stop=1</span><br><span class="line">    fi</span><br><span class="line"></span><br><span class="line">    ((iters++))</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># final solution</span></span></span><br><span class="line">bedops -u $solns \</span><br><span class="line"><span class="meta">   &gt;</span><span class="bash"> <span class="variable">$out</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Clean up</span></span></span><br><span class="line">rm -r $tmpd</span><br><span class="line"></span><br><span class="line">exit 0</span><br></pre></td></tr></table></figure>
<h3 id="outreach"><a class="markdownIt-Anchor" href="#outreach"></a> Outreach</h3>
<p>Well, then we complete the main upstream ATAC-seq data analysis. For downstream analysis, one may want to annotate peaks, find enriched motifs of TFs, compare peaks under different conditions and combine ATAC-seq data with other data types like RNA-seq etc.</p>
<h2 id="change-log"><a class="markdownIt-Anchor" href="#change-log"></a> Change log</h2>
<ul>
<li>20190103: create the note.</li>
<li>20190320: complete the note.</li>
<li>20190401: update the section of “Merging BAMs”. Add some discussions about “pooling replicates”</li>
</ul>

      
    </div>

    

    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/blog/tags/peak/" rel="tag"># peak</a>
          
            <a href="/blog/tags/ATAC-seq/" rel="tag"># ATAC-seq</a>
          
            <a href="/blog/tags/MACS/" rel="tag"># MACS</a>
          
            <a href="/blog/tags/bowtie2/" rel="tag"># bowtie2</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/blog/2019/03/Calculate-FRiP-score/" rel="next" title="Calculate FRiP score">
                <i class="fa fa-chevron-left"></i> Calculate FRiP score
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>


  </div>


          </div>
          

  
    <div class="comments" id="comments">
      <div id="disqus_thread">
        <noscript>
          Please enable JavaScript to view the
          <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a>
        </noscript>
      </div>
    </div>

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            Overview
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope="" itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image" src="/blog/images/portrait.jpg" alt="Yiwei Niu">
            
              <p class="site-author-name" itemprop="name">Yiwei Niu</p>
              <p class="site-description motion-element" itemprop="description">a beginner, a learner</p>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/blog/archives">
                
                    <span class="site-state-item-count">23</span>
                    <span class="site-state-item-name">posts</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-categories">
                  <a href="/blog/categories/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">22</span>
                    <span class="site-state-item-name">categories</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-tags">
                  <a href="/blog/tags/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">44</span>
                    <span class="site-state-item-name">tags</span>
                  </a>
                </div>
              
            </nav>
          

          
            <div class="feed-link motion-element">
              <a href="/blog/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          
            <div class="links-of-author motion-element">
              
                <span class="links-of-author-item">
                  <a href="mailto:xiaohuwangwang@qq.com" target="_blank" title="E-Mail"><i class="fa fa-fw fa-envelope"></i>E-Mail</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="https://github.com/YiweiNiu" target="_blank" title="GitHub"><i class="fa fa-fw fa-github"></i>GitHub</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="http://weibo.com/u/3049858680" target="_blank" title="Weibo"><i class="fa fa-fw fa-weibo"></i>Weibo</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="https://twitter.com/ywniu" target="_blank" title="Twitter"><i class="fa fa-fw fa-twitter"></i>Twitter</a>
                  
                </span>
              
            </div>
          

          
          

          
          
            <div class="links-of-blogroll motion-element links-of-blogroll-block">
              <div class="links-of-blogroll-title">
                <i class="fa  fa-fw fa-link"></i>
                Links
              </div>
              <ul class="links-of-blogroll-list">
                
                  <li class="links-of-blogroll-item">
                    <a href="https://kaigedong.github.io/" title="Kaige Dong" target="_blank">Kaige Dong</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://homolog.us/blogs/" title="Homolog.us" target="_blank">Homolog.us</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="https://cgatoxford.wordpress.com/" title="CGAT" target="_blank">CGAT</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://www.gettinggeneticsdone.com/" title="Getting Genetics Done" target="_blank">Getting Genetics Done</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://www.acgt.me/" title="ACGT" target="_blank">ACGT</a>
                  </li>
                
              </ul>
            </div>
          

          
            
          
          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#atac-seq-overview"><span class="nav-number">1.</span> <span class="nav-text"> ATAC-seq overview</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#experimental-design"><span class="nav-number">2.</span> <span class="nav-text"> Experimental design</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#data-analysis"><span class="nav-number">3.</span> <span class="nav-text"> Data analysis</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#quality-control"><span class="nav-number">3.1.</span> <span class="nav-text"> Quality control</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#alignment-and-filter"><span class="nav-number">3.2.</span> <span class="nav-text"> Alignment and filter</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#mitochondrial-reads"><span class="nav-number">3.2.1.</span> <span class="nav-text"> Mitochondrial reads</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#pcr-duplicates"><span class="nav-number">3.2.2.</span> <span class="nav-text"> PCR duplicates</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#non-unique-alignments"><span class="nav-number">3.2.3.</span> <span class="nav-text"> Non-unique alignments</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#others"><span class="nav-number">3.2.4.</span> <span class="nav-text"> Others</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#merging-bams-optional"><span class="nav-number">3.2.5.</span> <span class="nav-text"> Merging BAMs (optional)</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#shifting-reads"><span class="nav-number">3.3.</span> <span class="nav-text"> Shifting reads</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#peak-calling-using-macs2"><span class="nav-number">3.4.</span> <span class="nav-text"> Peak calling using MACS2</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#creating-browser-tracks"><span class="nav-number">3.5.</span> <span class="nav-text"> Creating browser tracks</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#quality-check"><span class="nav-number">3.6.</span> <span class="nav-text"> Quality check</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#fragmentinsert-size"><span class="nav-number">3.6.1.</span> <span class="nav-text"> Fragment/Insert size</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#mitochondrial-reads-2"><span class="nav-number">3.6.2.</span> <span class="nav-text"> %mitochondrial reads</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#library-complexity"><span class="nav-number">3.6.3.</span> <span class="nav-text"> Library complexity</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#fraction-of-reads-in-peaks-frip"><span class="nav-number">3.6.4.</span> <span class="nav-text"> Fraction of reads in peaks (FRiP)</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#transcription-start-site-tss-enrichment"><span class="nav-number">3.6.5.</span> <span class="nav-text"> Transcription start site (TSS) enrichment</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#blacklist-filtering-for-peaks"><span class="nav-number">3.7.</span> <span class="nav-text"> Blacklist filtering for peaks</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#merging-peaks-optional"><span class="nav-number">3.8.</span> <span class="nav-text"> Merging peaks (optional)</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#outreach"><span class="nav-number">3.9.</span> <span class="nav-text"> Outreach</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#change-log"><span class="nav-number">4.</span> <span class="nav-text"> Change log</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; 2017 &mdash; <span itemprop="copyrightYear">2019</span>
  <span class="with-love" id="animate">
    <i class="fa fa-mars"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Yiwei Niu</span>

  

  
</div>




  <div class="powered-by">Powered by <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a></div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">Theme &mdash; <a class="theme-link" target="_blank" href="https://theme-next.org">NexT.Gemini</a></div>




        








        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
          <span id="scrollpercent"><span>0</span>%</span>
        
      </div>
    

    
	
    

    
      <div>
        <!-- Go to www.addthis.com/dashboard to customize your tools -->
<div class="addthis_inline_share_toolbox">
  <script type="text/javascript" src="//s7.addthis.com/js/300/addthis_widget.js#pubid=ra-5a7ad07b7823d03b" async="async"></script>
</div>

      </div>
    
  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>














  













  
  
    <script type="text/javascript" src="/blog/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/canvas-nest/canvas-nest.min.js"></script>
  


  


  <script type="text/javascript" src="/blog/js/src/utils.js?v=6.3.0"></script>

  <script type="text/javascript" src="/blog/js/src/motion.js?v=6.3.0"></script>



  
  


  <script type="text/javascript" src="/blog/js/src/affix.js?v=6.3.0"></script>

  <script type="text/javascript" src="/blog/js/src/schemes/pisces.js?v=6.3.0"></script>



  
  <script type="text/javascript" src="/blog/js/src/scrollspy.js?v=6.3.0"></script>
<script type="text/javascript" src="/blog/js/src/post-details.js?v=6.3.0"></script>



  


  <script type="text/javascript" src="/blog/js/src/bootstrap.js?v=6.3.0"></script>



  

  
    <script id="dsq-count-scr" src="https://yiwei-nius-note.disqus.com/count.js" async></script>
  

  
    <script type="text/javascript">
      var disqus_config = function () {
        this.page.url = 'https://yiweiniu.github.io/blog/2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/';
        this.page.identifier = '2019/03/ATAC-seq-data-analysis-from-FASTQ-to-peaks/';
        this.page.title = 'ATAC-seq data analysis: from FASTQ to peaks';
        };
      function loadComments () {
        var d = document, s = d.createElement('script');
        s.src = 'https://yiwei-nius-note.disqus.com/embed.js';
        s.setAttribute('data-timestamp', '' + +new Date());
        (d.head || d.body).appendChild(s);
      }
      
        loadComments();
      
    </script>
  












  





  

  
  <script src="https://cdn1.lncld.net/static/js/av-core-mini-0.6.4.js"></script>
  <script>AV.initialize("sxYNFK15bXIJqayGqYdiJNsw-gzGzoHsz", "inuUVGRl3VjbS7oJLrS8uYMe");</script>
  <script>
    function showTime(Counter) {
      var query = new AV.Query(Counter);
      var entries = [];
      var $visitors = $(".leancloud_visitors");

      $visitors.each(function () {
        entries.push( $(this).attr("id").trim() );
      });

      query.containedIn('url', entries);
      query.find()
        .done(function (results) {
          var COUNT_CONTAINER_REF = '.leancloud-visitors-count';

          if (results.length === 0) {
            $visitors.find(COUNT_CONTAINER_REF).text(0);
            return;
          }

          for (var i = 0; i < results.length; i++) {
            var item = results[i];
            var url = item.get('url');
            var time = item.get('time');
            var element = document.getElementById(url);

            $(element).find(COUNT_CONTAINER_REF).text(time);
          }
          for(var i = 0; i < entries.length; i++) {
            var url = entries[i];
            var element = document.getElementById(url);
            var countSpan = $(element).find(COUNT_CONTAINER_REF);
            if( countSpan.text() == '') {
              countSpan.text(0);
            }
          }
        })
        .fail(function (object, error) {
          console.log("Error: " + error.code + " " + error.message);
        });
    }

    function addCount(Counter) {
      var $visitors = $(".leancloud_visitors");
      var url = $visitors.attr('id').trim();
      var title = $visitors.attr('data-flag-title').trim();
      var query = new AV.Query(Counter);

      query.equalTo("url", url);
      query.find({
        success: function(results) {
          if (results.length > 0) {
            var counter = results[0];
            counter.fetchWhenSave(true);
            counter.increment("time");
            
            counter.save(null, {
              success: function(counter) {
                
                  var $element = $(document.getElementById(url));
                  $element.find('.leancloud-visitors-count').text(counter.get('time'));
                
              },
              error: function(counter, error) {
                console.log('Failed to save Visitor num, with error message: ' + error.message);
              }
            });
          } else {
            
              var newcounter = new Counter();
              /* Set ACL */
              var acl = new AV.ACL();
              acl.setPublicReadAccess(true);
              acl.setPublicWriteAccess(true);
              newcounter.setACL(acl);
              /* End Set ACL */
              newcounter.set("title", title);
              newcounter.set("url", url);
              newcounter.set("time", 1);
              newcounter.save(null, {
                success: function(newcounter) {
                  var $element = $(document.getElementById(url));
                  $element.find('.leancloud-visitors-count').text(newcounter.get('time'));
                },
                error: function(newcounter, error) {
                  console.log('Failed to create');
                }
              });
            
          }
        },
        error: function(error) {
          console.log('Error:' + error.code + " " + error.message);
        }
      });
    }

    $(function() {
      var Counter = AV.Object.extend("Counter");
      if ($('.leancloud_visitors').length == 1) {
        addCount(Counter);
      } else if ($('.post-title-link').length > 1) {
        showTime(Counter);
      }
    });
  </script>



  

  

  

  
  

  

  

  

  

  

</body>
</html>
