<!DOCTYPE html>












  


<html class="theme-next gemini use-motion" lang="en">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2"/>
<meta name="theme-color" content="#222">












<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />



  <meta name="google-site-verification" content="3dnqtuxhGQK5nK7aqRtU2c3DzZlaVtBMN6M39UoGcgY" />




















<link href="/blog/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/blog/css/main.css?v=6.3.0" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/blog/images/apple-touch-icon-next.png?v=6.3.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/blog/images/favicon-32x32-next.png?v=6.3.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/blog/images/favicon-16x16-next.png?v=6.3.0">


  <link rel="mask-icon" href="/blog/images/logo.svg?v=6.3.0" color="#222">









<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/blog/',
    scheme: 'Gemini',
    version: '6.3.0',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":true,"onmobile":false},
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="Intro From its Git Repo  LINKS is a scalable genomics application for scaffolding or re-scaffolding genome assembly drafts with long reads, such as those produced by Oxford Nanopore Technologies Ltd">
<meta name="keywords" content="bio-tools,genome assembly pipeline,hybrid genome assembly,scaffolding">
<meta property="og:type" content="article">
<meta property="og:title" content="Genome Assembly Pipeline: LINKS">
<meta property="og:url" content="https://yiweiniu.github.io/blog/2018/03/Genome-assembly-pipeline-LINKS/index.html">
<meta property="og:site_name" content="Yiwei Niu&#39;s Note">
<meta property="og:description" content="Intro From its Git Repo  LINKS is a scalable genomics application for scaffolding or re-scaffolding genome assembly drafts with long reads, such as those produced by Oxford Nanopore Technologies Ltd">
<meta property="og:locale" content="en">
<meta property="og:updated_time" content="2018-08-18T03:32:58.644Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Genome Assembly Pipeline: LINKS">
<meta name="twitter:description" content="Intro From its Git Repo  LINKS is a scalable genomics application for scaffolding or re-scaffolding genome assembly drafts with long reads, such as those produced by Oxford Nanopore Technologies Ltd">



  <link rel="alternate" href="/blog/atom.xml" title="Yiwei Niu's Note" type="application/atom+xml" />




  <link rel="canonical" href="https://yiweiniu.github.io/blog/2018/03/Genome-assembly-pipeline-LINKS/"/>



<script type="text/javascript" id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>Genome Assembly Pipeline: LINKS | Yiwei Niu's Note</title>
  




<script async src="https://www.googletagmanager.com/gtag/js?id=UA-122074944-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-122074944-1');
</script>






  <noscript>
  <style type="text/css">
    .use-motion .motion-element,
    .use-motion .brand,
    .use-motion .menu-item,
    .sidebar-inner,
    .use-motion .post-block,
    .use-motion .pagination,
    .use-motion .comments,
    .use-motion .post-header,
    .use-motion .post-body,
    .use-motion .collection-title { opacity: initial; }

    .use-motion .logo,
    .use-motion .site-title,
    .use-motion .site-subtitle {
      opacity: initial;
      top: initial;
    }

    .use-motion {
      .logo-line-before i { left: initial; }
      .logo-line-after i { right: initial; }
    }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="en">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/blog/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Yiwei Niu's Note</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
      
        <p class="site-subtitle">to share, to learn</p>
      
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="Toggle navigation bar">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home">
    <a href="/blog/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-home"></i> <br />Home</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-categories">
    <a href="/blog/categories/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-th"></i> <br />Categories</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-tags">
    <a href="/blog/tags" rel="section">
      <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />Tags</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">
    <a href="/blog/archives" rel="section">
      <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />Archives</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-about">
    <a href="https:/yiweiniu.github.io/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-user"></i> <br />About</a>
  </li>

      
      
    </ul>
  

  

  
</nav>



  



</div>
    </header>

    


    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
            

          
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="https://yiweiniu.github.io/blog/blog/2018/03/Genome-assembly-pipeline-LINKS/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Yiwei Niu">
      <meta itemprop="description" content="a beginner, a learner">
      <meta itemprop="image" content="/blog/images/portrait.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Yiwei Niu's Note">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">Genome Assembly Pipeline: LINKS
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2018-03-30 00:07:28" itemprop="dateCreated datePublished" datetime="2018-03-30T00:07:28+08:00">2018-03-30</time>
            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/blog/categories/genome-assembly/" itemprop="url" rel="index"><span itemprop="name">genome assembly</span></a></span>

                
                
                  , 
                
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/blog/categories/genome-assembly/hybrid-pipeline/" itemprop="url" rel="index"><span itemprop="name">hybrid pipeline</span></a></span>

                
                
              
            </span>
          

          
            
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
                <a href="/blog/2018/03/Genome-assembly-pipeline-LINKS/#comments" itemprop="discussionUrl">
                
                  <span class="post-comments-count disqus-comment-count"
                        data-disqus-identifier="2018/03/Genome-assembly-pipeline-LINKS/" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          

          
            <span class="post-meta-divider">|</span>
            <span class="post-meta-item-icon"
            >
            <i class="fa fa-eye"></i>
             Views:  
            <span class="busuanzi-value" id="busuanzi_value_page_pv" ></span>
            </span>
          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                <span title="Symbols count in article">11k</span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                <span title="Reading time">0:10</span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="intro"><a class="markdownIt-Anchor" href="#intro"></a> Intro</h2>
<p>From its <a href="http://www.bcgsc.ca/platform/bioinfo/software/links" target="_blank" rel="noopener">Git Repo</a></p>
<blockquote>
<p>LINKS is a scalable genomics application for scaffolding or re-scaffolding genome assembly drafts with long reads, such as those produced by Oxford Nanopore Technologies Ltd and Pacific Biosciences. It provides a generic alignment-free framework for scaffolding and can work on any sequences. It is versatile and supports not only long sequences as a source of long-range information, but also MPET pairs and linked-reads, such as those from the 10X Genomics GemCode and Chromium platform, via ARCS (<a href="http://www.bcgsc.ca/platform/bioinfo/software/arcs" target="_blank" rel="noopener">http://www.bcgsc.ca/platform/bioinfo/software/arcs</a>). Fill gaps in LINKS-derived scaffolds using Sealer (<a href="http://www.bcgsc.ca/platform/bioinfo/software/sealer" target="_blank" rel="noopener">http://www.bcgsc.ca/platform/bioinfo/software/sealer</a>).</p>
</blockquote>
<p>Its paper:</p>
<blockquote>
<p>Warren RL, Yang C, Vandervalk BP, Behsaz B, Lagman A, Jones SJM, Birol I. LINKS: Scalable, alignment-free scaffolding of draft genomes with long reads. GigaScience. 2015;4:35. doi:10.1186/s13742-015-0076-3</p>
</blockquote>
<p>My feelings:</p>
<ul>
<li>easy to use</li>
<li>support scaffolding and re-scaffolding</li>
<li>fast</li>
<li>comparatively good results (at least in my case, compared to <code>BESST</code> and <code>OPERA-LG</code>)</li>
<li>huge RAM consumption</li>
</ul>
<h2 id="general-usage"><a class="markdownIt-Anchor" href="#general-usage"></a> General usage</h2>
<h3 id="install"><a class="markdownIt-Anchor" href="#install"></a> Install</h3>
<p>Dowload the latest version from its <a href="https://github.com/bcgsc/LINKS/releases" target="_blank" rel="noopener">release</a> page.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> decompress</span></span><br><span class="line">tar -zxvf links_v1-8-6.tar.gz; cd links_v1.8.6</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> build BloomFilter PERL module</span></span><br><span class="line">cd lib/bloomfilter/swig</span><br><span class="line">swig -Wall -c++ -perl5 BloomFilter.i</span><br><span class="line">g++ -c BloomFilter_wrap.cxx -I/usr/lib64/perl5/CORE -fPIC -Dbool=char -O3</span><br><span class="line">g++ -Wall -shared BloomFilter_wrap.o -o BloomFilter.so -O3</span><br></pre></td></tr></table></figure>
<h3 id="parameter-setting"><a class="markdownIt-Anchor" href="#parameter-setting"></a> Parameter setting</h3>
<p><a href="https://github.com/bcgsc/LINKS#how-it-works" target="_blank" rel="noopener">Here</a> gives a general description of the way <code>LINKS</code> works.</p>
<p>Setting the parameters is crucial for scaffolding. Below I summarize some tips or practices when setting these parameters.</p>
<ol>
<li>scaffolding control</li>
</ol>
<ul>
<li><code>-a</code>: the maximum ratio between the best two contig pairs for a given seed/contig being extended (default: 0.3).</li>
<li><code>-l</code>: the minimum number of links (read pairs) a valid contig pair MUST have to be considered (default: 5).</li>
<li>
<blockquote>
<p>For example, contig A shares 4 links with B and 2 links with C, in this orientation. contig rA (reverse) also shares 3 links with D. When it’s time to extend contig A (with the options -l and -a set to 2 and 0.7, respectively), both contig pairs AB and AC are considered. Since C (second-best) has 2 links and B (best) has 4 (2/4) = 0.5 below the maximum ratio of 0.7, A will be linked with B in the scaffold and C will be kept for another extension. If AC had 3 links the resulting ratio (0.75), above the user-defined maximum 0.7 would have caused the extension to terminate at A, with both B and C considered for a different scaffold. A maximum links ratio of 1 (not recommended) means that the best two candidate contig pairs have the same number of links – LINKS will accept the first one since both have a valid gap/overlap. When a scaffold extension is terminated on one side, the scaffold is extended on the “left”, by looking for contig pairs that involve the reverse of the seed (in this example, rAD). With AB and AC having 4 and 2 links, respectively and rAD being the only pair on the left, the final scaffolds outputted by LINKS would be: rD-A-B and C.</p>
</blockquote>
</li>
</ul>
<ol start="2">
<li>k-mer length: <code>-k</code></li>
</ol>
<ul>
<li><code>-k</code>: k-mer value (default 15).</li>
<li>LINKS is a k-mer scaffolder, and the <code>-k</code> parameter controls the k-mer length.</li>
<li>Exploration of vast kmer space is expected to yield better scaffolding results.</li>
<li>You may increase <code>-k</code> to <code>21</code> while working with pacbio reads.</li>
<li>I also recommend correcting ONT reads if you can, it will allow you to choose higher k values and increase the specificity.</li>
<li>The sweet spot will be somewhere k 15-19 for 2Gb genome (assuming raw ONT reads).</li>
</ul>
<ol start="3">
<li>k-mer pairs extraction: <code>-d</code>, <code>-e</code> and <code>-t</code>. And <code>-t</code> and <code>-d</code> are important for memory usage.</li>
</ol>
<ul>
<li><code>-e</code>: error (%) allowed on -d distance   e.g. -e 0.1  == distance +/- 10% (default: 0.1)
<ul>
<li>In theory, the -e parameter will play an important role limiting linkages outside of the target range -d (+/-) -e %. This is especially true when using raw MPET for scaffolding, to limit spurious linkages by contaminating PETs.</li>
</ul>
</li>
<li><code>-d</code>: distance between the 5’-end of each pairs (default: 4000)</li>
<li><code>-t</code>: sliding window when extracting k-mer pairs from long reads (default: 2)</li>
<li>Because you want want to start with a low -d for scaffolding, you have to estimate how many minimum links (-l) would fit in a -d window +/- error -e given sliding window -t. For instance, it may not make sense to use -t 200, -d 500 at low coverages BUT if you have at least 10-fold coverage it might since, in principle, you should be able to derive sufficient k-mer pairs within same locus if there’s no bias in genome sequencing.</li>
<li>On the data side of things, reducing the coverage (using less long reads), and limiting to only the highest quality reads would help decrease RAM usage.</li>
<li>WARNING: Specifying many distances will require large amount of RAM, especially with low -t values.</li>
<li>As <code>-d</code> increases, <code>-t</code> must decrease (otherwise you’ll end up with too few pairs for scaffolding over larger kmer distances).</li>
</ul>
<h4 id="an-example"><a class="markdownIt-Anchor" href="#an-example"></a> An example</h4>
<p>The power of <code>LINKS</code> is in scaffolding using various distance constraints, iteratively.</p>
<ol>
<li>Running links multiple times (when working with large long read dataset / genomes are very big)</li>
</ol>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">nohup ./runIterativeLINKS.sh 17 beluga.fa 5 0.3 &amp;</span><br><span class="line">./LINKS -f $2 -s ont.fof -b links1 -d 1000 -t 10 -k $1 -l $3 -a $4</span><br><span class="line">./LINKS -f links1.scaffolds.fa -s ont.fof -b links2 -d 2500 -t 5 -k $1 -l $3 -a $4 -o 1 -r links1.bloom</span><br><span class="line">./LINKS -f links2.scaffolds.fa -s ont.fof -b links3 -d 5000 -t 5 -k $1 -l $3 -a $4 -o 2 -r links1.bloom</span><br><span class="line">./LINKS -f links3.scaffolds.fa -s ont.fof -b links4 -d 7500 -t 4 -k $1 -l $3 -a $4 -o 3 -r links1.bloom</span><br><span class="line">./LINKS -f links4.scaffolds.fa -s ont.fof -b links5 -d 10000 -t 4 -k $1 -l $3 -a $4 -o 4 -r links1.bloom</span><br><span class="line">./LINKS -f links5.scaffolds.fa -s ont.fof -b links6 -d 12500 -t 3 -k $1 -l $3 -a $4 -o 5 -r links1.bloom</span><br><span class="line">./LINKS -f links6.scaffolds.fa -s ont.fof -b links7 -d 15000 -t 3 -k $1 -l $3 -a $4 -o 6 -r links1.bloom</span><br><span class="line">./LINKS -f links7.scaffolds.fa -s ont.fof -b links8 -d 30000 -t 2 -k $1 -l $3 -a $4 -o 7 -r links1.bloom</span><br></pre></td></tr></table></figure>
<ol start="2">
<li>Running links iteratively in a single command (works best when genomes are small/RAM not limiting)</li>
</ol>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">./LINKS -f $2 -s ont.fof -b links1 -d 1000,5000,7500,10000,12500,15000,30000 -t 10,5,5,4,4,3,3,2 -k $1 -l $3 -a $4</span><br></pre></td></tr></table></figure>
<ul>
<li>the value of <code>-t</code> will determine the #kmer pairs extracted and incidentally the RAM used. this will vary from one dataset to the next.</li>
</ul>
<p>The example was from: <a href="https://github.com/bcgsc/LINKS/issues/20" target="_blank" rel="noopener">working with 2 Gb genome; stuck on bloom filter being built</a>.</p>
<p>It seems <code>LINKS</code> needs huge RAM, and you may also want to try <a href="https://github.com/bcgsc/LINKS/issues/20" target="_blank" rel="noopener">lrscaf</a>, another long reads scaffolding tool.</p>
<h2 id="in-practice"><a class="markdownIt-Anchor" href="#in-practice"></a> In practice</h2>
<h3 id="background"><a class="markdownIt-Anchor" href="#background"></a> background</h3>
<ul>
<li>An insect</li>
<li>The species: high heterogeneity, high AT, high repetition.</li>
<li>Genome size: male 790M, female 830M.</li>
</ul>
<h3 id="data"><a class="markdownIt-Anchor" href="#data"></a> data</h3>
<p>The Illumina data I used:</p>
<table>
<thead>
<tr>
<th>Source</th>
<th>Insert size (bp)</th>
<th>Avg. read size (bp)</th>
<th>Raw bases (G)</th>
<th>Raw reads (M)</th>
<th>Sequencing depth</th>
</tr>
</thead>
<tbody>
<tr>
<td>AV1, M</td>
<td>270</td>
<td>150</td>
<td>44.1</td>
<td>293.6</td>
<td>55.5</td>
</tr>
<tr>
<td>AV2, F</td>
<td>500</td>
<td>150</td>
<td>24.4</td>
<td>162.8</td>
<td>29.4</td>
</tr>
<tr>
<td>AV2, F</td>
<td>800</td>
<td>150</td>
<td>15.8</td>
<td>105.4</td>
<td>19.0</td>
</tr>
<tr>
<td>AV2, F</td>
<td>3k</td>
<td>114</td>
<td>10.4</td>
<td>91.8</td>
<td>12.5</td>
</tr>
<tr>
<td>AV2, F</td>
<td>5k</td>
<td>150</td>
<td>29.8</td>
<td>198.7</td>
<td>35.9</td>
</tr>
<tr>
<td>AV2, F</td>
<td>5k</td>
<td>114</td>
<td>11.5</td>
<td>101.2</td>
<td>13.8</td>
</tr>
<tr>
<td>AV2, F</td>
<td>10k</td>
<td>150</td>
<td>17.5</td>
<td>116.8</td>
<td>21.1</td>
</tr>
<tr>
<td>Total</td>
<td>-</td>
<td>-</td>
<td>153.5</td>
<td>1070.3</td>
<td>187.3</td>
</tr>
</tbody>
</table>
<p>And the PacBio data:</p>
<table>
<thead>
<tr>
<th>Source</th>
<th>Raw bases (G)</th>
<th>Raw reads (M)</th>
<th>Sequencing depth</th>
<th>Avg.read size (bp)</th>
<th>N50 (bp)</th>
<th>N90 (bp)</th>
<th>Note</th>
</tr>
</thead>
<tbody>
<tr>
<td>AV3, F</td>
<td>15.2</td>
<td>20.1</td>
<td>18.31</td>
<td>7550</td>
<td>10046</td>
<td>4558</td>
<td>20170111</td>
</tr>
<tr>
<td>AV4, F</td>
<td>45.2</td>
<td>4.6</td>
<td>54.46</td>
<td>9798</td>
<td>17348</td>
<td>5702</td>
<td>20171224</td>
</tr>
<tr>
<td>Total</td>
<td>60.4</td>
<td>24.7</td>
<td>72.77</td>
<td>9115</td>
<td>14630</td>
<td>5310</td>
<td>-</td>
</tr>
</tbody>
</table>
<p>Because we didn’t receive the data at the same time (first all Illumina data, then 20X PacBio data, finally another 50X PacBio data), we tried many different assembly strategies: Illumina-dominant, and PacBio-dominant. We found ‘PacBio-dominant pipelines’ produced significantly good results, so we gradually gave up other pipelines and focused on exploring PacBio assemblers.</p>
<p>This note is one of the attempts of ‘Illumina-dominant pipelines’, and I ran LINKS as a scaffoling tool, after runing MEGAHIT-SOAPdenovo successfully. I’ve tried so many with MEGAHIT, and I ran LINKS after every attempt. See <a href="https://yiweiniu.github.io/blog/2018/03/Genome-assembly-pipeline-MEGAHIT-SOAPdenovo-fusion/">Genome Assembly Pipeline: MEGAHIT &amp; SOAPdenovo-fusion</a>. I put two representative results here as a memo.</p>
<p>PS: I encountered a problem and solved with the help of the author: <a href="https://github.com/bcgsc/LINKS/issues/12" target="_blank" rel="noopener">LINKS termineted with no error message when I used hybrid reads to scaffold</a>.</p>
<p>The <code>-t</code> and <code>-d</code> are important for RAM consumption. I’ve tried <code>-t</code> with <code>5, 10, 20</code>, and <code>-t 20</code> works. <code>-d</code> didn’t matter in my case, so I just set lots of <code>-d</code>.</p>
<p>Scripts I used and the stats I got:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">LINKS_HOME=/software/links_v1.8.5</span><br><span class="line">WORK_DIR=/DenovoSeq</span><br><span class="line"></span><br><span class="line">echo Start time is `date +%Y/%m/%d--%H:%M`</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> re-scafflod the output of MEGAHIT small insert size and SOAP-fusion. use about 50X data</span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="variable">$LINKS_HOME</span>/LINKS -k 21 -f <span class="variable">$WORK_DIR</span>/MEGAHIT/small_insert.no270/SOAP-fusion/k41.scafSeq -s Pacbio.fof -b reLINKS2 -t 20 -d 4000,5000,6000,8000,10000,15000,20000</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> re-scaffold the output of MEGAHIT small insert size and SOAP-fusion, use about 70X data</span></span><br><span class="line"><span class="meta">$</span><span class="bash">LINKS_HOME/LINKS -k 21 -f <span class="variable">$WORK_DIR</span>/MEGAHIT/small_insert.no270/SOAP-fusion/k41.scafSeq -s Pacbio_171231.fof -b reLINKS_171231 -t 20 -d 4000,5000,6000,8000,10000,15000,20000</span></span><br><span class="line"></span><br><span class="line">echo Finish time is `date +%Y/%m/%d--%H:%M`</span><br></pre></td></tr></table></figure>
<p>And the input data:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># first batch PB data, about 20X</span></span></span><br><span class="line"><span class="meta">$</span><span class="bash"> cat Pacbio.fof</span></span><br><span class="line">/DenovoSeq/Third_rawData/av_20k.fasta</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># all PB data, about 70X</span></span></span><br><span class="line"><span class="meta">$</span><span class="bash"> cat Pacbio_171231.fof</span></span><br><span class="line">/DenovoSeq/Third_rawData/third_all.fasta</span><br></pre></td></tr></table></figure>
<p>The stats I got:</p>
<p>re-scaffloding the output of MEGAHIT small insert size and SOAP-fusion, use ~20X PB data:</p>
<table>
<thead>
<tr>
<th></th>
<th>MEGAHIT</th>
<th>SOAP-fusion</th>
<th>LINKS</th>
</tr>
</thead>
<tbody>
<tr>
<td>Size_includeN</td>
<td>582099585</td>
<td>765246417</td>
<td>776147223</td>
</tr>
<tr>
<td>Size_withoutN</td>
<td>582099585</td>
<td>574675978</td>
<td>574675978</td>
</tr>
<tr>
<td>Seq_Num</td>
<td>575808</td>
<td>377325</td>
<td>370033</td>
</tr>
<tr>
<td>Mean_Size</td>
<td>1010</td>
<td>2028</td>
<td>2097</td>
</tr>
<tr>
<td>Median_Size</td>
<td>433</td>
<td>393</td>
<td>389</td>
</tr>
<tr>
<td>Longest_Seq</td>
<td>55064</td>
<td>436254</td>
<td>621272</td>
</tr>
<tr>
<td>Shortest_Seq</td>
<td>200</td>
<td>200</td>
<td>200</td>
</tr>
<tr>
<td>GC_Content</td>
<td>31.43</td>
<td>31.42</td>
<td>31.42</td>
</tr>
<tr>
<td>N50</td>
<td>2158</td>
<td>33478</td>
<td>42519</td>
</tr>
<tr>
<td>N90</td>
<td>358</td>
<td>439</td>
<td>444</td>
</tr>
<tr>
<td>Gap</td>
<td>0</td>
<td>24.9</td>
<td>25.96</td>
</tr>
</tbody>
</table>
<p>re-scaffolding the output of MEGAHIT small insert size and SOAP-fusion, use ~70X PB data.</p>
<table>
<thead>
<tr>
<th></th>
<th>MEGAHIT</th>
<th>SOAP-fusion</th>
<th>LINKS</th>
</tr>
</thead>
<tbody>
<tr>
<td>Size_includeN</td>
<td>582099585</td>
<td>765246417</td>
<td>798803587</td>
</tr>
<tr>
<td>Size_withoutN</td>
<td>582099585</td>
<td>574675978</td>
<td>574675978</td>
</tr>
<tr>
<td>Seq_Num</td>
<td>575808</td>
<td>377325</td>
<td>359653</td>
</tr>
<tr>
<td>Mean_Size</td>
<td>1010</td>
<td>2028</td>
<td>2221</td>
</tr>
<tr>
<td>Median_Size</td>
<td>433</td>
<td>393</td>
<td>384</td>
</tr>
<tr>
<td>Longest_Seq</td>
<td>55064</td>
<td>436254</td>
<td>732454</td>
</tr>
<tr>
<td>Shortest_Seq</td>
<td>200</td>
<td>200</td>
<td>200</td>
</tr>
<tr>
<td>GC_Content</td>
<td>31.43</td>
<td>31.42</td>
<td>31.42</td>
</tr>
<tr>
<td>N50</td>
<td>2158</td>
<td>33478</td>
<td>68195</td>
</tr>
<tr>
<td>N90</td>
<td>358</td>
<td>439</td>
<td>453</td>
</tr>
<tr>
<td>Gap</td>
<td>0</td>
<td>24.9</td>
<td>28.06</td>
</tr>
</tbody>
</table>
<p>The last run took about <code>27 hours</code>.</p>
<p>As can be seen, adding 50X PacBio data did not help a lot. After this, we started to use ‘PacBio-dominant pipelines’ and have tried many different assemblers. But, LINKS is a good scaffolding tool if you assemble the genome with illumina data and want to further scaffold with &lt;20X long-reads.</p>
<p>This note can serve as a reference in case I will have to use it again.</p>
<h2 id="change-log"><a class="markdownIt-Anchor" href="#change-log"></a> Change log</h2>
<ul>
<li>20180307: createt the note.</li>
<li>20180709: update the ‘background information’.</li>
<li>20180817: change the ‘General usage’ part, add contents about ‘parameter setting’.</li>
</ul>

      
    </div>

    

    
    
    

    

    

    
      <div>
        <ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>Post author:  </strong>Yiwei Niu</li>
  <li class="post-copyright-link">
    <strong>Post link: </strong>
    <a href="https://yiweiniu.github.io/blog/2018/03/Genome-assembly-pipeline-LINKS/" title="Genome Assembly Pipeline: LINKS">https://yiweiniu.github.io/blog/2018/03/Genome-assembly-pipeline-LINKS/</a>
  </li>
  <li class="post-copyright-license">
    <strong>Copyright Notice:  </strong>All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="external nofollow" target="_blank">CC BY-NC-SA 4.0</a> unless stating additionally.</li>
</ul>

      </div>
    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/blog/tags/bio-tools/" rel="tag"># bio-tools</a>
          
            <a href="/blog/tags/genome-assembly-pipeline/" rel="tag"># genome assembly pipeline</a>
          
            <a href="/blog/tags/hybrid-genome-assembly/" rel="tag"># hybrid genome assembly</a>
          
            <a href="/blog/tags/scaffolding/" rel="tag"># scaffolding</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/blog/2018/03/Genome-assembly-pipeline-MEGAHIT-SOAPdenovo-fusion/" rel="next" title="Genome Assembly Pipeline: MEGAHIT & SOAPdenovo-fusion">
                <i class="fa fa-chevron-left"></i> Genome Assembly Pipeline: MEGAHIT & SOAPdenovo-fusion
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/blog/2018/03/Genome-assembly-pipeline-BESST/" rel="prev" title="Genome Assembly Pipeline: BESST">
                Genome Assembly Pipeline: BESST <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>


  </div>


          </div>
          

  
    <div class="comments" id="comments">
      <div id="disqus_thread">
        <noscript>
          Please enable JavaScript to view the
          <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a>
        </noscript>
      </div>
    </div>

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            Overview
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image"
                src="/blog/images/portrait.jpg"
                alt="Yiwei Niu" />
            
              <p class="site-author-name" itemprop="name">Yiwei Niu</p>
              <p class="site-description motion-element" itemprop="description">a beginner, a learner</p>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/blog/archives">
                
                    <span class="site-state-item-count">20</span>
                    <span class="site-state-item-name">posts</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-categories">
                  <a href="/blog/categories/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">17</span>
                    <span class="site-state-item-name">categories</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-tags">
                  <a href="/blog/tags/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">35</span>
                    <span class="site-state-item-name">tags</span>
                  </a>
                </div>
              
            </nav>
          

          
            <div class="feed-link motion-element">
              <a href="/blog/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          
            <div class="links-of-author motion-element">
              
                <span class="links-of-author-item">
                  <a href="mailto:xiaohuwangwang@qq.com" target="_blank" title="E-Mail"><i class="fa fa-fw fa-envelope"></i>E-Mail</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="https://github.com/YiweiNiu" target="_blank" title="GitHub"><i class="fa fa-fw fa-github"></i>GitHub</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="http://weibo.com/u/3049858680" target="_blank" title="Weibo"><i class="fa fa-fw fa-weibo"></i>Weibo</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="https://twitter.com/ywniu" target="_blank" title="Twitter"><i class="fa fa-fw fa-twitter"></i>Twitter</a>
                  
                </span>
              
            </div>
          

          
          

          
          
            <div class="links-of-blogroll motion-element links-of-blogroll-block">
              <div class="links-of-blogroll-title">
                <i class="fa  fa-fw fa-link"></i>
                Links
              </div>
              <ul class="links-of-blogroll-list">
                
                  <li class="links-of-blogroll-item">
                    <a href="https://kaigedong.github.io/" title="Kaige Dong" target="_blank">Kaige Dong</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://homolog.us/blogs/" title="Homolog.us" target="_blank">Homolog.us</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="https://cgatoxford.wordpress.com/" title="CGAT" target="_blank">CGAT</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://www.gettinggeneticsdone.com/" title="Getting Genetics Done" target="_blank">Getting Genetics Done</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://www.acgt.me/" title="ACGT" target="_blank">ACGT</a>
                  </li>
                
              </ul>
            </div>
          

          
            
          
          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#intro"><span class="nav-number">1.</span> <span class="nav-text"> Intro</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#general-usage"><span class="nav-number">2.</span> <span class="nav-text"> General usage</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#install"><span class="nav-number">2.1.</span> <span class="nav-text"> Install</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#parameter-setting"><span class="nav-number">2.2.</span> <span class="nav-text"> Parameter setting</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#an-example"><span class="nav-number">2.2.1.</span> <span class="nav-text"> An example</span></a></li></ol></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#in-practice"><span class="nav-number">3.</span> <span class="nav-text"> In practice</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#background"><span class="nav-number">3.1.</span> <span class="nav-text"> background</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#data"><span class="nav-number">3.2.</span> <span class="nav-text"> data</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#change-log"><span class="nav-number">4.</span> <span class="nav-text"> Change log</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; 2017 &mdash; <span itemprop="copyrightYear">2018</span>
  <span class="with-love" id="animate">
    <i class="fa fa-mars"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Yiwei Niu</span>

  

  
</div>




  <div class="powered-by">Powered by <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a></div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">Theme &mdash; <a class="theme-link" target="_blank" href="https://theme-next.org">NexT.Gemini</a></div>




        
<div class="busuanzi-count">
  <script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script>

  
    <span class="site-uv" title="Total Visitors">
      <i class="fa fa-user"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_uv"></span>
    </span>
  

  
    <span class="site-pv" title="Total Views">
      <i class="fa fa-eye"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_pv"></span>
    </span>
  
</div>









        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
          <span id="scrollpercent"><span>0</span>%</span>
        
      </div>
    

    
	
    

    
      <div>
        <!-- Go to www.addthis.com/dashboard to customize your tools -->
<div class="addthis_inline_share_toolbox">
  <script type = "text/javascript" src = "//s7.addthis.com/js/300/addthis_widget.js#pubid=ra-5a7ad07b7823d03b" async = "async" ></script>
</div>

      </div>
    
  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>


























  
  
    <script type="text/javascript" src="/blog/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  


  


  <script type="text/javascript" src="/blog/js/src/utils.js?v=6.3.0"></script>

  <script type="text/javascript" src="/blog/js/src/motion.js?v=6.3.0"></script>



  
  


  <script type="text/javascript" src="/blog/js/src/affix.js?v=6.3.0"></script>

  <script type="text/javascript" src="/blog/js/src/schemes/pisces.js?v=6.3.0"></script>



  
  <script type="text/javascript" src="/blog/js/src/scrollspy.js?v=6.3.0"></script>
<script type="text/javascript" src="/blog/js/src/post-details.js?v=6.3.0"></script>



  


  <script type="text/javascript" src="/blog/js/src/bootstrap.js?v=6.3.0"></script>



  

  
    <script id="dsq-count-scr" src="https://yiwei-nius-note.disqus.com/count.js" async></script>
  

  
    <script type="text/javascript">
      var disqus_config = function () {
        this.page.url = 'https://yiweiniu.github.io/blog/2018/03/Genome-assembly-pipeline-LINKS/';
        this.page.identifier = '2018/03/Genome-assembly-pipeline-LINKS/';
        this.page.title = 'Genome Assembly Pipeline: LINKS';
        };
      function loadComments () {
        var d = document, s = d.createElement('script');
        s.src = 'https://yiwei-nius-note.disqus.com/embed.js';
        s.setAttribute('data-timestamp', '' + +new Date());
        (d.head || d.body).appendChild(s);
      }
      
        loadComments();
      
    </script>
  












  





  

  

  

  

  
  

  

  

  

  

  
  <style>
    .copy-btn {
      display: inline-block;
      padding: 6px 12px;
      font-size: 13px;
      font-weight: 700;
      line-height: 20px;
      color: #333;
      white-space: nowrap;
      vertical-align: middle;
      cursor: pointer;
      background-color: #eee;
      background-image: linear-gradient(#fcfcfc, #eee);
      border: 1px solid #d5d5d5;
      border-radius: 3px;
      user-select: none;
      outline: 0;
    }

    .highlight-wrap .copy-btn {
      transition: opacity .3s ease-in-out;
      opacity: 0;
      padding: 2px 6px;
      position: absolute;
      right: 4px;
      top: 8px;
    }

    .highlight-wrap:hover .copy-btn,
    .highlight-wrap .copy-btn:focus {
      opacity: 1
    }

    .highlight-wrap {
      position: relative;
    }
  </style>
  <script>
    $('.highlight').each(function (i, e) {
      var $wrap = $('<div>').addClass('highlight-wrap')
      $(e).after($wrap)
      $wrap.append($('<button>').addClass('copy-btn').append('Copy').on('click', function (e) {
        var code = $(this).parent().find('.code').find('.line').map(function (i, e) {
          return $(e).text()
        }).toArray().join('\n')
        var ta = document.createElement('textarea')
        document.body.appendChild(ta)
        ta.style.position = 'absolute'
        ta.style.top = '0px'
        ta.style.left = '0px'
        ta.value = code
        ta.select()
        ta.focus()
        var result = document.execCommand('copy')
        document.body.removeChild(ta)
        
          if(result)$(this).text('Copied')
          else $(this).text('Copy failed')
        
        $(this).blur()
      })).on('mouseleave', function (e) {
        var $b = $(this).find('.copy-btn')
        setTimeout(function () {
          $b.text('Copy')
        }, 300)
      }).append(e)
    })
  </script>


</body>
</html>
