<!DOCTYPE html>



  


<html class="theme-next gemini use-motion" lang="_en">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2"/>
<meta name="theme-color" content="#222">












<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />



  <meta name="google-site-verification" content="3dnqtuxhGQK5nK7aqRtU2c3DzZlaVtBMN6M39UoGcgY" />




















<link href="/blog/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/blog/css/main.css?v=6.1.0" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/blog/images/apple-touch-icon-next.png?v=6.1.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/blog/images/favicon-32x32-next.png?v=6.1.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/blog/images/favicon-16x16-next.png?v=6.1.0">


  <link rel="mask-icon" href="/blog/images/logo.svg?v=6.1.0" color="#222">









<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/blog/',
    scheme: 'Gemini',
    version: '6.1.0',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":true,"onmobile":false},
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="Intro From its git repo:  Flye is a de novo assembler for long and noisy reads, such as those produced by PacBio and Oxford Nanopore Technologies. The algorithm uses an A-Bruijn graph to find the ove">
<meta name="keywords" content="bio-tools,genome assembly pipeline,TGS genome assembly">
<meta property="og:type" content="article">
<meta property="og:title" content="Genome Assembly Pipeline: Flye">
<meta property="og:url" content="https://yiweiniu.github.io/blog/2018/03/Genome-assembly-pipeline-Flye/index.html">
<meta property="og:site_name" content="Yiwei Niu&#39;s Note">
<meta property="og:description" content="Intro From its git repo:  Flye is a de novo assembler for long and noisy reads, such as those produced by PacBio and Oxford Nanopore Technologies. The algorithm uses an A-Bruijn graph to find the ove">
<meta property="og:locale" content="_en">
<meta property="og:updated_time" content="2018-07-02T01:34:06.111Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Genome Assembly Pipeline: Flye">
<meta name="twitter:description" content="Intro From its git repo:  Flye is a de novo assembler for long and noisy reads, such as those produced by PacBio and Oxford Nanopore Technologies. The algorithm uses an A-Bruijn graph to find the ove">



  <link rel="alternate" href="/blog/atom.xml" title="Yiwei Niu's Note" type="application/atom+xml" />




  <link rel="canonical" href="https://yiweiniu.github.io/blog/2018/03/Genome-assembly-pipeline-Flye/"/>



<script type="text/javascript" id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>Genome Assembly Pipeline: Flye | Yiwei Niu's Note</title>
  









  <noscript>
  <style type="text/css">
    .use-motion .motion-element,
    .use-motion .brand,
    .use-motion .menu-item,
    .sidebar-inner,
    .use-motion .post-block,
    .use-motion .pagination,
    .use-motion .comments,
    .use-motion .post-header,
    .use-motion .post-body,
    .use-motion .collection-title { opacity: initial; }

    .use-motion .logo,
    .use-motion .site-title,
    .use-motion .site-subtitle {
      opacity: initial;
      top: initial;
    }

    .use-motion {
      .logo-line-before i { left: initial; }
      .logo-line-after i { right: initial; }
    }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="_en">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"> 

<div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/blog/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Yiwei Niu's Note</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle">to share, to learn</p>
      
  </div>

  <div class="site-nav-toggle">
    <button aria-label="Toggle navigation bar">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        
          
  
  <li class="menu-item menu-item-home menu-item-active">
    <a href="/blog/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-home"></i> <br />Home</a>
</li>

      
        
        
          
  
  <li class="menu-item menu-item-categories menu-item-active">
    <a href="/blog/categories/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-th"></i> <br />Categories</a>
</li>

      
        
        
          
  
  <li class="menu-item menu-item-tags">
    <a href="/blog/tags/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />Tags</a>
</li>

      
        
        
          
  
  <li class="menu-item menu-item-archives">
    <a href="/blog/archives/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />Archives</a>
</li>

      
        
        
          
  
  <li class="menu-item menu-item-about">
    <a href="https:/yiweiniu.github.io/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-user"></i> <br />About</a>
</li>

      

      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br />Search</a>
        </li>
      
    </ul>
  

  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off"
             placeholder="Searching..." spellcheck="false"
             type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>


  



 </div>
    </header>

    


    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
            

          
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="https://yiweiniu.github.io/blog/blog/2018/03/Genome-assembly-pipeline-Flye/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Yiwei Niu">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/blog/images/portrait.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Yiwei Niu's Note">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">Genome Assembly Pipeline: Flye</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              
              <time title="Post created" itemprop="dateCreated datePublished" datetime="2018-03-30T00:33:08+08:00">2018-03-30</time>
            

            
            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/blog/categories/genome-assembly/" itemprop="url" rel="index"><span itemprop="name">genome assembly</span></a></span>

                
                
                  , 
                
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/blog/categories/genome-assembly/TGS-pipeline/" itemprop="url" rel="index"><span itemprop="name">TGS pipeline</span></a></span>

                
                
              
            </span>
          

          
            
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
                <a href="/blog/2018/03/Genome-assembly-pipeline-Flye/#comments" itemprop="discussionUrl">
                  <span class="post-comments-count disqus-comment-count"
                        data-disqus-identifier="2018/03/Genome-assembly-pipeline-Flye/" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          

          
            <span class="post-meta-divider">|</span>
            <span class="post-meta-item-icon"
            >
            <i class="fa fa-eye"></i>
             Views: 
            <span class="busuanzi-value" id="busuanzi_value_page_pv" ></span>
            </span>
          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                <span title="Symbols count in article">8.9k</span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                <span title="Reading time">0:09</span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="intro"><a class="markdownIt-Anchor" href="#intro"></a> Intro</h2>
<p>From <a href="https://github.com/fenderglass/Flye" target="_blank" rel="noopener">its git repo</a>:</p>
<blockquote>
<p>Flye is a de novo assembler for long and noisy reads, such as those produced by PacBio and Oxford Nanopore Technologies. The algorithm uses an A-Bruijn graph to find the overlaps between reads and does not require them to be error-corrected. After the initial assembly, Flye performs an extra repeat classification and analysis step to improve the structural accuracy of the resulting sequence. The package also includes a polisher module, which produces the final assembly of high nucleotide-level quality.</p>
</blockquote>
<p>This tool is now on biRxiv:</p>
<blockquote>
<p>Kolmogorov M, Yuan J, Lin Y, Pevzner P. Assembly of Long Error-Prone Reads Using Repeat Graphs. bioRxiv. 2018 Jan 12:247148. doi:10.1101/247148</p>
</blockquote>
<p>My feelings:</p>
<ul>
<li>easy to use</li>
<li>comparatively good results, good N50, good completeness</li>
<li>not too many parameters to be tested</li>
</ul>
<h2 id="general-usage"><a class="markdownIt-Anchor" href="#general-usage"></a> General usage</h2>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br></pre></td><td class="code"><pre><span class="line"># install</span><br><span class="line">git clone https://github.com/fenderglass/Flye Flye-2.3.3</span><br><span class="line">cd Flye-2.3.3</span><br><span class="line">python setup.py build</span><br><span class="line"></span><br><span class="line"># usage</span><br><span class="line">$ ./bin/flye -h</span><br><span class="line">usage: flye (--pacbio-raw | --pacbio-corr | --nano-raw |</span><br><span class="line">       --nano-corr | --subassemblies) file1 [file_2 ...]</span><br><span class="line">       --genome-size size --out-dir dir_path [--threads int]</span><br><span class="line">       [--iterations int] [--min-overlap int] [--resume]</span><br><span class="line">       [--debug] [--version] [--help]</span><br><span class="line"></span><br><span class="line">Assembly of long and error-prone reads</span><br><span class="line"></span><br><span class="line">optional arguments:</span><br><span class="line">  -h, --help            show this help message and exit</span><br><span class="line">  --pacbio-raw path [path ...]</span><br><span class="line">                        PacBio raw reads</span><br><span class="line">  --pacbio-corr path [path ...]</span><br><span class="line">                        PacBio corrected reads</span><br><span class="line">  --nano-raw path [path ...]</span><br><span class="line">                        ONT raw reads</span><br><span class="line">  --nano-corr path [path ...]</span><br><span class="line">                        ONT corrected reads</span><br><span class="line">  --subassemblies path [path ...]</span><br><span class="line">                        high-quality contig-like input</span><br><span class="line">  -g size, --genome-size size</span><br><span class="line">                        estimated genome size (for example, 5m or 2.6g)</span><br><span class="line">  -o path, --out-dir path</span><br><span class="line">                        Output directory</span><br><span class="line">  -t int, --threads int</span><br><span class="line">                        number of parallel threads (default: 1)</span><br><span class="line">  -i int, --iterations int</span><br><span class="line">                        number of polishing iterations (default: 1)</span><br><span class="line">  -m int, --min-overlap int</span><br><span class="line">                        minimum overlap between reads (default: auto)</span><br><span class="line">  --resume              resume from the last completed stage</span><br><span class="line">  --resume-from stage_name</span><br><span class="line">                        resume from a custom stage</span><br><span class="line">  --debug               enable debug output</span><br><span class="line">  -v, --version         show program&apos;s version number and exit</span><br><span class="line"></span><br><span class="line">Input reads could be in FASTA or FASTQ format, uncompressed</span><br><span class="line">or compressed with gz. Currenlty, raw and corrected reads</span><br><span class="line">from PacBio and ONT are supported. Additionally, --subassemblies</span><br><span class="line">option does a consensus assembly of high-quality input contigs.</span><br><span class="line">You may specify multiple fles with reads (separated by spaces).</span><br><span class="line">Mixing different read types is not yet supported.</span><br><span class="line"></span><br><span class="line">You must provide an estimate of the genome size as input,</span><br><span class="line">which is used for solid k-mers selection. The estimate could</span><br><span class="line">be rough (e.g. withing 0.5x-2x range) and does not affect</span><br><span class="line">the other assembly stages. Standard size modificators are</span><br><span class="line">supported (e.g. 5m or 2.6g)</span><br><span class="line"></span><br><span class="line"># E. coli P6-C4 PacBio data</span><br><span class="line">wget https://zenodo.org/record/1172816/files/E.coli_PacBio_40x.fasta</span><br><span class="line">flye --pacbio-raw E.coli_PacBio_40x.fasta --out-dir out_pacbio --genome-size 5m --threads 4</span><br><span class="line"></span><br><span class="line"># E. coli Oxford Nanopore Technologies data</span><br><span class="line">wget https://zenodo.org/record/1172816/files/Loman_E.coli_MAP006-1_2D_50x.fasta</span><br><span class="line">flye --nano-raw Loman_E.coli_MAP006-1_2D_50x.fasta --out-dir out_nano --genome-size 5m --threads 4</span><br></pre></td></tr></table></figure>
<p>See <a href="https://github.com/fenderglass/Flye/blob/flye/docs/USAGE.md" target="_blank" rel="noopener">Flye manual</a> for full usage.</p>
<h2 id="in-practice"><a class="markdownIt-Anchor" href="#in-practice"></a> In practice</h2>
<h3 id="an-insect"><a class="markdownIt-Anchor" href="#an-insect"></a> An insect</h3>
<ul>
<li>The species: high heterogeneity, high AT, high repetition.</li>
<li>Genome size: male 790M, female 830M.</li>
<li>Data used：about 70X PacBio long-reads.</li>
<li>OS environment: CentOS6.6  86_64  glibc-2.12. QSUB grid system. 15 Fat nodes (2TB RAM, 40 CPU) and 10 Blade nodes (156G RAM, 24 CPU).</li>
</ul>
<h4 id="version-232-gd46edb7"><a class="markdownIt-Anchor" href="#version-232-gd46edb7"></a> version 2.3.2-gd46edb7</h4>
<ul>
<li><code>Flye</code> version: 2.3.2-gd46edb7</li>
</ul>
<p>I didn’t test all the parameters. Below is the results based on default settings.</p>
<p>command:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">flye --pacbio-raw $DATADIR/third/third_all.fasta --out-dir run1 --genome-size 850m --threads 24</span><br></pre></td></tr></table></figure>
<p>stats:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"># contig</span><br><span class="line">Size_includeN 724744485</span><br><span class="line">Size_withoutN 724744485</span><br><span class="line">Seq_Num 17602</span><br><span class="line">Mean_Size 41173</span><br><span class="line">Median_Size 17572</span><br><span class="line">Longest_Seq 1648999</span><br><span class="line">Shortest_Seq  55</span><br><span class="line">GC_Content  31.55</span><br><span class="line">N50 91066</span><br><span class="line">N90 17456</span><br><span class="line">Gap 0.0</span><br><span class="line"></span><br><span class="line"># scaffold</span><br><span class="line">Size_includeN 724753785</span><br><span class="line">Size_withoutN 724744485</span><br><span class="line">Seq_Num 17509</span><br><span class="line">Mean_Size 41393</span><br><span class="line">Median_Size 17532</span><br><span class="line">Longest_Seq 1648999</span><br><span class="line">Shortest_Seq  55</span><br><span class="line">GC_Content  31.55</span><br><span class="line">N50 92367</span><br><span class="line">N90 17530</span><br><span class="line">Gap 0.0</span><br></pre></td></tr></table></figure>
<h4 id="version-233-g47cdd0b"><a class="markdownIt-Anchor" href="#version-233-g47cdd0b"></a> Version 2.3.3-g47cdd0b</h4>
<p><code>Flye</code> 2.3.3 have two updates appealing to me:</p>
<ul>
<li>Automatic selection of minimum overlap parameter based on read length</li>
<li>Minimap2 updated</li>
</ul>
<p>Because I’ve run <code>Canu</code> before, and <code>Flye</code> can start from raw data and corrected data, I’ll test <code>Flye</code> for both.</p>
<h5 id="from-raw-data"><a class="markdownIt-Anchor" href="#from-raw-data"></a> From raw data</h5>
<p>Commands:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">$</span><span class="bash">TOOLDIR/Flye-2.3.3/bin/flye --pacbio-raw third_all.fasta --out-dir run2 --genome-size 830m --threads 40</span></span><br></pre></td></tr></table></figure>
<p>Stats:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> contigs</span></span><br><span class="line">Size_includeN 787629166</span><br><span class="line">Size_withoutN 787629166</span><br><span class="line">Seq_Num 14846</span><br><span class="line">Mean_Size 53053</span><br><span class="line">Median_Size 20542</span><br><span class="line">Longest_Seq 1636300</span><br><span class="line">Shortest_Seq  12</span><br><span class="line">GC_Content  31.6</span><br><span class="line">N50 121564</span><br><span class="line">L50 1699</span><br><span class="line">N90 25419</span><br><span class="line">Gap 0.0</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> scaffolds</span></span><br><span class="line">Size_includeN 787632766</span><br><span class="line">Size_withoutN 787629166</span><br><span class="line">Seq_Num 14810</span><br><span class="line">Mean_Size 53182</span><br><span class="line">Median_Size 20465</span><br><span class="line">Longest_Seq 1692734</span><br><span class="line">Shortest_Seq  12</span><br><span class="line">GC_Content  31.6</span><br><span class="line">N50 122313</span><br><span class="line">L50 1680</span><br><span class="line">N90 25437</span><br><span class="line">Gap 0.0</span><br></pre></td></tr></table></figure>
<h5 id="from-corrected-data-from-canu-about-33x"><a class="markdownIt-Anchor" href="#from-corrected-data-from-canu-about-33x"></a> From corrected data from <code>Canu</code> (about 33X)</h5>
<p>Commands:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">$</span><span class="bash">TOOLDIR/Flye-2.3.3/bin/flye --pacbio-corr canu.correctedReads.fasta.gz --out-dir run3 --genome-size 830m --threads 40</span></span><br></pre></td></tr></table></figure>
<p>Stats:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> contigs</span></span><br><span class="line">Size_includeN 833065987</span><br><span class="line">Size_withoutN 833065987</span><br><span class="line">Seq_Num 17536</span><br><span class="line">Mean_Size 47506</span><br><span class="line">Median_Size 26593</span><br><span class="line">Longest_Seq 1145994</span><br><span class="line">Shortest_Seq  518</span><br><span class="line">GC_Content  31.47</span><br><span class="line">N50 88680</span><br><span class="line">L50 2594</span><br><span class="line">N90 22129</span><br><span class="line">Gap 0.0</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> scaffolds</span></span><br><span class="line">Size_includeN 833070387</span><br><span class="line">Size_withoutN 833065987</span><br><span class="line">Seq_Num 17492</span><br><span class="line">Mean_Size 47625</span><br><span class="line">Median_Size 26602</span><br><span class="line">Longest_Seq 1145994</span><br><span class="line">Shortest_Seq  518</span><br><span class="line">GC_Content  31.47</span><br><span class="line">N50 89165</span><br><span class="line">L50 2581</span><br><span class="line">N90 22242</span><br><span class="line">Gap 0.0</span><br></pre></td></tr></table></figure>
<h3 id="a-plant"><a class="markdownIt-Anchor" href="#a-plant"></a> A plant</h3>
<ul>
<li>The species: high heterogeneity, high repetition.</li>
<li>Genome size: 2.1G.</li>
<li>Data used：more than 100X PacBio long reads.</li>
<li>OS environment: CentOS6.6  86_64  glibc-2.12. QSUB grid system. 15 Fat nodes (2TB RAM, 40 CPU) and 10 Blade nodes (156G RAM, 24 CPU).</li>
</ul>
<h4 id="run1-more-than-100x-data"><a class="markdownIt-Anchor" href="#run1-more-than-100x-data"></a> run1, more than 100X data</h4>
<p>commands:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">$</span><span class="bash">path2flye --pacbio-raw <span class="variable">$WORKDIR</span>/data/Pacbio/all.fasta --out-dir run1 --genome-size 2g --threads 30</span></span><br></pre></td></tr></table></figure>
<p>But I came across a memory issue: <a href="https://github.com/fenderglass/Flye/issues/46" target="_blank" rel="noopener">ERROR: Caught unhandled exception: std::bad_alloc in both 2.3.2 and 2.3.3</a>. And the author suggested me to downsample the data.</p>
<p>And I asked him that what’s the difference: using all raw data (say 100X) and using downsampling data (say longest 50X)? He said “You might have extra connectivity information in these 100x reads (you can resolve more repeats, for example). But some studies suggest (Canu paper, for example) that you don’t really need more than 40x in general (but it, of course, also depends on the genome complexity, ploidy etc…). Plus, extra coverage helps to get a good final consensus.”</p>
<h4 id="run2-with-about-50x-data"><a class="markdownIt-Anchor" href="#run2-with-about-50x-data"></a> run2, with about 50X data</h4>
<p>I used <a href="https://github.com/yechengxi/AssemblyUtility" target="_blank" rel="noopener">SelectLongestReads</a> to downsample about 50X data and ran <code>Flye</code> again.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> run1.1, extract 50X data</span></span><br><span class="line">SelectLongestReads sum 100000000000 longest 1 o Pacbio_50x.fasta f all.fasta</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> remove @ from the fasta</span></span><br><span class="line">replace_@_in_fasta_header.py Pacbio_50x.fasta &gt; Pacbio_50x_no_@.fasta</span><br><span class="line"></span><br><span class="line"><span class="meta">$</span><span class="bash">path2flye --pacbio-raw Pacbio_50x_no_@.fasta --out-dir run1.1 --genome-size 2g --threads 30 --resume</span></span><br></pre></td></tr></table></figure>
<p>The reason why I removed the <code>@</code> from the headers was because I encountered another problem: <a href="https://github.com/fenderglass/Flye/issues/48" target="_blank" rel="noopener">ERROR: parse error in 1-consensus/consensus.fasta on line 1: empty sequence</a>. It seemed that <code>Flye</code> would ignore these headers.</p>
<p>And the stats I got:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> contigs</span></span><br><span class="line">Size_includeN 1640872256</span><br><span class="line">Size_withoutN 1640872256</span><br><span class="line">Seq_Num 9843</span><br><span class="line">Mean_Size 166704</span><br><span class="line">Median_Size 72841</span><br><span class="line">Longest_Seq 8808184</span><br><span class="line">Shortest_Seq  139</span><br><span class="line">GC_Content  37.78</span><br><span class="line">N50 398108</span><br><span class="line">L50 1119</span><br><span class="line">N90 83615</span><br><span class="line">Gap 0.0</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> scaffolds</span></span><br><span class="line">Size_includeN 1640874656</span><br><span class="line">Size_withoutN 1640872256</span><br><span class="line">Seq_Num 9819</span><br><span class="line">Mean_Size 167112</span><br><span class="line">Median_Size 72812</span><br><span class="line">Longest_Seq 8808184</span><br><span class="line">Shortest_Seq  139</span><br><span class="line">GC_Content  37.78</span><br><span class="line">N50 399898</span><br><span class="line">L50 1114</span><br><span class="line">N90 83809</span><br><span class="line">Gap 0.0</span><br></pre></td></tr></table></figure>
<p>Not bad.</p>
<h4 id="run3-with-corrected-data-from-canu-about-37x"><a class="markdownIt-Anchor" href="#run3-with-corrected-data-from-canu-about-37x"></a> run3, with corrected data from <code>Canu</code> (about 37X)</h4>
<p>The <code>Canu</code> version was <code>1.7</code>.</p>
<p>commands:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> run2, pacbio corrected by canu, defalut</span></span><br><span class="line"><span class="meta">$</span><span class="bash">path2flye --pacbio-corr canu.correctedReads.fasta.gz --out-dir run2 --genome-size 2g --threads 30</span></span><br></pre></td></tr></table></figure>
<p>stats:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> contigs</span></span><br><span class="line">Size_includeN 1886541389</span><br><span class="line">Size_withoutN 1886541389</span><br><span class="line">Seq_Num 13177</span><br><span class="line">Mean_Size 143169</span><br><span class="line">Median_Size 70083</span><br><span class="line">Longest_Seq 2690265</span><br><span class="line">Shortest_Seq  110</span><br><span class="line">GC_Content  38.01</span><br><span class="line">N50 310885</span><br><span class="line">L50 1669</span><br><span class="line">N90 70697</span><br><span class="line">Gap 0.0</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> scaffolds</span></span><br><span class="line">Size_includeN 1886554189</span><br><span class="line">Size_withoutN 1886541389</span><br><span class="line">Seq_Num 13049</span><br><span class="line">Mean_Size 144574</span><br><span class="line">Median_Size 70202</span><br><span class="line">Longest_Seq 2690265</span><br><span class="line">Shortest_Seq  110</span><br><span class="line">GC_Content  38.01</span><br><span class="line">N50 315687</span><br><span class="line">L50 1648</span><br><span class="line">N90 71378</span><br><span class="line">Gap 0.0</span><br></pre></td></tr></table></figure>
<h2 id="useful-links"><a class="markdownIt-Anchor" href="#useful-links"></a> Useful links</h2>
<ul>
<li><a href="https://github.com/fenderglass/Flye/blob/flye/docs/USAGE.md" target="_blank" rel="noopener">Flye manual</a></li>
</ul>
<h2 id="change-log"><a class="markdownIt-Anchor" href="#change-log"></a> Change log</h2>
<ul>
<li>20180314: create the note.</li>
<li>20180428: test version 2.3.3, and run from corrected reads of <code>Canu</code></li>
<li>20180630: add the part of ‘A plant’.</li>
</ul>

      
    </div>

    

    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/blog/tags/bio-tools/" rel="tag"># bio-tools</a>
          
            <a href="/blog/tags/genome-assembly-pipeline/" rel="tag"># genome assembly pipeline</a>
          
            <a href="/blog/tags/TGS-genome-assembly/" rel="tag"># TGS genome assembly</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/blog/2018/03/Genome-assembly-pipeline-miniasm-Racon/" rel="next" title="Genome Assembly Pipeline: miniasm & Racon">
                <i class="fa fa-chevron-left"></i> Genome Assembly Pipeline: miniasm & Racon
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/blog/2018/04/How-to-install-perl-modules/" rel="prev" title="How to Install Perl Modules">
                How to Install Perl Modules <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
        <!-- Go to www.addthis.com/dashboard to customize your tools -->
<div class="addthis_inline_share_toolbox">
  <script type = "text/javascript" src = "//s7.addthis.com/js/300/addthis_widget.js#pubid=ra-5a7ad07b7823d03b" async = "async" ></script>
</div>

      
    </div>
  </div>


          </div>
          

  
    <div class="comments" id="comments">
      <div id="disqus_thread">
        <noscript>
          Please enable JavaScript to view the
          <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a>
        </noscript>
      </div>
    </div>

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            Overview
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image"
                src="/blog/images/portrait.jpg"
                alt="Yiwei Niu" />
            
              <p class="site-author-name" itemprop="name">Yiwei Niu</p>
              <p class="site-description motion-element" itemprop="description">a beginner, a learner</p>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/blog/archives/">
                
                    <span class="site-state-item-count">19</span>
                    <span class="site-state-item-name">posts</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-categories">
                  <a href="/blog/categories/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">15</span>
                    <span class="site-state-item-name">categories</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-tags">
                  <a href="/blog/tags/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">29</span>
                    <span class="site-state-item-name">tags</span>
                  </a>
                </div>
              
            </nav>
          

          
            <div class="feed-link motion-element">
              <a href="/blog/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          
            <div class="links-of-author motion-element">
              
                <span class="links-of-author-item">
                  <a href="mailto:xiaohuwangwang@qq.com" target="_blank" title="E-Mail"><i class="fa fa-fw fa-envelope"></i>E-Mail</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="https://github.com/YiweiNiu" target="_blank" title="GitHub"><i class="fa fa-fw fa-github"></i>GitHub</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="http://weibo.com/u/3049858680" target="_blank" title="Weibo"><i class="fa fa-fw fa-weibo"></i>Weibo</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="https://twitter.com/ywniu" target="_blank" title="Twitter"><i class="fa fa-fw fa-twitter"></i>Twitter</a>
                  
                </span>
              
            </div>
          

          
          

          
          
            <div class="links-of-blogroll motion-element links-of-blogroll-block">
              <div class="links-of-blogroll-title">
                <i class="fa  fa-fw fa-link"></i>
                Links
              </div>
              <ul class="links-of-blogroll-list">
                
                  <li class="links-of-blogroll-item">
                    <a href="https://kaigedong.github.io/" title="Kaige Dong" target="_blank">Kaige Dong</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://homolog.us/blogs/" title="Homolog.us" target="_blank">Homolog.us</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="https://cgatoxford.wordpress.com/" title="CGAT" target="_blank">CGAT</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://www.gettinggeneticsdone.com/" title="Getting Genetics Done" target="_blank">Getting Genetics Done</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://www.acgt.me/" title="ACGT" target="_blank">ACGT</a>
                  </li>
                
              </ul>
            </div>
          

          
            
          
          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#intro"><span class="nav-number">1.</span> <span class="nav-text"> Intro</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#general-usage"><span class="nav-number">2.</span> <span class="nav-text"> General usage</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#in-practice"><span class="nav-number">3.</span> <span class="nav-text"> In practice</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#an-insect"><span class="nav-number">3.1.</span> <span class="nav-text"> An insect</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#version-232-gd46edb7"><span class="nav-number">3.1.1.</span> <span class="nav-text"> version 2.3.2-gd46edb7</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#version-233-g47cdd0b"><span class="nav-number">3.1.2.</span> <span class="nav-text"> Version 2.3.3-g47cdd0b</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#from-raw-data"><span class="nav-number">3.1.2.1.</span> <span class="nav-text"> From raw data</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#from-corrected-data-from-canu-about-33x"><span class="nav-number">3.1.2.2.</span> <span class="nav-text"> From corrected data from Canu (about 33X)</span></a></li></ol></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#a-plant"><span class="nav-number">3.2.</span> <span class="nav-text"> A plant</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#run1-more-than-100x-data"><span class="nav-number">3.2.1.</span> <span class="nav-text"> run1, more than 100X data</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#run2-with-about-50x-data"><span class="nav-number">3.2.2.</span> <span class="nav-text"> run2, with about 50X data</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#run3-with-corrected-data-from-canu-about-37x"><span class="nav-number">3.2.3.</span> <span class="nav-text"> run3, with corrected data from Canu (about 37X)</span></a></li></ol></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#useful-links"><span class="nav-number">4.</span> <span class="nav-text"> Useful links</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#change-log"><span class="nav-number">5.</span> <span class="nav-text"> Change log</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; 2017 &mdash; <span itemprop="copyrightYear">2018</span>
  <span class="with-love" id="animate">
    <i class="fa fa-mars"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Yiwei Niu</span>

  

  
</div>




  <div class="powered-by">Powered by <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a></div>








        
<div class="busuanzi-count">
  <script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script>

  
    <span class="site-uv" title="Total Visitors">
      <i class="fa fa-user"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_uv"></span>
    </span>
  

  
    <span class="site-pv" title="Total Views">
      <i class="fa fa-eye"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_pv"></span>
    </span>
  
</div>









        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
          <span id="scrollpercent"><span>0</span>%</span>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>


























  
  
    <script type="text/javascript" src="/blog/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  


  


  <script type="text/javascript" src="/blog/js/src/utils.js?v=6.1.0"></script>

  <script type="text/javascript" src="/blog/js/src/motion.js?v=6.1.0"></script>



  
  


  <script type="text/javascript" src="/blog/js/src/affix.js?v=6.1.0"></script>

  <script type="text/javascript" src="/blog/js/src/schemes/pisces.js?v=6.1.0"></script>



  
  <script type="text/javascript" src="/blog/js/src/scrollspy.js?v=6.1.0"></script>
<script type="text/javascript" src="/blog/js/src/post-details.js?v=6.1.0"></script>



  


  <script type="text/javascript" src="/blog/js/src/bootstrap.js?v=6.1.0"></script>



  

  
    <script id="dsq-count-scr" src="https://yiwei-nius-note.disqus.com/count.js" async></script>
  

  
    <script type="text/javascript">
      var disqus_config = function () {
        this.page.url = 'https://yiweiniu.github.io/blog/2018/03/Genome-assembly-pipeline-Flye/';
        this.page.identifier = '2018/03/Genome-assembly-pipeline-Flye/';
        this.page.title = 'Genome Assembly Pipeline: Flye';
      };
      function loadComments () {
        var d = document, s = d.createElement('script');
        s.src = 'https://yiwei-nius-note.disqus.com/embed.js';
        s.setAttribute('data-timestamp', '' + +new Date());
        (d.head || d.body).appendChild(s);
      }
      
        loadComments();
      
    </script>
  





	





  












  

  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/blog/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  

  

  
  

  

  

  

  

</body>
</html>
