<!DOCTYPE html>












  


<html class="theme-next gemini use-motion" lang="en">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2"/>
<meta name="theme-color" content="#222">












<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />



  <meta name="google-site-verification" content="3dnqtuxhGQK5nK7aqRtU2c3DzZlaVtBMN6M39UoGcgY" />




















<link href="/blog/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/blog/css/main.css?v=6.3.0" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/blog/images/apple-touch-icon-next.png?v=6.3.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/blog/images/favicon-32x32-next.png?v=6.3.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/blog/images/favicon-16x16-next.png?v=6.3.0">


  <link rel="mask-icon" href="/blog/images/logo.svg?v=6.3.0" color="#222">









<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/blog/',
    scheme: 'Gemini',
    version: '6.3.0',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":true,"onmobile":false},
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="ID mapping is annoying but we have to face very often. This note is a collection of methods to deal with this trouble.  R (Bioconductor) There are lots of annotation packages in Bioconductor and they">
<meta name="keywords" content="ID conversion,bioconductor,biomaRt,OrgDb,API">
<meta property="og:type" content="article">
<meta property="og:title" content="Biological ID Conversion">
<meta property="og:url" content="https://yiweiniu.github.io/blog/2018/09/Biological-ID-Conversion/index.html">
<meta property="og:site_name" content="Yiwei Niu&#39;s Note">
<meta property="og:description" content="ID mapping is annoying but we have to face very often. This note is a collection of methods to deal with this trouble.  R (Bioconductor) There are lots of annotation packages in Bioconductor and they">
<meta property="og:locale" content="en">
<meta property="og:updated_time" content="2018-09-18T08:22:22.499Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Biological ID Conversion">
<meta name="twitter:description" content="ID mapping is annoying but we have to face very often. This note is a collection of methods to deal with this trouble.  R (Bioconductor) There are lots of annotation packages in Bioconductor and they">



  <link rel="alternate" href="/blog/atom.xml" title="Yiwei Niu's Note" type="application/atom+xml" />




  <link rel="canonical" href="https://yiweiniu.github.io/blog/2018/09/Biological-ID-Conversion/"/>



<script type="text/javascript" id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>Biological ID Conversion | Yiwei Niu's Note</title>
  




<script async src="https://www.googletagmanager.com/gtag/js?id=UA-122074944-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-122074944-1');
</script>






  <noscript>
  <style type="text/css">
    .use-motion .motion-element,
    .use-motion .brand,
    .use-motion .menu-item,
    .sidebar-inner,
    .use-motion .post-block,
    .use-motion .pagination,
    .use-motion .comments,
    .use-motion .post-header,
    .use-motion .post-body,
    .use-motion .collection-title { opacity: initial; }

    .use-motion .logo,
    .use-motion .site-title,
    .use-motion .site-subtitle {
      opacity: initial;
      top: initial;
    }

    .use-motion {
      .logo-line-before i { left: initial; }
      .logo-line-after i { right: initial; }
    }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="en">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/blog/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Yiwei Niu's Note</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
      
        <p class="site-subtitle">to share, to learn</p>
      
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="Toggle navigation bar">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home">
    <a href="/blog/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-home"></i> <br />Home</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-categories">
    <a href="/blog/categories/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-th"></i> <br />Categories</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-tags">
    <a href="/blog/tags" rel="section">
      <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />Tags</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">
    <a href="/blog/archives" rel="section">
      <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />Archives</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-about">
    <a href="https:/yiweiniu.github.io/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-user"></i> <br />About</a>
  </li>

      
      
    </ul>
  

  

  
</nav>



  



</div>
    </header>

    


    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
            

          
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="https://yiweiniu.github.io/blog/blog/2018/09/Biological-ID-Conversion/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Yiwei Niu">
      <meta itemprop="description" content="a beginner, a learner">
      <meta itemprop="image" content="/blog/images/portrait.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Yiwei Niu's Note">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">Biological ID Conversion
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2018-09-18 15:54:27" itemprop="dateCreated datePublished" datetime="2018-09-18T15:54:27+08:00">2018-09-18</time>
            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/blog/categories/bioinformatics/" itemprop="url" rel="index"><span itemprop="name">bioinformatics</span></a></span>

                
                
                  , 
                
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/blog/categories/bioinformatics/basic/" itemprop="url" rel="index"><span itemprop="name">basic</span></a></span>

                
                
              
            </span>
          

          
            
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
                <a href="/blog/2018/09/Biological-ID-Conversion/#comments" itemprop="discussionUrl">
                
                  <span class="post-comments-count disqus-comment-count"
                        data-disqus-identifier="2018/09/Biological-ID-Conversion/" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          

          
            <span class="post-meta-divider">|</span>
            <span class="post-meta-item-icon"
            >
            <i class="fa fa-eye"></i>
             Views:  
            <span class="busuanzi-value" id="busuanzi_value_page_pv" ></span>
            </span>
          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                <span title="Symbols count in article">16k</span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                <span title="Reading time">0:14</span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <p>ID mapping is annoying but we have to face very often. This note is a collection of methods to deal with this trouble.</p>
<h2 id="r-bioconductor"><a class="markdownIt-Anchor" href="#r-bioconductor"></a> R (Bioconductor)</h2>
<p>There are lots of annotation packages in Bioconductor and they contain various kinds of annotation we need and we don’t need. Different series of annotation packages may have different design purpose, and these differences should be considered when in practice.</p>
<p>For ID conversion, two main resources can be used: <code>biomaRt</code>, the R interface of <a href="http://www.biomart.org/" target="_blank" rel="noopener">BioMart</a>, and various specialized annotation packages.</p>
<h3 id="biomart"><a class="markdownIt-Anchor" href="#biomart"></a> biomaRt</h3>
<p>Ref: <a href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/biomaRt.html" target="_blank" rel="noopener">The biomaRt users guide</a></p>
<p><code>biomaRt</code> is a R interface to <a href="http://www.biomart.org/" target="_blank" rel="noopener">BioMart</a> databases. It’s very powerful and ID conversion is only one of many applications.</p>
<blockquote>
<p>The package enables retrieval of large amounts of data in a uniform way without the need to know the underlying database schemas or write complex SQL queries. Examples of BioMart databases are Ensembl, COSMIC, Uniprot, HGNC, Gramene, Wormbase and dbSNP mapped to Ensembl. These major databases give biomaRt users direct access to a diverse set of data and enable a wide range of powerful online queries from gene annotation to database mining. <em>via: <a href="http://www.bioconductor.org/packages/release/bioc/html/biomaRt.html" target="_blank" rel="noopener">http://www.bioconductor.org/packages/release/bioc/html/biomaRt.html</a></em></p>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">library</span>(<span class="string">"biomaRt"</span>)</span><br></pre></td></tr></table></figure>
<p>Display all available BioMart web services</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">listMarts()</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">               biomart               version</span><br><span class="line">1 ENSEMBL_MART_ENSEMBL      Ensembl Genes 93</span><br><span class="line">2   ENSEMBL_MART_MOUSE      Mouse strains 93</span><br><span class="line">3     ENSEMBL_MART_SNP  Ensembl Variation 93</span><br><span class="line">4 ENSEMBL_MART_FUNCGEN Ensembl Regulation 93</span><br></pre></td></tr></table></figure>
<p>Choose to query the Ensembl BioMart database.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ensembl=useMart(<span class="string">"ensembl"</span>)</span><br></pre></td></tr></table></figure>
<p>Look at which datasets are available in the selected BioMart by using the function</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">listDatasets(ensembl)[1:5, ]</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">                           dataset                                                  description                 version</span><br><span class="line">1       acarolinensis_gene_ensembl                               Anole lizard genes (AnoCar2.0)               AnoCar2.0</span><br><span class="line">2        amelanoleuca_gene_ensembl                                        Panda genes (ailMel1)                 ailMel1</span><br><span class="line">3          amexicanus_gene_ensembl                                  Cave fish genes (AstMex102)               AstMex102</span><br><span class="line">4          anancymaae_gene_ensembl                           Ma&apos;s night monkey genes (Anan_2.0)                Anan_2.0</span><br><span class="line">5      aplatyrhynchos_gene_ensembl                                    Duck genes (BGI_duck_1.0)            BGI_duck_1.0</span><br></pre></td></tr></table></figure>
<p>Update the Mart object using the function <code>useDataset()</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ensembl = useDataset(<span class="string">"hsapiens_gene_ensembl"</span>, mart=ensembl)</span><br></pre></td></tr></table></figure>
<p>Or alternatively if the dataset one wants to use is known in advance, we can select a BioMart database and dataset in one step by:</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ensembl = useMart(<span class="string">"ensembl"</span>, dataset=<span class="string">"hsapiens_gene_ensembl"</span>)</span><br></pre></td></tr></table></figure>
<p>Shows all available filters in the selected dataset</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">filters = listFilters(ensembl)</span><br><span class="line">filters[<span class="number">1</span>:<span class="number">5</span>,]</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">##              name              description</span><br><span class="line">## 1 chromosome_name Chromosome/scaffold name</span><br><span class="line">## 2           start                    Start</span><br><span class="line">## 3             end                      End</span><br><span class="line">## 4      band_start               Band Start</span><br><span class="line">## 5        band_end                 Band End</span><br></pre></td></tr></table></figure>
<p>Displays all available attributes in the selected dataset</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">attributes = listAttributes(ensembl)</span><br><span class="line">attributes[<span class="number">1</span>:<span class="number">5</span>,]</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">                           name                  description         page</span><br><span class="line">1               ensembl_gene_id               Gene stable ID feature_page</span><br><span class="line">2       ensembl_gene_id_version       Gene stable ID version feature_page</span><br><span class="line">3         ensembl_transcript_id         Transcript stable ID feature_page</span><br><span class="line">4 ensembl_transcript_id_version Transcript stable ID version feature_page</span><br><span class="line">5            ensembl_peptide_id            Protein stable ID feature_page</span><br></pre></td></tr></table></figure>
<p>Annotate a set of Affymetrix identifiers with HUGO symbol and chromosomal locations of corresponding genes</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">affyids=c(&quot;202763_at&quot;,&quot;209310_s_at&quot;,&quot;207500_at&quot;)</span><br><span class="line">getBM(attributes=c(&apos;affy_hg_u133_plus_2&apos;, &apos;entrezgene&apos;), </span><br><span class="line">       filters = &apos;affy_hg_u133_plus_2&apos;, </span><br><span class="line">       values = affyids, </span><br><span class="line">       mart = ensembl)</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">  affy_hg_u133_plus_2 entrezgene</span><br><span class="line">1         209310_s_at        837</span><br><span class="line">2           207500_at        838</span><br><span class="line">3           202763_at        836</span><br></pre></td></tr></table></figure>
<p>Retrieve all HUGO gene symbols of genes that are located on chromosomes 17,20 or Y, and are associated with specific GO terms.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">go=c(<span class="string">"GO:0051330"</span>,<span class="string">"GO:0000080"</span>,<span class="string">"GO:0000114"</span>,<span class="string">"GO:0000082"</span>)</span><br><span class="line">chrom=c(<span class="number">17</span>,<span class="number">20</span>,<span class="string">"Y"</span>)</span><br><span class="line">getBM(attributes= <span class="string">"hgnc_symbol"</span>,</span><br><span class="line">       filters=c(<span class="string">"go"</span>,<span class="string">"chromosome_name"</span>),</span><br><span class="line">       values=list(go, chrom), mart=ensembl)</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">  hgnc_symbol</span><br><span class="line">1     RPS6KB1</span><br><span class="line">2        CDC6</span><br><span class="line">3        RPA1</span><br><span class="line">4        CDK3</span><br><span class="line">5        MCM8</span><br><span class="line">6       CRLF3</span><br></pre></td></tr></table></figure>
<p>Annotate a set of EntrezGene identifiers with GO annotation.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">entrez=c(<span class="string">"673"</span>,<span class="string">"837"</span>)</span><br><span class="line">goids = getBM(attributes = c(<span class="string">'entrezgene'</span>, <span class="string">'go_id'</span>), </span><br><span class="line">               filters = <span class="string">'entrezgene'</span>, </span><br><span class="line">               values = entrez, </span><br><span class="line">               mart = ensembl)</span><br><span class="line">head(goids)</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">  entrezgene      go_id</span><br><span class="line">1        673 GO:0000166</span><br><span class="line">2        673 GO:0004672</span><br><span class="line">3        673 GO:0004674</span><br><span class="line">4        673 GO:0005524</span><br><span class="line">5        673 GO:0006468</span><br><span class="line">6        673</span><br></pre></td></tr></table></figure>
<p>Ensembl id to gene symbol and entrez id.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">library</span>(biomaRt)</span><br><span class="line">ensembl = useMart(<span class="string">"ensembl"</span>, dataset = <span class="string">"hsapiens_gene_ensembl"</span>)</span><br><span class="line"></span><br><span class="line">ensg = c(<span class="string">'ENSG00000242268.2'</span>, <span class="string">'ENSG00000158486.13'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># get stable id</span></span><br><span class="line">ensg.no_version = sapply(strsplit(as.character(ensg),<span class="string">"\\."</span>),<span class="string">"[["</span>,<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line">getBM(attributes = c(<span class="string">'ensembl_gene_id'</span>, <span class="string">'entrezgene'</span>, <span class="string">'hgnc_symbol'</span>), filters = <span class="string">'ensembl_gene_id'</span>, values=ensg.no_version, mart=ensembl)</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">  ensembl_gene_id entrezgene hgnc_symbol</span><br><span class="line">1 ENSG00000158486      55567       DNAH3</span><br><span class="line">2 ENSG00000242268         NA   LINC02082</span><br></pre></td></tr></table></figure>
<p>If you do not want to <code>NA</code>, use <code>na.omit</code> to remove those genes that can’t be transformed.</p>
<h3 id="orgdb-packages-bitr"><a class="markdownIt-Anchor" href="#orgdb-packages-bitr"></a> OrgDb packages + bitr</h3>
<p>Ref: <a href="https://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html#bitr-biological-id-translator" target="_blank" rel="noopener">clusterProfiler - bitr</a></p>
<p>orgDb packages are gene-centric annotation packages at organism level, such as <code>org.Hs.eg.db</code>, <code>org.Mmu.eg.db</code>.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">library</span>(org.Hs.eg.db)</span><br></pre></td></tr></table></figure>
<p>Here is <code>org.Hs.eg</code> package. We can see all the resources used to build this package.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">&gt; org.Hs.eg.db</span><br><span class="line">OrgDb object:</span><br><span class="line">| DBSCHEMAVERSION: <span class="number">2.1</span></span><br><span class="line">| Db type: OrgDb</span><br><span class="line">| Supporting package: AnnotationDbi</span><br><span class="line">| DBSCHEMA: HUMAN_DB</span><br><span class="line">| ORGANISM: Homo sapiens</span><br><span class="line">| SPECIES: Human</span><br><span class="line">| EGSOURCEDATE: <span class="number">2018</span>-Apr4</span><br><span class="line">| EGSOURCENAME: Entrez Gene</span><br><span class="line">| EGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA</span><br><span class="line">| CENTRALID: EG</span><br><span class="line">| TAXID: <span class="number">9606</span></span><br><span class="line">| GOSOURCENAME: Gene Ontology</span><br><span class="line">| GOSOURCEURL: ftp://ftp.geneontology.org/pub/go/godatabase/archive/latest-lite/</span><br><span class="line">| GOSOURCEDATE: <span class="number">2018</span>-Mar28</span><br><span class="line">| GOEGSOURCEDATE: <span class="number">2018</span>-Apr4</span><br><span class="line">| GOEGSOURCENAME: Entrez Gene</span><br><span class="line">| GOEGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA</span><br><span class="line">| KEGGSOURCENAME: KEGG GENOME</span><br><span class="line">| KEGGSOURCEURL: ftp://ftp.genome.jp/pub/kegg/genomes</span><br><span class="line">| KEGGSOURCEDATE: <span class="number">2011</span>-Mar15</span><br><span class="line">| GPSOURCENAME: UCSC Genome Bioinformatics (Homo sapiens)</span><br><span class="line">| GPSOURCEURL: </span><br><span class="line">| GPSOURCEDATE: <span class="number">2018</span>-Mar26</span><br><span class="line">| ENSOURCEDATE: <span class="number">2017</span>-Dec04</span><br><span class="line">| ENSOURCENAME: Ensembl</span><br><span class="line">| ENSOURCEURL: ftp://ftp.ensembl.org/pub/current_fasta</span><br><span class="line">| UPSOURCENAME: Uniprot</span><br><span class="line">| UPSOURCEURL: http://www.UniProt.org/</span><br><span class="line">| UPSOURCEDATE: Mon Apr  <span class="number">9</span> <span class="number">20</span>:<span class="number">58</span>:<span class="number">54</span> <span class="number">2018</span></span><br><span class="line"></span><br><span class="line">Please see: help(<span class="string">'select'</span>) <span class="keyword">for</span> usage information</span><br></pre></td></tr></table></figure>
<p>Use <code>keytypes()</code> to list all supporting types.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">keytypes(org.Hs.eg.db)</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"> [1] &quot;ACCNUM&quot;       &quot;ALIAS&quot;        &quot;ENSEMBL&quot;      &quot;ENSEMBLPROT&quot;  &quot;ENSEMBLTRANS&quot; &quot;ENTREZID&quot;     &quot;ENZYME&quot;       &quot;EVIDENCE&quot;    </span><br><span class="line"> [9] &quot;EVIDENCEALL&quot;  &quot;GENENAME&quot;     &quot;GO&quot;           &quot;GOALL&quot;        &quot;IPI&quot;          &quot;MAP&quot;          &quot;OMIM&quot;         &quot;ONTOLOGY&quot;    </span><br><span class="line">[17] &quot;ONTOLOGYALL&quot;  &quot;PATH&quot;         &quot;PFAM&quot;         &quot;PMID&quot;         &quot;PROSITE&quot;      &quot;REFSEQ&quot;       &quot;SYMBOL&quot;       &quot;UCSCKG&quot;      </span><br><span class="line">[25] &quot;UNIGENE&quot;      &quot;UNIPROT&quot;</span><br></pre></td></tr></table></figure>
<p>Key types supported by differenct packages can be different.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">keytypes(org.Ss.eg.db)</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"> [1] &quot;ACCNUM&quot;      &quot;ALIAS&quot;       &quot;ENTREZID&quot;    &quot;ENZYME&quot;      &quot;EVIDENCE&quot;    &quot;EVIDENCEALL&quot; &quot;GENENAME&quot;    &quot;GO&quot;          &quot;GOALL&quot;      </span><br><span class="line">[10] &quot;ONTOLOGY&quot;    &quot;ONTOLOGYALL&quot; &quot;PATH&quot;        &quot;PMID&quot;        &quot;REFSEQ&quot;      &quot;SYMBOL&quot;      &quot;UNIGENE&quot;     &quot;UNIPROT&quot;</span><br></pre></td></tr></table></figure>
<p>Convert Ensembl ids to entrez id and gene symbol.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">library</span>(clusterProfiler)</span><br><span class="line"><span class="keyword">library</span>(org.Hs.eg.db)</span><br><span class="line"></span><br><span class="line">ensg = c(<span class="string">'ENSG00000242268.2'</span>, <span class="string">'ENSG00000158486.13'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># remove version number</span></span><br><span class="line">ensg.no_version = sapply(strsplit(as.character(ensg),<span class="string">"\\."</span>),<span class="string">"[["</span>,<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line">bitr(ensg.no_version, fromType=<span class="string">"ENSEMBL"</span>, toType=c(<span class="string">"ENTREZID"</span>, <span class="string">"SYMBOL"</span>), OrgDb=<span class="string">"org.Hs.eg.db"</span>)</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&apos;select()&apos; returned 1:1 mapping between keys and columns</span><br><span class="line">          ENSEMBL  ENTREZID    SYMBOL</span><br><span class="line">1 ENSG00000242268 100507661 LINC02082</span><br><span class="line">2 ENSG00000158486     55567     DNAH3</span><br></pre></td></tr></table></figure>
<h2 id="ncbi-gene-data"><a class="markdownIt-Anchor" href="#ncbi-gene-data"></a> NCBI gene DATA</h2>
<p>Sometimes we want to have all information on local disks and use in-house scripts to do the conversion. <a href="ftp://ftp.ncbi.nih.gov/gene/DATA" target="_blank" rel="noopener">ftp://ftp.ncbi.nih.gov/gene/DATA</a> provide most up-to-date and comprehensive collections of gene-centric information.</p>
<blockquote>
<p>By incorporating the data from LocusLink in an Entrez database with gene-specific data from other species, you now have a single point of lookup for gene-specific information for the taxa within the scope of the RefSeq project. You also have more immediate access to related data that was cumbersome to maintain independent of Entrez, and can harness the power of Entrez-based tools such as Entrez Programming Utilities (E-Utilities) and MyNCBI.  <em>via: <a href="https://www.ncbi.nlm.nih.gov/entrez/query/static/help/LL2G.html#files" target="_blank" rel="noopener">https://www.ncbi.nlm.nih.gov/entrez/query/static/help/LL2G.html#files</a></em></p>
</blockquote>
<p>This <a href="ftp://ftp.ncbi.nih.gov/gene/DATA/README" target="_blank" rel="noopener">README</a> discribes all the files included. Here is a short summary.</p>
<table>
<thead>
<tr>
<th>Entrez Gene file name</th>
<th>Comments</th>
</tr>
</thead>
<tbody>
<tr>
<td>DATA/ASN_BINARY</td>
<td>Files in this directory contain comprehensive extractions from Entrez Gene in ASN.1 format.</td>
</tr>
<tr>
<td>DATA/GENE_INFO</td>
<td>extractions from Entrez Gene in the same format as the gene_info file. Each file contains a subset of data for the species or taxonomic group indicated by the file name.</td>
</tr>
<tr>
<td>DATA/expression</td>
<td>reports of normalized RNA expression levels computed from RNA-seq data for human, mouse, and rat genes.</td>
</tr>
<tr>
<td>gene2accession</td>
<td>a comprehensive report of the accessions that are related to a GeneID.  It includes sequences from the international sequence collaboration, Swiss-Prot, and RefSeq. The RefSeq subset of this file is also available as gene2refseq… If you want to convert any accessions into GeneIDs, this one file should suffice.</td>
</tr>
<tr>
<td>gene2ensembl</td>
<td>This file reports matches between NCBI and Ensembl annotation based on comparison of rna and protein features.</td>
</tr>
<tr>
<td>gene2vega</td>
<td>This file reports matches between NCBI and Vega annotation.</td>
</tr>
<tr>
<td>gene2go</td>
<td>GeneID/GO ID/Evidence Code. Consolidated summary based on gene_association files from the GO Consortium and Entrez Gene’s gene_info file.</td>
</tr>
<tr>
<td>gene2pubmed</td>
<td>gene2pubmed includes the identifier for the species of the GeneID (i.e. the Taxonomy ID).</td>
</tr>
<tr>
<td>gene2refseq</td>
<td>This file is the RefSeq subset of gene2accession. The file in Entrez Gene does not include information about secondary accessions. This function is now provided from the RefSeq ftp site, as documented in the current release notes: <a href="ftp://ftp.ncbi.nlm.nih.gov/refseq/release/release-notes/RefSeq-release#.txt" target="_blank" rel="noopener">ftp://ftp.ncbi.nlm.nih.gov/refseq/release/release-notes/RefSeq-release#.txt</a>, where # is the value of the current release number.</td>
</tr>
<tr>
<td>gene2sts</td>
<td>GeneID/UniSTS marker ID relationship</td>
</tr>
<tr>
<td>gene2unigene</td>
<td>GeneID/UniGene cluster relationship</td>
</tr>
<tr>
<td>gene_group</td>
<td>report of genes and their relationships to other genes</td>
</tr>
<tr>
<td>gene_orthologs</td>
<td>report of orthologous genes</td>
</tr>
<tr>
<td>gene_history</td>
<td>comprehensive information about GeneIDs that are no longer current</td>
</tr>
<tr>
<td>gene_info</td>
<td>GeneID, names, map locations, and database cross-reference.</td>
</tr>
<tr>
<td>gene_neighbors</td>
<td>reports neighboring genes for all genes placed on a given genomic sequence.</td>
</tr>
<tr>
<td>gene_refseq_uniprotkb_collab</td>
<td>report of the relationship between NCBI Reference Sequence protein accessions and UniProtKB protein accessions</td>
</tr>
<tr>
<td>mim2gene_medgen</td>
<td>report of the relationship between MIM numbers (OMIM), GeneIDs, and Records in MedGen</td>
</tr>
</tbody>
</table>
<h2 id="api"><a class="markdownIt-Anchor" href="#api"></a> API</h2>
<p>Many databases provide APIs to help access their data and some of them can be used for id conversion. But I do not recommend to use these APIs directly if one dose not want to spend much time on this job, as they can be changed over time and users have to be familiar with the data structure provided. Many commonly used APIs have external software or packages to access, and you may use Google to find them before using the APIs.</p>
<h3 id="ensembl-rest-api"><a class="markdownIt-Anchor" href="#ensembl-rest-api"></a> Ensembl REST API</h3>
<p><a href="http://rest.ensembl.org/" target="_blank" rel="noopener">Ensembl REST API</a> provides many user-friendly interfaces to retrive information. And there are three APIs for cross biological id mapping.</p>
<ul>
<li><a href="http://rest.ensembl.org/documentation/info/xref_external" target="_blank" rel="noopener">GET xrefs/symbol/:species/:symbol</a> looks up an external symbol and returns all Ensembl objects linked to it.</li>
<li><a href="http://rest.ensembl.org/documentation/info/xref_id" target="_blank" rel="noopener">GET xrefs/id/:id</a> performs lookups of Ensembl Identifiers and retrieve their external references in other databases.</li>
<li><a href="http://rest.ensembl.org/documentation/info/xref_name" target="_blank" rel="noopener">GET xrefs/name/:species/:name</a> performs a lookup based upon the primary accession or display label of an external reference and returning the information we hold about the entry.</li>
</ul>
<p>I guess <code>biomaRt</code> aforementioned is actually a well-capsulated software that communicates with databases through APIs.</p>
<h3 id="kegg-api"><a class="markdownIt-Anchor" href="#kegg-api"></a> KEGG API</h3>
<p><a href="https://www.kegg.jp/kegg/rest/keggapi.html" target="_blank" rel="noopener">KEGG API</a> is a REST-stype Application Programming Interface to the KEGG database resource.</p>
<p>We can use this API by <code>bitr_kegg</code> in <code>clusterProfiler</code> package or <code>KEGGREST</code> package.</p>
<h4 id="bitr_kegg"><a class="markdownIt-Anchor" href="#bitr_kegg"></a> bitr_kegg</h4>
<p>Ref: <a href="https://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html#bitr_kegg-converting-biological-ids-using-kegg-api" target="_blank" rel="noopener">https://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html#bitr_kegg-converting-biological-ids-using-kegg-api</a></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">library</span>(clusterProfiler)</span><br><span class="line"></span><br><span class="line">hg = c(<span class="string">"4597"</span>, <span class="string">"7111"</span>, <span class="string">"5266"</span>, <span class="string">"2175"</span>, <span class="string">"755"</span>, <span class="string">"23046"</span>)</span><br><span class="line"></span><br><span class="line">bitr_kegg(hg, fromType=<span class="string">'kegg'</span>, toType=<span class="string">'ncbi-proteinid'</span>, organism=<span class="string">'hsa'</span>)</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">   kegg ncbi-proteinid</span><br><span class="line">1  2175      NP_000126</span><br><span class="line">2 23046   NP_001239029</span><br><span class="line">3  4597      NP_002452</span><br><span class="line">4  5266      NP_002629</span><br><span class="line">5  7111   NP_001159588</span><br><span class="line">6   755      NP_004919</span><br></pre></td></tr></table></figure>
<blockquote>
<p>The ID type (both <code>fromType</code> &amp; <code>toType</code>) should be one of ‘kegg’, ‘ncbi-geneid’, ‘ncbi-proteinid’ or ‘uniprot’. The ‘kegg’ is the primary ID used in KEGG database. The data source of KEGG was from NCBI. A rule of thumb for the ‘kegg’ ID is <code>entrezgene</code> ID for eukaryote species and <code>Locus</code> ID for prokaryotes.</p>
</blockquote>
<blockquote>
<p><mark>Many prokaryote species don’t have entrezgene ID available</mark>. For example we can check the gene information of <code>ece:Z5100</code> in <a href="http://www.genome.jp/dbget-bin/www_bget?ece:Z5100" target="_blank" rel="noopener">http://www.genome.jp/dbget-bin/www_bget?ece:Z5100</a>, which have <code>NCBI-ProteinID</code> and <code>UnitProt</code> links in the <code>Other DBs</code> Entry, but not <code>NCBI-GeneID</code>.</p>
</blockquote>
<blockquote>
<p>The full list of KEGG supported organisms can be accessed via <a href="http://www.genome.jp/kegg/catalog/org_list.html" target="_blank" rel="noopener">http://www.genome.jp/kegg/catalog/org_list.html</a>.</p>
</blockquote>
<h4 id="keggrest"><a class="markdownIt-Anchor" href="#keggrest"></a> KEGGREST</h4>
<p><a href="http://bioconductor.org/packages/KEGGREST/" target="_blank" rel="noopener">KEGGREST</a> provides a client interface to the KEGG REST server. And <code>keggConv()</code> can be used for converting identifiers.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">library</span>(KEGGREEST)</span><br></pre></td></tr></table></figure>
<p>Convert between KEGG identifiers and outside identifiers.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">keggConv(<span class="string">"ncbi-proteinid"</span>, c(<span class="string">"hsa:10458"</span>, <span class="string">"ece:Z5100"</span>))</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">                 hsa:10458                  ece:Z5100 </span><br><span class="line">&quot;ncbi-proteinid:NP_059345&quot;  &quot;ncbi-proteinid:AAG58814&quot;</span><br></pre></td></tr></table></figure>
<p>…or get the mapping for an entire species:</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">head(keggConv(<span class="string">"eco"</span>, <span class="string">"ncbi-geneid"</span>))</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">ncbi-geneid:944742 ncbi-geneid:945803 ncbi-geneid:947498 ncbi-geneid:945198 ncbi-geneid:944747 ncbi-geneid:944749 </span><br><span class="line">       &quot;eco:b0001&quot;        &quot;eco:b0002&quot;        &quot;eco:b0003&quot;        &quot;eco:b0004&quot;        &quot;eco:b0005&quot;        &quot;eco:b0006&quot;</span><br></pre></td></tr></table></figure>
<p>Reversing the arguments does the opposite mapping:</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">head(keggConv(<span class="string">"ncbi-geneid"</span>, <span class="string">"eco"</span>))</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">           eco:b0001            eco:b0002            eco:b0003            eco:b0004            eco:b0005            eco:b0006 </span><br><span class="line">&quot;ncbi-geneid:944742&quot; &quot;ncbi-geneid:945803&quot; &quot;ncbi-geneid:947498&quot; &quot;ncbi-geneid:945198&quot; &quot;ncbi-geneid:944747&quot; &quot;ncbi-geneid:944749&quot;</span><br></pre></td></tr></table></figure>
<h2 id="web-server"><a class="markdownIt-Anchor" href="#web-server"></a> Web Server</h2>
<ul>
<li><a href="https://david.ncifcrf.gov/conversion.jsp" target="_blank" rel="noopener">DAVID - Gene ID Conversion Tool</a>. easy to use, but a bit old.</li>
<li><a href="http://www.ensembl.org/biomart/martview" target="_blank" rel="noopener">BioMart - Ensembl</a>. up-to-date and powerful.</li>
</ul>
<h2 id="change-log"><a class="markdownIt-Anchor" href="#change-log"></a> Change log</h2>
<ul>
<li>20180918: create the note.</li>
</ul>

      
    </div>

    

    
    
    

    

    

    
      <div>
        <ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>Post author:  </strong>Yiwei Niu</li>
  <li class="post-copyright-link">
    <strong>Post link: </strong>
    <a href="https://yiweiniu.github.io/blog/2018/09/Biological-ID-Conversion/" title="Biological ID Conversion">https://yiweiniu.github.io/blog/2018/09/Biological-ID-Conversion/</a>
  </li>
  <li class="post-copyright-license">
    <strong>Copyright Notice:  </strong>All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="external nofollow" target="_blank">CC BY-NC-SA 4.0</a> unless stating additionally.</li>
</ul>

      </div>
    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/blog/tags/ID-conversion/" rel="tag"># ID conversion</a>
          
            <a href="/blog/tags/bioconductor/" rel="tag"># bioconductor</a>
          
            <a href="/blog/tags/biomaRt/" rel="tag"># biomaRt</a>
          
            <a href="/blog/tags/OrgDb/" rel="tag"># OrgDb</a>
          
            <a href="/blog/tags/API/" rel="tag"># API</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/blog/2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/" rel="next" title="Remove Microbial Contamination in Reads">
                <i class="fa fa-chevron-left"></i> Remove Microbial Contamination in Reads
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>


  </div>


          </div>
          

  
    <div class="comments" id="comments">
      <div id="disqus_thread">
        <noscript>
          Please enable JavaScript to view the
          <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a>
        </noscript>
      </div>
    </div>

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            Overview
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image"
                src="/blog/images/portrait.jpg"
                alt="Yiwei Niu" />
            
              <p class="site-author-name" itemprop="name">Yiwei Niu</p>
              <p class="site-description motion-element" itemprop="description">a beginner, a learner</p>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/blog/archives">
                
                    <span class="site-state-item-count">20</span>
                    <span class="site-state-item-name">posts</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-categories">
                  <a href="/blog/categories/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">17</span>
                    <span class="site-state-item-name">categories</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-tags">
                  <a href="/blog/tags/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">35</span>
                    <span class="site-state-item-name">tags</span>
                  </a>
                </div>
              
            </nav>
          

          
            <div class="feed-link motion-element">
              <a href="/blog/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          
            <div class="links-of-author motion-element">
              
                <span class="links-of-author-item">
                  <a href="mailto:xiaohuwangwang@qq.com" target="_blank" title="E-Mail"><i class="fa fa-fw fa-envelope"></i>E-Mail</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="https://github.com/YiweiNiu" target="_blank" title="GitHub"><i class="fa fa-fw fa-github"></i>GitHub</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="http://weibo.com/u/3049858680" target="_blank" title="Weibo"><i class="fa fa-fw fa-weibo"></i>Weibo</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="https://twitter.com/ywniu" target="_blank" title="Twitter"><i class="fa fa-fw fa-twitter"></i>Twitter</a>
                  
                </span>
              
            </div>
          

          
          

          
          
            <div class="links-of-blogroll motion-element links-of-blogroll-block">
              <div class="links-of-blogroll-title">
                <i class="fa  fa-fw fa-link"></i>
                Links
              </div>
              <ul class="links-of-blogroll-list">
                
                  <li class="links-of-blogroll-item">
                    <a href="https://kaigedong.github.io/" title="Kaige Dong" target="_blank">Kaige Dong</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://homolog.us/blogs/" title="Homolog.us" target="_blank">Homolog.us</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="https://cgatoxford.wordpress.com/" title="CGAT" target="_blank">CGAT</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://www.gettinggeneticsdone.com/" title="Getting Genetics Done" target="_blank">Getting Genetics Done</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://www.acgt.me/" title="ACGT" target="_blank">ACGT</a>
                  </li>
                
              </ul>
            </div>
          

          
            
          
          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#r-bioconductor"><span class="nav-number">1.</span> <span class="nav-text"> R (Bioconductor)</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#biomart"><span class="nav-number">1.1.</span> <span class="nav-text"> biomaRt</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#orgdb-packages-bitr"><span class="nav-number">1.2.</span> <span class="nav-text"> OrgDb packages + bitr</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#ncbi-gene-data"><span class="nav-number">2.</span> <span class="nav-text"> NCBI gene DATA</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#api"><span class="nav-number">3.</span> <span class="nav-text"> API</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#ensembl-rest-api"><span class="nav-number">3.1.</span> <span class="nav-text"> Ensembl REST API</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#kegg-api"><span class="nav-number">3.2.</span> <span class="nav-text"> KEGG API</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#bitr_kegg"><span class="nav-number">3.2.1.</span> <span class="nav-text"> bitr_kegg</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#keggrest"><span class="nav-number">3.2.2.</span> <span class="nav-text"> KEGGREST</span></a></li></ol></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#web-server"><span class="nav-number">4.</span> <span class="nav-text"> Web Server</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#change-log"><span class="nav-number">5.</span> <span class="nav-text"> Change log</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; 2017 &mdash; <span itemprop="copyrightYear">2018</span>
  <span class="with-love" id="animate">
    <i class="fa fa-mars"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Yiwei Niu</span>

  

  
</div>




  <div class="powered-by">Powered by <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a></div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">Theme &mdash; <a class="theme-link" target="_blank" href="https://theme-next.org">NexT.Gemini</a></div>




        
<div class="busuanzi-count">
  <script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script>

  
    <span class="site-uv" title="Total Visitors">
      <i class="fa fa-user"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_uv"></span>
    </span>
  

  
    <span class="site-pv" title="Total Views">
      <i class="fa fa-eye"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_pv"></span>
    </span>
  
</div>









        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
          <span id="scrollpercent"><span>0</span>%</span>
        
      </div>
    

    
	
    

    
      <div>
        <!-- Go to www.addthis.com/dashboard to customize your tools -->
<div class="addthis_inline_share_toolbox">
  <script type = "text/javascript" src = "//s7.addthis.com/js/300/addthis_widget.js#pubid=ra-5a7ad07b7823d03b" async = "async" ></script>
</div>

      </div>
    
  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>


























  
  
    <script type="text/javascript" src="/blog/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  


  


  <script type="text/javascript" src="/blog/js/src/utils.js?v=6.3.0"></script>

  <script type="text/javascript" src="/blog/js/src/motion.js?v=6.3.0"></script>



  
  


  <script type="text/javascript" src="/blog/js/src/affix.js?v=6.3.0"></script>

  <script type="text/javascript" src="/blog/js/src/schemes/pisces.js?v=6.3.0"></script>



  
  <script type="text/javascript" src="/blog/js/src/scrollspy.js?v=6.3.0"></script>
<script type="text/javascript" src="/blog/js/src/post-details.js?v=6.3.0"></script>



  


  <script type="text/javascript" src="/blog/js/src/bootstrap.js?v=6.3.0"></script>



  

  
    <script id="dsq-count-scr" src="https://yiwei-nius-note.disqus.com/count.js" async></script>
  

  
    <script type="text/javascript">
      var disqus_config = function () {
        this.page.url = 'https://yiweiniu.github.io/blog/2018/09/Biological-ID-Conversion/';
        this.page.identifier = '2018/09/Biological-ID-Conversion/';
        this.page.title = 'Biological ID Conversion';
        };
      function loadComments () {
        var d = document, s = d.createElement('script');
        s.src = 'https://yiwei-nius-note.disqus.com/embed.js';
        s.setAttribute('data-timestamp', '' + +new Date());
        (d.head || d.body).appendChild(s);
      }
      
        loadComments();
      
    </script>
  












  





  

  

  

  

  
  

  

  

  

  

  
  <style>
    .copy-btn {
      display: inline-block;
      padding: 6px 12px;
      font-size: 13px;
      font-weight: 700;
      line-height: 20px;
      color: #333;
      white-space: nowrap;
      vertical-align: middle;
      cursor: pointer;
      background-color: #eee;
      background-image: linear-gradient(#fcfcfc, #eee);
      border: 1px solid #d5d5d5;
      border-radius: 3px;
      user-select: none;
      outline: 0;
    }

    .highlight-wrap .copy-btn {
      transition: opacity .3s ease-in-out;
      opacity: 0;
      padding: 2px 6px;
      position: absolute;
      right: 4px;
      top: 8px;
    }

    .highlight-wrap:hover .copy-btn,
    .highlight-wrap .copy-btn:focus {
      opacity: 1
    }

    .highlight-wrap {
      position: relative;
    }
  </style>
  <script>
    $('.highlight').each(function (i, e) {
      var $wrap = $('<div>').addClass('highlight-wrap')
      $(e).after($wrap)
      $wrap.append($('<button>').addClass('copy-btn').append('Copy').on('click', function (e) {
        var code = $(this).parent().find('.code').find('.line').map(function (i, e) {
          return $(e).text()
        }).toArray().join('\n')
        var ta = document.createElement('textarea')
        document.body.appendChild(ta)
        ta.style.position = 'absolute'
        ta.style.top = '0px'
        ta.style.left = '0px'
        ta.value = code
        ta.select()
        ta.focus()
        var result = document.execCommand('copy')
        document.body.removeChild(ta)
        
          if(result)$(this).text('Copied')
          else $(this).text('Copy failed')
        
        $(this).blur()
      })).on('mouseleave', function (e) {
        var $b = $(this).find('.copy-btn')
        setTimeout(function () {
          $b.text('Copy')
        }, 300)
      }).append(e)
    })
  </script>


</body>
</html>
