<!DOCTYPE html>












  


<html class="theme-next gemini use-motion" lang="en">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2"/>
<meta name="theme-color" content="#222">












<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />



  <meta name="google-site-verification" content="3dnqtuxhGQK5nK7aqRtU2c3DzZlaVtBMN6M39UoGcgY" />




















<link href="/blog/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/blog/css/main.css?v=6.3.0" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/blog/images/apple-touch-icon-next.png?v=6.3.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/blog/images/favicon-32x32-next.png?v=6.3.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/blog/images/favicon-16x16-next.png?v=6.3.0">


  <link rel="mask-icon" href="/blog/images/logo.svg?v=6.3.0" color="#222">









<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/blog/',
    scheme: 'Gemini',
    version: '6.3.0',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":true,"onmobile":false},
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="Purpose in short: I’ve got both illumina (PE and MPE) and PacBio reads of an insect for de novo geome assembly. Since whole bodies were used for extracted DNA, I thought there was some contamination i">
<meta name="keywords" content="bio-tools,contamination,reads,BBDuk,MashMap,long-reads,long-reads alignment">
<meta property="og:type" content="article">
<meta property="og:title" content="Remove Microbial Contamination in Reads">
<meta property="og:url" content="https://yiweiniu.github.io/blog/2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/index.html">
<meta property="og:site_name" content="Yiwei Niu&#39;s Note">
<meta property="og:description" content="Purpose in short: I’ve got both illumina (PE and MPE) and PacBio reads of an insect for de novo geome assembly. Since whole bodies were used for extracted DNA, I thought there was some contamination i">
<meta property="og:locale" content="en">
<meta property="og:image" content="https://yiweiniu.github.io/blog/2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/blast_1532401032_2561.png">
<meta property="og:image" content="https://yiweiniu.github.io/blog/2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/1532401743_26861.png">
<meta property="og:image" content="https://yiweiniu.github.io/blog/2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/1532401866_17878.png">
<meta property="og:updated_time" content="2018-08-16T05:37:35.411Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Remove Microbial Contamination in Reads">
<meta name="twitter:description" content="Purpose in short: I’ve got both illumina (PE and MPE) and PacBio reads of an insect for de novo geome assembly. Since whole bodies were used for extracted DNA, I thought there was some contamination i">
<meta name="twitter:image" content="https://yiweiniu.github.io/blog/2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/blast_1532401032_2561.png">



  <link rel="alternate" href="/blog/atom.xml" title="Yiwei Niu's Note" type="application/atom+xml" />




  <link rel="canonical" href="https://yiweiniu.github.io/blog/2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/"/>



<script type="text/javascript" id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>Remove Microbial Contamination in Reads | Yiwei Niu's Note</title>
  




<script async src="https://www.googletagmanager.com/gtag/js?id=UA-122074944-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-122074944-1');
</script>






  <noscript>
  <style type="text/css">
    .use-motion .motion-element,
    .use-motion .brand,
    .use-motion .menu-item,
    .sidebar-inner,
    .use-motion .post-block,
    .use-motion .pagination,
    .use-motion .comments,
    .use-motion .post-header,
    .use-motion .post-body,
    .use-motion .collection-title { opacity: initial; }

    .use-motion .logo,
    .use-motion .site-title,
    .use-motion .site-subtitle {
      opacity: initial;
      top: initial;
    }

    .use-motion {
      .logo-line-before i { left: initial; }
      .logo-line-after i { right: initial; }
    }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="en">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/blog/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Yiwei Niu's Note</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
      
        <p class="site-subtitle">to share, to learn</p>
      
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="Toggle navigation bar">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home">
    <a href="/blog/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-home"></i> <br />Home</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-categories">
    <a href="/blog/categories/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-th"></i> <br />Categories</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-tags">
    <a href="/blog/tags" rel="section">
      <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />Tags</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">
    <a href="/blog/archives" rel="section">
      <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />Archives</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-about">
    <a href="https:/yiweiniu.github.io/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-user"></i> <br />About</a>
  </li>

      
      
    </ul>
  

  

  
</nav>



  



</div>
    </header>

    


    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
            

          
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="https://yiweiniu.github.io/blog/blog/2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Yiwei Niu">
      <meta itemprop="description" content="a beginner, a learner">
      <meta itemprop="image" content="/blog/images/portrait.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Yiwei Niu's Note">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">Remove Microbial Contamination in Reads
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2018-07-25 18:37:59" itemprop="dateCreated datePublished" datetime="2018-07-25T18:37:59+08:00">2018-07-25</time>
            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/blog/categories/reads/" itemprop="url" rel="index"><span itemprop="name">reads</span></a></span>

                
                
                  , 
                
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/blog/categories/reads/contamination/" itemprop="url" rel="index"><span itemprop="name">contamination</span></a></span>

                
                
              
            </span>
          

          
            
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
                <a href="/blog/2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/#comments" itemprop="discussionUrl">
                
                  <span class="post-comments-count disqus-comment-count"
                        data-disqus-identifier="2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          

          
            <span class="post-meta-divider">|</span>
            <span class="post-meta-item-icon"
            >
            <i class="fa fa-eye"></i>
             Views:  
            <span class="busuanzi-value" id="busuanzi_value_page_pv" ></span>
            </span>
          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                <span title="Symbols count in article">32k</span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                <span title="Reading time">0:29</span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <p>Purpose in short: I’ve got both illumina (PE and MPE) and PacBio reads of an insect for <em>de novo</em> geome assembly. Since whole bodies were used for extracted DNA, I thought there was some contamination in the raw sequencing reads, so I want to remove them before assembling.</p>
<p>The main tool for illumina reads I used was <code>BBDuk</code>, which belongs to the <a href="https://sourceforge.net/projects/bbmap/" target="_blank" rel="noopener">BBTools suite</a>. <code>Seal</code> and <code>BBSplit</code> can also do this, but <code>BBDuk</code> is the most suitable for my purpose. See discussions here: <a href="https://www.biostars.org/p/165059/" target="_blank" rel="noopener">Question: How to remove contamination from the transcriptome assembly</a>. <code>BBTools</code> is quite versatile, fast, and convenient! Though now I don’t fully understand every kit in it, the experience was very good from the tools I’ve tried. The author Brian Bushnell is very active and responsive. Here is a handy summary of <code>BBMap</code>: <a href="http://seqanswers.com/forums/showthread.php?t=58221" target="_blank" rel="noopener">Yes … BBMap can do that!</a>.</p>
<p>But for PacBio long reads, I didn’t find any tools specialized for that. I tried several tools including <code>mapPacBio.sh</code> (based on <code>bbmap</code>, and is tuned for the error profile of long reads.), <code>blasr</code>, <code>MashMap</code>, and <code>minimap2</code>. At last, I found <code>minimap2</code> best met my needs.</p>
<h2 id="prepare-the-contamination-library"><a class="markdownIt-Anchor" href="#prepare-the-contamination-library"></a> Prepare the contamination library</h2>
<ul>
<li>
<p>Download all the sequences of bacteria, viral, fungi, protozoa, and archaea using <a href="https://github.com/sschmeier/refseq2kraken" target="_blank" rel="noopener">refseq2kraken</a>. See note <a href="/blog/2018/06/Detect-Microbial-Contamination-in-Contigs-by-Kraken/" title="Detect Microbial Contamination in Contigs by Kraken">Detect Microbial Contamination in Contigs by Kraken</a> for details. We can also use <a href="https://github.com/kblin/ncbi-genome-download" target="_blank" rel="noopener">ncbi-genome-download</a> to download them.</p>
  <figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> folder structures</span></span><br><span class="line"><span class="meta">$</span><span class="bash"> ls genomes/refseq/</span></span><br><span class="line">archaea  bacteria  fungi  protozoa  viral</span><br></pre></td></tr></table></figure>
</li>
<li>
<p>Merge all the sequences into one single Fasta.</p>
</li>
<li>
<p>Since I’ve got the mitochondrion DNA sequences of the target organism, I added them in the contaminant library to remove reads from mtDNA too. This library is refered as to <code>$contaminants</code> below.</p>
</li>
<li>
<p>After discussing with the author of <code>MashMap</code>, I included several insects’ genomes of the same order of my target insect to improve the specificity of aligners. This library is refered as to <code>$contaminants2</code>.</p>
</li>
</ul>
<h2 id="bbduk-for-short-reads"><a class="markdownIt-Anchor" href="#bbduk-for-short-reads"></a> BBDuk for short reads</h2>
<p><a href="http://seqanswers.com/forums/showthread.php?t=42776" target="_blank" rel="noopener">BBDuk</a> is a member of the <a href="https://sourceforge.net/projects/bbmap/" target="_blank" rel="noopener">BBTools</a> package. “Duk” stands for Decontamination Using Kmers. BBDuk is extremely fast, scalable, and memory-efficient, while maintaining greater sensitivity and specificity than other tools.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">WORKDIR=/parastor300/niuyw/Project/beetle_genome_171231/data/second</span><br><span class="line"></span><br><span class="line">ANNODIR=/home/niuyw/RefData</span><br><span class="line">TOOLDIR=/home/niuyw/software</span><br><span class="line">path2java=$TOOLDIR/jre1.8.0_111/bin/java</span><br><span class="line"></span><br><span class="line">DATADIR=/home/niuyw/Project/beetle_genome_171231/data/second</span><br><span class="line"></span><br><span class="line">PPN=24</span><br><span class="line"></span><br><span class="line">path2bbduk=$TOOLDIR/bbmap/bbduk.sh</span><br><span class="line">contaminants=$ANNODIR/NCBI_contaminants/contaminants_4_bettle_genome.fa</span><br><span class="line"></span><br><span class="line">for sample in 270B 500B 800B 3k_1 5k-1 5k-2 10k</span><br><span class="line">do</span><br><span class="line"><span class="meta">$</span><span class="bash">path2bbduk ref=<span class="variable">$&#123;contaminants&#125;</span> threads=<span class="variable">$&#123;PPN&#125;</span> ordered=t k=31 <span class="keyword">in</span>=<span class="variable">$&#123;DATADIR&#125;</span>/TrimGalore/<span class="variable">$&#123;sample&#125;</span>_R1_val_1.fq.gz in2=<span class="variable">$&#123;DATADIR&#125;</span>/TrimGalore/<span class="variable">$&#123;sample&#125;</span>_R2_val_2.fq.gz out=<span class="variable">$&#123;DATADIR&#125;</span>/bbduk/<span class="variable">$&#123;sample&#125;</span>.R1.fq out2=<span class="variable">$&#123;DATADIR&#125;</span>/bbduk/<span class="variable">$&#123;sample&#125;</span>.R2.fq outm=<span class="variable">$&#123;DATADIR&#125;</span>/bbduk/<span class="variable">$&#123;sample&#125;</span>.bad.R1.fq outm2=<span class="variable">$&#123;DATADIR&#125;</span>/bbduk/<span class="variable">$&#123;sample&#125;</span>.bad.R2.fq</span></span><br><span class="line">done</span><br></pre></td></tr></table></figure>
<p>The ‘good’ reads will be in <code>${sample}.R*.fq</code>, and the ‘bad’ ones will be in <code>${sample}.bad.R*.fq</code>.</p>
<h2 id="long-reads"><a class="markdownIt-Anchor" href="#long-reads"></a> Long reads</h2>
<h3 id="mappacbio"><a class="markdownIt-Anchor" href="#mappacbio"></a> mapPacBio</h3>
<p><a href="http://seqanswers.com/forums/showthread.php?p=133568#post133568" target="_blank" rel="noopener">mapPacBio.sh</a> is based on <code>bbmap</code>, and is tuned for the error profile of long reads.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">WORKDIR=/parastor300/niuyw/Project/beetle_genome_171231</span><br><span class="line"></span><br><span class="line">ANNODIR=/home/niuyw/RefData</span><br><span class="line"></span><br><span class="line">TOOLDIR=/home/niuyw/software</span><br><span class="line">path2java=$TOOLDIR/jre1.8.0_111/bin/java</span><br><span class="line"></span><br><span class="line">DATADIR=/home/niuyw/Project/beetle_genome_171231/data/third</span><br><span class="line"></span><br><span class="line">PPN=24</span><br><span class="line"></span><br><span class="line">path2bbmap=$TOOLDIR/bbmap</span><br><span class="line">contaminants=$ANNODIR/NCBI_contaminants/contaminants_4_bettle_genome.fa</span><br><span class="line"></span><br><span class="line"><span class="meta">$</span><span class="bash">path2bbmap/mapPacBio.sh threads=<span class="variable">$&#123;PPN&#125;</span> ref=<span class="variable">$&#123;contaminants&#125;</span> <span class="keyword">in</span>=third_all.fasta maxlen=6000 out=bbmap.sam</span></span><br><span class="line"><span class="meta">$</span><span class="bash">path2bbmap/reformat.sh unmappedonly <span class="keyword">in</span>=bbmap.sam out=good.fa</span></span><br></pre></td></tr></table></figure>
<p>The output of <code>bbmap</code> or <code>mapPacBio.sh</code> is <code>SAM</code>, and both mapped and unmapped reads are saved in one file. <code>reformat.sh</code> was used to extract mapped reads and transfrom it to <code>FASTA</code>. See this: <a href="https://www.biostars.org/p/233118/" target="_blank" rel="noopener">Question: bbmap command to extract mapped and unmapped pair end reads</a>.</p>
<p>But <code>mapPacBio</code> was very very very slow, even I used 24 threads. After about 25 days, the ouput was about 4.6G (my input was about 60G!), though it was still increasing, I killed the job.</p>
<h3 id="blasr"><a class="markdownIt-Anchor" href="#blasr"></a> blasr</h3>
<p><a href="https://github.com/PacificBiosciences/blasr" target="_blank" rel="noopener">blasr</a> may be the one of the first long-read aligner <sup class="footnote-ref"><a href="#fn1" id="fnref1">[1]</a></sup>. So I gave it a shot.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">$</span><span class="bash">TOOLDIR/anaconda2/bin/blasr third_all.fasta <span class="variable">$contaminants</span> --nproc <span class="variable">$PPN</span> --out blasr.bad --unaligned blasr.good</span></span><br></pre></td></tr></table></figure>
<p>Related: <a href="https://github.com/PacificBiosciences/blasr/issues/347" target="_blank" rel="noopener">output unmapped reads</a>.</p>
<p>But it went wrong:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">[INFO] 2018-07-02T18:56:18 [blasr] started.</span><br><span class="line">ERROR! Reading fasta files greater than 4Gbytes is not supported.</span><br></pre></td></tr></table></figure>
<p>What?! My fasta was about 60G. If I wanted use <code>blasr</code>, I had to split the input into small ones, but I didn’t want to.</p>
<h3 id="mashmap"><a class="markdownIt-Anchor" href="#mashmap"></a> MashMap</h3>
<p><a href="https://github.com/marbl/MashMap" target="_blank" rel="noopener">MashMap</a> is a fast approximate aligner for long DNA sequences <sup class="footnote-ref"><a href="#fn2" id="fnref2">[2]</a></sup>. It’s very fast and is easy to use.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">PPN=8</span><br><span class="line"><span class="meta">$</span><span class="bash">TOOLDIR/mashmap-Linux64-v2.0/mashmap -r <span class="variable">$contaminants</span> -q third_all.fasta -o mashmap.out</span></span><br></pre></td></tr></table></figure>
<p><code>-s</code> is the minimum query length (default is 5000), and <code>--pi</code> is the minimum identity to be reported (default is 85).</p>
<p>The output of <code>MashMap</code> is like this. Separated by space, it is <code>query name, length, 0-based start, end, strand, target name, length, start, end and mapping nucleotide identity</code> in turn.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">m161109_080520_42256_c101052872550000001823247601061737_s1_p0/104/24935_31494 6559 0 4999 - NC_037282.1 2038340 596479 601478 82.1711</span><br><span class="line">m161109_080520_42256_c101052872550000001823247601061737_s1_p0/357/31541_41029 9488 0 4999 - NC_004326.2 1343557 117176 122175 81.9933</span><br><span class="line">m161109_080520_42256_c101052872550000001823247601061737_s1_p0/562/0_12626 12626 7626 12625 + NC_001224.1 85779 50909 55908 82.2314</span><br></pre></td></tr></table></figure>
<p>Here is also a thread talking about contamination: <a href="https://github.com/marbl/MashMap/issues/6" target="_blank" rel="noopener">Decontamination of bacterial sequences in an assembly</a>. The author suggested that “One more thing that may be helpful for you is to include the representative genome (corresponding to your assembly) in the database as well. This would help improve the specificity of the method for correct portions of your assembly.”</p>
<p>But, the default parameters <code>MashMap</code> use are a bit loose, I wanted to use more strict parameters. When I lowered the <code>-s</code> and <code>--pi</code>, the program needed huge RAM and couldn’t run on our machine.</p>
<p>I’ve reported this issue to the author: <a href="https://github.com/marbl/MashMap/issues/6#issuecomment-402911053" target="_blank" rel="noopener">Decontamination of bacterial sequences in an assembly</a>.</p>
<p>And, there were also some questions about its alignment: <a href="https://github.com/marbl/MashMap/issues/12" target="_blank" rel="noopener">Questions about the alignment of MashMap</a>.</p>
<p>I first ran <code>MashMap</code> with different parameters and got several outputs:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> run1, with default parameters</span></span><br><span class="line"><span class="meta">$</span><span class="bash">path2mashmap -r <span class="variable">$contaminants</span> -q third_all.fasta -o mashmap.out</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> run2, with -s 2500 --pi 80</span></span><br><span class="line"><span class="meta">$</span><span class="bash">path2mashmap -t 8 -r <span class="variable">$contaminants</span> -q third_all.fasta -s 2500 --pi 80 -o mashmap2.out</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> run3, with -s 500 --pi 85</span></span><br><span class="line"><span class="meta">$</span><span class="bash">path2mashmap -t 8 -r <span class="variable">$contaminants</span> -q third_all.fasta -s 500 --pi 85 -o mashmap3.out</span></span><br></pre></td></tr></table></figure>
<p>And the outputs from three runs varied:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> there are 6633142 sequences of input</span></span><br><span class="line"><span class="meta">$</span><span class="bash"> grep -c <span class="string">'&gt;'</span> third_all.fasta</span></span><br><span class="line">6633142</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> run1</span></span><br><span class="line">cut -f 1 -d ' ' mashmap.out |sort|uniq|wc -l</span><br><span class="line">463569</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> run2</span></span><br><span class="line">cut -f 1 -d ' ' mashmap2.out |sort|uniq|wc -l</span><br><span class="line">2821004</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> run3</span></span><br><span class="line">cut -f 1 -d ' ' mashmap3.out |sort|uniq|wc -l</span><br><span class="line">6189307</span><br></pre></td></tr></table></figure>
<p>As can be seen, nearly all the sequences were aligned to contaminant library. That really shocked me!</p>
<p>Then I checked the top 10 sequences with highest identity and top 10 ones with loweset identity from the first run using <code>blastn</code>. The highest ones were fine. There were some differences between hits reported by blastn and MashMap, but maybe it’s because they used different databases. But the loweset ones were problematic. Most of them were ‘No significant similarity found’ when default parameters of <code>blastn</code> were used. And when I unselected ‘Low complexity regions’, the alignments were unreliable. There maybe something with ‘low complexity regions’ or ‘repeat’ things.</p>
<p>And the author explained:</p>
<blockquote>
<p>Mashmap identity is an estimate based on Jaccard similarity- not the precise identity; unfortunately the Jaccard-similarity based metric delivers poor specificity in cases when the source of reads is absent from database. See if you can include an insect reference genome (s) in the reference list to avoid this.</p>
</blockquote>
<p>Then I added several insects’ genomes of the same order of the target insect into the contaminant library and ran <code>MapshMap</code> again:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">$</span><span class="bash">path2mashmap -t 20 -r <span class="variable">$contaminants2</span> -q third_all.fasta -s 500 --pi 80 -o mashmap4.out</span></span><br></pre></td></tr></table></figure>
<p>For each read, it will be one of two states: mapped or unmapped, and for the mapped, it will be one of three states: mapped to insects (good), mapped to contaminants (bad) and mapped to both (ambivalent). So I counted reads in each categories:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">No. of total reads: 6633142</span><br><span class="line">  No. of reads in the mashmap.output: 6214500</span><br><span class="line">    No. of good: 853515</span><br><span class="line">    No. of bad: 138799</span><br><span class="line">    No. of ambivalent: 5222186</span><br><span class="line">  No. of reads not in the mashmap.output: 418642</span><br></pre></td></tr></table></figure>
<p>As can be seen, majority of reads had been mapped ambivalently, and here is a example of such reads:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">m161123_064622_42256_c101049952550000001823247601061783_s1_p0/52396/4594_8610 4016 3500 4015 + NW_017852934.1 2683736 1681820 1682319 79.4204</span><br><span class="line">m161123_064622_42256_c101049952550000001823247601061783_s1_p0/52396/4594_8610 4016 1000 1999 + NW_019280650.1 1003565 813077 813577 78.0766</span><br><span class="line">m161123_064622_42256_c101049952550000001823247601061783_s1_p0/52396/4594_8610 4016 2500 2999 - LJIG01019880.1 38067 34865 35364 79.3626</span><br><span class="line">m161123_064622_42256_c101049952550000001823247601061783_s1_p0/52396/4594_8610 4016 500 999 + NC_007418.3 31381287 24981081 24981580 79.5573</span><br><span class="line">m161123_064622_42256_c101049952550000001823247601061783_s1_p0/52396/4594_8610 4016 3000 3499 - kraken:taxid|76857|NZ_CP022123.1 2521394 1537365 1537864 81.3159</span><br><span class="line">m161123_064622_42256_c101049952550000001823247601061783_s1_p0/52396/4594_8610 4016 0 499 - kraken:taxid|1202539|NC_018417.1 157543 40864 41363 79.4397</span><br><span class="line">m161123_064622_42256_c101049952550000001823247601061783_s1_p0/52396/4594_8610 4016 1500 2499 + kraken:taxid|1936081|NZ_CP019389.1 3752836 1813582 1814081 76.7726</span><br></pre></td></tr></table></figure>
<p>So it’s very hard to extract good ones.</p>
<p>After all this, I got two points:</p>
<ul>
<li>I should add some insects’ genomes into the library to reduce the false positive hits.</li>
<li>Alignment from <code>MashMap</code> maybe not so reliable ().</li>
</ul>
<h3 id="minimap2"><a class="markdownIt-Anchor" href="#minimap2"></a> minimap2</h3>
<p><a href="https://github.com/lh3/minimap2" target="_blank" rel="noopener">minimap2</a> is a versatile pairwise aligner for genomic and spliced nucleotide sequences created by Heng Li <sup class="footnote-ref"><a href="#fn3" id="fnref3">[3]</a></sup>. It’s easy to use and runs very fast.</p>
<h4 id="mimimap2"><a class="markdownIt-Anchor" href="#mimimap2"></a> mimimap2</h4>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">PPN=24</span><br><span class="line"><span class="meta">$</span><span class="bash">TOOLDIR/minimap2-2.8_x64-linux/minimap2 -x map-pb <span class="variable">$contaminants</span> third_all.fasta -t <span class="variable">$PPN</span> -a -Q &gt; minimap.sam</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> -a: output the SAM format</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> -Q:not output base quality <span class="keyword">in</span> SAM</span></span><br></pre></td></tr></table></figure>
<p>I had about 60G input, and <code>minimap2</code> was so fast, finished after 77 CPU hours (4 real hours, 24 threads).</p>
<p>The output was very huge (~685G!), and it’s like this (truncated, and the sequences were replaced by ‘seq’.):</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">m54174_171023_074758/9962478/22561_26454  4 * 0 0 * * 0 0 seq *</span><br><span class="line">m54174_171023_074758/9962478/22561_26454  4 * 0 0 * * 0 0 seq *</span><br><span class="line">m54174_171023_074758/9962478/22561_26454  16  kraken:taxid|1094466|NC_017025.1  1304571 1 3661S5M2D6M4I8M1D4M1D13M1D11M1I16M1D13M1D13M1D27M1D4M1D4M2I4M2D12M2D9M1D2M1D13M1D12M1D14M1D13M1D6M2I7M1D7M  * 0 0 seq * NM:i:43 ms:i:220  AS:i:220  nn:i:0  tp:A:P  cm:i:3  s1:i:40 s2:i:0  dv:f:0.0200</span><br><span class="line">m54174_171023_074758/9962478/22561_26454  4 * 0 0 * * 0 0 seq *</span><br><span class="line">m54174_171023_074758/9962478/22561_26454  16  kraken:taxid|1323664|NZ_CP012748.1  2198444 1 903S12M1I12M1I12M1I4M1I6M1I2M1I12M1I39M1I9M1I6M5I4M4I7M1I2M1I5M1D6M1I12M1I12M1I12M1I13M1I11M2I5M1D8M2I22M1I13M1I23M1I12M1I12M1I10M1I3M2I10M3I5M1I6M1I3M1I16M1D6M1I12M2I6M1D3M1I2M1I7M1D4M1I5M1I7M1I12M1I6M7I3M1I3M1D6M1I11M1I12M1I12M2464S  * 0 0 seq * NM:i:95 ms:i:432  AS:i:432  nn:i:0  tp:A:P  cm:i:4  s1:i:49 s2:i:77 dv:f:0.0701 SA:Z:kraken:taxid|1323664|NZ_CP012748.1,2198449,-,1461S464M47I1921S,15,95;kraken:taxid|1323664|NZ_CP012748.1,2198443,-,1853S470M47I1523S,1,95;kraken:taxid|1323664|NZ_CP012748.1,2198444,-,683S469M51I2690S,4,106;kraken:taxid|1323664|NZ_CP012748.1,2198449,-,2305S464M17I1107S,16,105;kraken:taxid|1323664|NZ_CP012748.1,2198456,-,2607S450M11I825S,22,93;kraken:taxid|1323664|NZ_CP012748.1,2198451,-,430S460M63I2940S,3,105;kraken:taxid|1323664|NZ_CP012748.1,2198449,-,13S464M94I3322S,11,133;</span><br><span class="line">m54174_171023_074758/9962478/22561_26454  2064  kraken:taxid|1323664|NZ_CP012748.1  2198449 15  1461H7M1I12M1I7M1D4M1I12M1I4M1D10M1I10M1D24M1I18M2I9M1I8M1I12M2I9M1I2M6I7M1I13M5I5M1I7M1D4M1I10M1I2M1I12M1I12M1I12M2I13M1I11M1I11M1I5M1I7M1I5M1I8M1I12M1I5M1D4M1I2M1I4M2I2M1I7M5D5M1D7M2I21M1I24M4I8M1I12M1I12M1I4M1I20M1I12M1921H  * 0 0 seq * NM:i:95 ms:i:420  AS:i:420  nn:i:0  tp:A:P  cm:i:6  s1:i:80 s2:i:0  dv:f:0.0526 SA:Z:kraken:taxid|1323664|NZ_CP012748.1,2198444,-,903S469M57I2464S,1,95;kraken:taxid|1323664|NZ_CP012748.1,2198443,-,1853S470M47I1523S,1,95;kraken:taxid|1323664|NZ_CP012748.1,2198444,-,683S469M51I2690S,4,106;kraken:taxid|1323664|NZ_CP012748.1,2198449,-,2305S464M17I1107S,16,105;kraken:taxid|1323664|NZ_CP012748.1,2198456,-,2607S450M11I825S,22,93;kraken:taxid|1323664|NZ_CP012748.1,2198451,-,430S460M63I2940S,3,105;kraken:taxid|1323664|NZ_CP012748.1,2198449,-,13S464M94I3322S,11,133;</span><br><span class="line">m54174_171023_074758/9962478/22561_26454  272 kraken:taxid|1323664|NZ_CP012748.1  2198445 0 1010S9M1I2M1I4M1I8M1I12M1I5M1D6M1I12M1I39M1I11M2I3M1I10M1I22M1I13M1I23M1I12M1I12M1I10M1I3M2I11M1I1M2I3M1I6M1I3M1I16M1D6M1I13M4D7M1D5M2I7M1D4M1I5M1I6M1I13M1I6M1I9M1I2M1D7M1I11M1I10M1I15M1I10M1I13M1I12M1I12M1I7M1D4M1I13M2386S * 0 0 * *   NM:i:86 ms:i:418  AS:i:418  nn:i:0  tp:A:S  cm:i:6  s1:i:77 dv:f:0.0589</span><br><span class="line">m54174_171023_074758/9962478/22561_26454  2064  kraken:taxid|1323664|NZ_CP012748.1  2198443 1 1853H19M1I22M4I8M1I12M1I38M1I18M2I9M1I3M1I5M1I12M1I24M1I12M2I24M2I3M1I3M1D5M3I12M1I13M2I3M1I8M1I11M2I8M1D3M1I2M1I11M1I12M1I12M1I5M1D3M1D2M1I12M1I12M1I13M2I1M3I4M1I10M1D14M2I2M1I4M1D4M1I4M1I8M1I5M1D4M1I2M1I12M2I12M1I7M1D10M1523H * 0 0 seq * NM:i:95 ms:i:414  AS:i:414  nn:i:0  tp:A:P  cm:i:7  s1:i:60 s2:i:97 dv:f:0.0466 SA:Z:kraken:taxid|1323664|NZ_CP012748.1,2198444,-,903S469M57I2464S,1,95;kraken:taxid|1323664|NZ_CP012748.1,2198449,-,1461S464M47I1921S,15,95;kraken:taxid|1323664|NZ_CP012748.1,2198444,-,683S469M51I2690S,4,106;kraken:taxid|1323664|NZ_CP012748.1,2198449,-,2305S464M17I1107S,16,105;kraken:taxid|1323664|NZ_CP012748.1,2198456,-,2607S450M11I825S,22,93;kraken:taxid|1323664|NZ_CP012748.1,2198451,-,430S460M63I2940S,3,105;kraken:taxid|1323664|NZ_CP012748.1,2198449,-,13S464M94I3322S,11,133;</span><br></pre></td></tr></table></figure>
<p>We can clearly found that the same sequence had been mapped to different reference sequences, and it’s been reported several times, with the exactly same line content.</p>
<p>This is a limitation of <code>minimap2</code>, as reported: <a href="https://github.com/lh3/minimap2/issues/164" target="_blank" rel="noopener">Multiple empty hits?</a> and <a href="https://github.com/lh3/minimap2/issues/141" target="_blank" rel="noopener">How does using a multi part index affect the accuracy?</a>. Specifically, when a huge reference is used, <code>minimap2</code> will split it into multiple parts and align all queries against each part independently. For most parts, <code>minimap2</code> will print unmapped records.</p>
<p>The <strong>good news</strong> is that <a href="https://github.com/lh3/minimap2/releases/tag/v2.12" target="_blank" rel="noopener">Minimap2-2.12 (r827)</a> had addressed this bug.</p>
<h4 id="minimap2-arm"><a class="markdownIt-Anchor" href="#minimap2-arm"></a> minimap2-arm</h4>
<p><a href="https://github.com/hasindu2008/minimap2" target="_blank" rel="noopener">minimap2-arm</a> is a solution provieded by Hasindu Gamaarachchi, and it merges the results from a multi-part index to achieve a considerably similar output from a single-part index. So I cloned this modified <code>minimap2</code> to run the job anain.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> install</span></span><br><span class="line">git clone https://github.com/hasindu2008/minimap2 minimap2-arm &amp;&amp; cd minimap2-arm &amp;&amp; git checkout multipart-merge-tmp &amp;&amp; make</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> run</span></span><br><span class="line"><span class="meta">$</span><span class="bash">TOOLDIR/minimap2-arm/minimap2 -x map-pb -I 500G -t <span class="variable">$PPN</span> -a -Q --multi-prefix tmp <span class="variable">$contaminants2</span> third_all.fasta &gt; minimap2.sam</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> --multi-prefix: <span class="built_in">enable</span> mergine</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> -I: split index <span class="keyword">for</span> every ~500G input bases, this number is far more than the reference.</span></span><br></pre></td></tr></table></figure>
<p>I counted reads in each categories like I did in <code>MashMap</code> part:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">No. of total reads: 6633142</span><br><span class="line">  No. of reads in the SAM: 6633142</span><br><span class="line">    No. of mapped: 661646</span><br><span class="line">      No. of good: 490150</span><br><span class="line">      No. of bad: 125390</span><br><span class="line">      No. of ambivalent: 46106</span><br><span class="line">    No. of Unmapped: 5971496</span><br><span class="line">  No. of reads not in the SAM: 0</span><br></pre></td></tr></table></figure>
<p>Then I kept the ‘unmapped’, ‘good’ and ‘ambivalent’ reads for downstream analysis.</p>
<p>To parse the result and get the clean sequences, I used a simple python script to extract clean fasta and bad fasta ids. (appendix 1)</p>
<p>To validate the effciency, I checked several sequences by <code>blastn</code> manually.</p>
<p>Belows are two examples.</p>
<p>This is the alignment of ‘m161109_080520_42256_c101052872550000001823247601061737_s1_p0/28/0_4873’ by <code>minimap2</code>.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">$</span><span class="bash"> grep <span class="string">'m161109_080520_42256_c101052872550000001823247601061737_s1_p0/28/0_4873'</span> minimap2.sam</span></span><br><span class="line">m161109_080520_42256_c101052872550000001823247601061737_s1_p0/28/0_4873 16  kraken:taxid|66084|NC_012416.1  1207118 5 2514S9M1D7M1D9M1I8M1D15M1I6M1I7M1D5M1D15M1D8M1I4M1D11M1D12M1I17M1D5M1D4M1D5M1I3M2D18M1I1M1I33M1I13M2I13M1D6M1I29M1I4M1D26M1D8M1I27M1D3M1D12M1I4M1I8M1I8M1I12M1D3M1D8M1I16M1I12M1I11M1I2M2D6M2D3M1D31M1D9M2I33M1D16M1I15M1D11M2D9M1D25M2I5M1D3M2I11M1D3M1D14M1I8M1D4M1D19M4I29M1D11M1D8M1D2M1D5M1I2M4I9M1D15M1I19M2I7M1I6M1I28M1D3M1D3M1I16M1D2M1D3M1I6M1I14M2I12M1D28M1D3M1D6M2I5M1I32M1D8M1D25M1D8M1D30M1I11M1I1M1I16M1D14M1D17M1D19M1I32M1D14M1D6M1D12M1I50M1D26M1I8M1D3M1D12M1D13M1I8M1D18M1D12M1I24M1D4M2D6M3I10M1D3M2D18M1D3M1I5M1D5M1I3M2D15M1I5M1I7M1I9M1D2M3I31M1I13M1D6M1I22M2D2I3M1I6M1I10M1I5M1D6M2D10M1I16M1I13M1I3M1I13M1I3M3I2M1I11M1I7M1D4M2I8M1D9M1I17M1I3M1I7M2I3M1I20M1D2M3I5M2I5M1I13M5I3M2I6M1I17M2I3M1I5M1D10M2I11M1D13M1D2M1I8M1D9M1D8M1I3M1I12M1I3M1D3M1D5M1I6M1I8M1I5M1I5M1I10M1I4M1I8M1I3M1I7M1I5M1D8M1D8M1I6M1D10M2I9M1I1M1I9M1D10M1D3M1D4M1D6M1D7M2D5M2I5M2I2M2D2M1I16M1I5M14I4M2D3M1I11M1D6M1I1M1I9M1I14M2I6M2I14M1D2M1D6M1D9M2D1M1D22M1D12M1D7M1I13M2D8M1I3M1I15M1I18M1I17M9S  * 0 0 seq * NM:i:424  ms:i:2078 AS:i:2078 nn:i:0  tp:A:P  cm:i:28 s1:i:477  s2:i:474  dv:f:0.0866 SA:Z:kraken:taxid|66084|NC_012416.1,1360016,-,616S1173M32I3052S,9,195;</span><br><span class="line">m161109_080520_42256_c101052872550000001823247601061737_s1_p0/28/0_4873 272 kraken:taxid|1236909|NC_021089.1  1067958 0 2514S9M1D7M1D9M1I8M1D15M1I6M1I7M1D5M1D15M1D8M1I4M1D11M1D12M1I17M1D5M1D4M1D8M7D4M3D15M1I1M1I33M1I13M2I13M1D6M1I29M1I4M1D26M1D8M1I27M1D3M1D12M1I4M1I8M1I8M1I12M1D3M1D8M1I16M1I12M1I11M1I2M2D6M2D3M1D31M1D9M2I33M1D16M1I15M1D11M2D9M1D25M2I5M1D3M2I11M1D3M1D14M1I8M1D4M1D19M4I29M1D11M1D8M1D2M1D5M1I2M4I9M1D15M1I19M2I7M1I6M1I28M1D3M1D3M1I16M1D2M1D3M1I6M1I18M1I3M1I5M1D28M1D3M1D6M2I5M1I32M1D3M1D30M1D8M1D30M1I11M1I1M1I16M1D14M1D22M1D14M1I32M1D14M1D6M1D13M1I49M1D26M1I8M1D3M1D10M1D15M1I8M1D18M1D12M1I25M1D3M2D6M3I10M1D3M2D17M1D4M1I5M1D5M1I3M2D15M1I5M1I7M1I9M1D2M3I35M1I9M1D6M1I22M1I11M1I10M1I5M1D6M2D10M1I16M1I13M1I3M1I13M1I3M3I2M1I11M1I7M1D4M2I8M1D9M1I17M1I3M1I7M2I3M1I20M1D2M3I5M2I5M1I13M5I3M2I6M1I17M2I3M1I5M1D10M2I11M1D13M1D2M1I8M1D9M1D8M1I3M1I12M1I4M1D2M1D5M1I6M1I8M1I5M1I5M1I10M1I4M1I8M1I3M1I7M1I5M1D8M1D8M1I6M1D10M2I9M1I1M1I9M1D13M2D4M1D6M1D7M2D5M2I11M1I16M1I5M14I4M2D4M1I10M1D6M1I1M1I9M1I14M2I6M2I14M1D2M1D6M1D8M2D2M1D22M1D12M1D7M1I12M1D1M1D8M1D8M1I3M1I1M1I5M1I18M1I17M9S * 0 0 *   NM:i:435  ms:i:2028 AS:i:2028 nn:i:0  tp:A:S  cm:i:28 s1:i:474  dv:f:0.0866</span><br><span class="line">m161109_080520_42256_c101052872550000001823247601061737_s1_p0/28/0_4873 272 kraken:taxid|225364|NZ_LK055284.1 991010  0 2514S9M1D7M1D9M1I8M1D15M1I6M1I7M1D5M1D15M1D8M1I4M1D11M1D12M1I17M1D5M1D4M1D8M7D4M3D15M1I1M1I33M1I13M2I13M1D6M1I29M1I4M1D26M1D8M1I27M1D3M1D12M1I4M1I8M1I8M1I12M1D3M1D8M1I16M1I12M1I3M1I3M1D4M1I2M2D6M3D34M1D9M2I33M1D16M1I15M1D11M2D9M1D25M2I6M1D2M2I11M1D3M1D14M1I8M1D4M1D19M4I29M1D11M1D8M1D2M1D5M1I2M4I9M1D15M1I19M2I7M1I6M1I24M1D7M1D4M1I15M1D2M1D3M1I6M1I18M1I3M1I5M1D28M1D3M1D6M2I5M1I32M1D3M1D30M1D8M1D30M1I11M1I1M1I16M1D14M1D22M1D14M1I32M1D12M1D8M1D13M1I49M1D26M1I8M1D3M1D10M1D15M1I8M1D18M1D12M1I25M1D3M2D6M3I10M1D3M2D17M1D4M1I5M1D5M1I3M2D15M1I5M1I7M1I9M1D2M3I31M1I13M1D6M1I22M2D2I3M1I6M1I10M1I5M1D6M2D10M1I16M1I13M1I3M1I13M1I3M3I2M1I11M1I7M1D4M2I8M1D10M1I16M1I3M1I7M2I3M1I20M1D2M3I5M2I5M1I13M5I3M2I6M1I17M2I3M1I5M1D10M2I11M1D13M1D2M1I8M1D9M1D8M1I3M1I12M1I3M1D3M1D5M1I6M1I8M1I5M1I5M1I10M1I4M1I8M1I3M1I7M1I5M1D8M1D8M1I6M1D10M2I9M1I1M1I9M1D10M1D3M1D4M1D6M1D7M2D5M2I5M2I2M2D2M1I16M1I5M14I4M2D4M1I10M1D6M1I1M1I3M1I14M2I3M1I4M1D4M2I14M1D2M1D6M1D8M2D2M1D22M1D12M1D7M1I13M2D8M1I3M1I15M1I18M1I17M9S *   NM:i:444  ms:i:1984 AS:i:1984 nn:i:0  tp:A:S  cm:i:24 s1:i:437  dv:f:0.0923</span><br><span class="line">m161109_080520_42256_c101052872550000001823247601061737_s1_p0/28/0_4873 272 kraken:taxid|163164|NC_002978.6 1039834 0 2514S9M1D7M1D9M1I8M1D15M1I6M1I7M1D5M1D15M1D8M1I4M1D11M1D12M1I17M1D5M1D4M1D8M7D4M3D15M1I1M1I33M1I13M2I13M1D6M1I29M1I4M1D26M1D8M1I27M1D3M1D12M1I4M1I8M1I8M1I12M1D3M1D8M1I16M1I12M1I3M1I3M1D4M1I2M2D6M3D34M1D9M2I33M1D16M1I15M1D11M2D9M1D25M2I6M1D2M2I11M1D3M1D14M1I8M1D4M1D19M4I29M1D11M1D8M1D2M1D5M1I2M4I9M1D15M1I19M2I7M1I6M1I24M1D7M1D4M1I15M1D2M1D3M1I6M1I18M1I3M1I5M1D28M1D3M1D6M2I5M1I32M1D3M1D30M1D8M1D30M1I11M1I1M1I16M1D14M1D22M1D14M1I32M1D12M1D8M1D13M1I49M1D26M1I8M1D3M1D10M1D15M1I8M1D18M1D12M1I25M1D3M2D6M3I10M1D3M2D17M1D4M1I5M1D5M1I3M2D15M1I5M1I7M1I9M1D2M3I31M1I13M1D6M1I22M2D2I3M1I6M1I10M1I5M1D6M2D10M1I16M1I13M1I3M1I13M1I3M3I2M1I11M1I7M1D4M2I8M1D10M1I16M1I3M1I7M2I3M1I20M1D2M3I5M2I5M1I13M5I3M2I6M1I17M2I3M1I5M1D10M2I11M1D13M1D2M1I8M1D9M1D8M1I3M1I12M1I3M1D3M1D5M1I6M1I8M1I5M1I5M1I10M1I4M1I8M1I3M1I7M1I5M1D8M1D8M1I6M1D10M2I9M1I1M1I9M1D10M1D3M1D4M1D6M1D7M2D5M2I5M2I2M2D2M1I16M1I5M14I4M2D4M1I10M1D6M1I1M1I3M1I14M2I3M1I4M1D4M2I14M1D2M1D6M1D8M2D2M1D22M1D12M1D7M1I13M2D8M1I3M1I15M1I18M1I17M9S * 0   NM:i:444  ms:i:1984 AS:i:1984 nn:i:0  tp:A:S  cm:i:25 s1:i:452  dv:f:0.0908</span><br><span class="line">m161109_080520_42256_c101052872550000001823247601061737_s1_p0/28/0_4873 272 kraken:taxid|1633785|NZ_CP011148.1  1039878 0 2514S9M1D7M1D9M1I8M1D15M1I6M1I7M1D5M1D15M1D8M1I4M1D11M1D12M1I17M1D5M1D4M1D8M7D4M3D15M1I1M1I33M1I13M2I13M1D6M1I29M1I4M1D26M1D8M1I27M1D3M1D12M1I4M1I8M1I8M1I12M1D3M1D8M1I16M1I12M1I3M1I3M1D4M1I2M2D6M3D34M1D9M2I33M1D16M1I15M1D11M2D9M1D25M2I6M1D2M2I11M1D3M1D14M1I8M1D4M1D19M4I29M1D11M1D8M1D2M1D5M1I2M4I9M1D15M1I19M2I7M1I6M1I24M1D7M1D4M1I15M1D2M1D3M1I6M1I18M1I3M1I5M1D28M1D3M1D6M2I5M1I32M1D3M1D30M1D8M1D30M1I11M1I1M1I16M1D14M1D22M1D14M1I32M1D12M1D8M1D13M1I49M1D26M1I8M1D3M1D10M1D15M1I8M1D18M1D12M1I25M1D3M2D6M3I10M1D3M2D17M1D4M1I5M1D5M1I3M2D15M1I5M1I7M1I9M1D2M3I31M1I13M1D6M1I22M2D2I3M1I6M1I10M1I5M1D6M2D10M1I16M1I13M1I3M1I13M1I3M3I2M1I11M1I7M1D4M2I8M1D10M1I16M1I3M1I7M2I3M1I20M1D2M3I5M2I5M1I13M5I3M2I6M1I17M2I3M1I5M1D10M2I11M1D13M1D2M1I8M1D9M1D8M1I3M1I12M1I3M1D3M1D5M1I6M1I8M1I5M1I5M1I10M1I4M1I8M1I3M1I7M1I5M1D8M1D8M1I6M1D10M2I9M1I1M1I9M1D10M1D3M1D4M1D6M1D7M2D5M2I5M2I2M2D2M1I16M1I5M14I4M2D4M1I10M1D6M1I1M1I3M1I14M2I3M1I4M1D4M2I14M1D2M1D6M1D8M2D2M1D22M1D12M1D7M1I13M2D8M1I3M1I15M1I18M1I17M9S *   NM:i:446  ms:i:1972 AS:i:1972 nn:i:0  tp:A:S  cm:i:25 s1:i:452  dv:f:0.0908</span><br><span class="line">m161109_080520_42256_c101052872550000001823247601061737_s1_p0/28/0_4873 2064  kraken:taxid|66084|NC_012416.1  1360016 9 616H6M1I3M1I9M1D12M1I3M1I4M1I1M1I6M3I5M1D10M1D2M1D4M1D4M1I28M2I7M4I3M2I10M1I1M1I4M1D4M6I7M1I6M1I9M1I2M1I13M1I1M1I8M1D4M1I8M1I15M1D3M1I15M1D5M1I1M3D4M1I10M1I8M2I7M1I9M1D4M1I7M1I12M2D11M1I27M1D21M1D5M2I23M5I3M1I9M1I6M3I14M1I6M1D3M1D43M1I2M1D4M1D8M1D21M2I16M1I7M1D5M1D5M1I11M1I21M1D6M1D2M1D7M2I3M1I15M3I2M1I1M2I6M2I4M1I6M1D9M1D62M1I10M1D3M1I11M2D7M1D13M1I4M1I10M1D4M1I8M1D15M1D12M1D10M1D10M1D10M1I12M1D2M2D1M1D14M1I8M1D6M1D11M2D15M1I13M1I14M1I8M1I32M1D4M1I18M1D30M2D5M1D10M1D5M1D11M1D13M1D6M1I7M1I4M1D13M1I1M2D18M2D8M1I8M1I3M1I7M3052H * 0 0 seq * NM:i:195  ms:i:1194 AS:i:1194 nn:i:0  tp:A:P  cm:i:21 s1:i:287  s2:i:284  dv:f:0.0502 SA:Z:kraken:taxid|66084|NC_012416.1,1207118,-,2514S2299M51I9S,5,424;</span><br><span class="line">m161109_080520_42256_c101052872550000001823247601061737_s1_p0/28/0_4873 272 kraken:taxid|1633785|NZ_CP011148.1  1051218 0 616S6M1I3M1I9M1D12M1I3M1I4M1I1M1I6M3I5M1D10M1D2M1D4M1D4M1I28M2I7M4I3M2I10M1I1M1I4M1D4M6I7M1I6M1I9M1I2M1I13M1I1M1I8M1D4M1I8M1I15M1D3M1I15M1D7M2D4M1I10M1I8M2I7M1I9M1D4M1I7M1I12M2D11M1I27M1D29M1I21M5I8M1I4M1I6M3I14M1I6M1D3M1D43M1I2M1D4M1D8M1D21M2I16M1I7M1D5M1D5M1I11M1I21M1D6M1D2M1D7M2I3M1I15M3I2M1I2M4I9M1I6M1D9M1D62M1I10M1D3M1I11M2D5M1D15M1I4M1I10M1D4M1I8M1D15M1D12M1D10M1D10M1D10M1I12M1D2M2D1M1D14M1I8M1D6M1D11M2D15M1I13M1I14M1I8M1I32M1D4M1I18M1D30M2D5M1D10M1D5M1D11M1D13M1D6M1I7M1I4M1D13M1I1M2D18M2D8M1I8M1I3M1I7M3052S * 0 0 * * NM:i:200  ms:i:1164 AS:i:1164 nn:i:0  tp:A:S  cm:i:22 s1:i:284  dv:f:0.0655</span><br></pre></td></tr></table></figure>
<p>This is the results of <code>blastn</code>, and I marked them with taxonomy ID. The results agreed well with that of <code>minimap2</code>.</p>
<img src="/blog/2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/blast_1532401032_2561.png">
<p>Here is another example.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">$</span><span class="bash"> grep <span class="string">'m161109_080520_42256_c101052872550000001823247601061737_s1_p0/796/0_3902'</span> minimap2.sam</span></span><br><span class="line">m161109_080520_42256_c101052872550000001823247601061737_s1_p0/796/0_3902  16  kraken:taxid|573570|NZ_CP016796.1 218142  29  390S6M1D15M1D10M3I6M1D15M1I3M2I11M3I4M1I14M3I4M1I9M1D7M1D3M2I9M3I20M2I6M1D15M1I3M2I17M3I9M2I4M1I5M1D11M3I9M3I4M4I5M1I10M1I12M1I6M3D10M2D5M2D12M2I11M3177S * 0 0 seq * NM:i:85 ms:i:178  AS:i:178  nn:i:0  tp:A:P  cm:i:5  s1:i:93 s2:i:52 dv:f:0.0291 SA:Z:kraken:taxid|573570|NZ_CP016796.1,218159,-,296S287M9I3310S,22,82;</span><br><span class="line">m161109_080520_42256_c101052872550000001823247601061737_s1_p0/796/0_3902  2064  kraken:taxid|573570|NZ_CP016796.1 218159  22  296H9M3I6M4I7M1D11M3I3M1I3M1D11M3I8M1I10M3I6M1D11M3I13M1I5M3I6M1D15M1I3M2I11M4D4M5D7M5D11M3I4M1I9M1D7M1D3M2I6M3I23M2I6M1D17M5D4M2D5M2D13M3310H  * 0 0 seq * NM:i:82 ms:i:178  AS:i:178  nn:i:0  tp:A:P  cm:i:7  s1:i:166  s2:i:120  dv:f:0.0235 SA:Z:kraken:taxid|573570|NZ_CP016796.1,218142,-,390S304M31I3177S,29,85;</span><br></pre></td></tr></table></figure>
<p>And, I got different results using <code>blastn</code>.</p>
<img src="/blog/2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/1532401743_26861.png">
<p>And the detailed alignment of the top 1.</p>
<img src="/blog/2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/1532401866_17878.png">
<p>But when I checked the raw sequence, there are lots of repeats. I don’t know what it is.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;</span><span class="bash">m161109_080520_42256_c101052872550000001823247601061737_s1_p0/796/0_3902 RQ=0.860</span></span><br><span class="line">CAGAAGATGAGATTTAAATTGCATCCCAATTGTCATATATAATTAATCATATAAATTATATATGTGTAAG</span><br><span class="line">TTTTTTTTTTTTGTGAACATTATAGAAATAATAATAATATTCCATGGCCCTGACAAAGGGCCGAGTATTT</span><br><span class="line">ACTCAATTTGGACAAAGAATAAAAAATTTAAGCCTTCCTTTCAGTTTCTTGGGCAACAGAACATATTTGG</span><br><span class="line">TAAAACACCACCTTGTGCAATCGTACACCAAACAGAAGTTTAATTTATTCTTCGTCCACCATCATACGAA</span><br><span class="line">TTGCAATTGCAAATGTCTATGGAATAATACGAGTTTTCTTGTTGTCCACAACTCTTATGAGCAGCATTAC</span><br><span class="line">CGCGCCAATTCTAGTACTATCAAGCAGCTAAATATTACCCATGACAGCAGCCAAATCATAAACTGGTGCA</span><br><span class="line">CCAGCTCCAACACGTTCAGCATAATTTCCTTTGAAAAAATTTAAAAATTTTGAAAAATTCAAAATTTTGA</span><br><span class="line">AAAATTCAAAAATTTTGAAAAATTACAACAAATTTGAAAAAAATTCAAAAATTTGAAAAAATTCAATAAA</span><br><span class="line">ATTTTGAAAAAATTCAAAAAAATTTTAAAAAATCTCAAAAATTTTGAAAATTCATAAATTTTGAAAAAAT</span><br><span class="line">TCAAAACCTTTTGAAAACAAAATCAAAATTTGAAAAAAATTCAAAAATTTTGAAAAATTCAAAAATTTGA</span><br><span class="line">AAAACCATTCAAAAATTTTGAAAAAAAATTCAAAAATTTTGAAAAAATTCAAAAATTTTGAAATAAAAAA</span><br><span class="line">TTTCAAAAATTTTGAAAAAAATTCAAAAATTTTGAAAAATCTCAAAAAATTTTGAAAAAGTTCAATGAAA</span><br><span class="line">TTTTGAAAAAGTTTAAAAGATTTTGAAAAAAATACCAAAAATTTTGAAAAAATCAAGAATTTTGAAAAAA</span><br><span class="line">ATTCCAAAATTTTGAAAAAATTCAATTAAATTTTGAAAAATTCAAAATCTTTGAAAAAAATTCAACAAAT</span><br><span class="line">TTTGAAAAAATTCACAAAAATTTTGAAAAAATTCAAAAATTTTGAAAAATTCAAAAATTTTTGAAAAAAC</span><br><span class="line">AAACTCAAAATTTTGAAAAAATTCAAAAATTTGAAAAAATTCAAAAAATTTGAAAAAATTCAAAAACTTT</span><br><span class="line">TGAAAAAAATTCAAAAATCTTTGAAAAAATTCAAAAATTTTTGAAAAAATTCAAAAATTTTGAAAAAAAT</span><br><span class="line">TCAAAAATTTTGAAAAAATTCAAAAATTTTTGAAAAAATCAAAAATTTGAAAAAATTCCAAAAATTTTTG</span><br><span class="line">AAAAAATTCAAAAAATTTTCGAAAAATTCAATTTTGAAAAATTCAAAAATTTACTGAAAAAATTTCAAAA</span><br><span class="line">ATTTTGAAAAAATTCAATAAATTTTGAAAAAATTCAAACATTTTGAAAAATTCAAAAATTTTGAAAAAAT</span><br><span class="line">CAACAAATTTTTGAAAAAATTTCAAAAATTTTTGAAAAATCAAAAAATTTTTTGAAAAAAATTCACGCCA</span><br><span class="line">AATTTTGAAAAAATTTAAAAATTTTGAAAAAATTAAATAAATTTTAAAATTTGTATGATTTTTCAAATTT</span><br><span class="line">TTGATAAGTTTTCATTTTGAAAAATTTAAATTTTTTCAAAATTTTTTGAATTTTTTCAAAATTTTGAATT</span><br><span class="line">TTTTCACCCAATTTTTGAATTTTCTTTCAAAAAATTTTTAGAAATTAACAAATAACCTATTTCAAATTTT</span><br><span class="line">TGAATTTTTCAAATTTTTGATTTTTTCAAAATTTTGAATTTTTTTTCAAAATTTTGAATTTTTAAATACA</span><br><span class="line">AATTTTTGAATTTTTCAAAATTGTTTGAATTTTTATCAAAATTTTTGAATTTTTTTAAAAATTTTTGATT</span><br><span class="line">TTTTCAAAGGAAATTATTGCTGAAGCCGTGTTGGAGCTGTGCACCAGTTTTATTTGGCTGCCTGTTCATG</span><br><span class="line">GAACTTTTAGCTGCTGAGTACTAAGAATTGGCGGGTAATGCTGTCAGTGACAGAACAGAAAACTCGTATT</span><br><span class="line">ATTTCCTAGACATTTGCAATTGGCAATTCAGTAAATGACGAACCAATTAAATAATACTTCTGTCTGGTGT</span><br><span class="line">TTACGAATGCACAAGGTGAGTGTTTTACCAAATAACAAGCTGTTCTGATGCCTCAACGAAAACTGAAAGA</span><br><span class="line">AGGCTTAAATTTTTTTATTCCTTTGTCCAAATTAGTAAACTACTTCGGCCCTTTTCAGGGCCATAATATT</span><br><span class="line">CATTATCTATTTTCACTCAGATGTTCGCAAAAAAAAAACTTACACATATAATAATTTTATATATTAATTT</span><br><span class="line">GCAAATTTGGATGCAATTTAAAATTGAATGATAAAGTGCAATGGTGTTGTCTAGTCTACAAAAATTCTAT</span><br><span class="line">AAACGTACACAAAAAAAATATTCGCTAAATTGAATTGTTGATAAAAAAAATATTTTTACTTAGAAAATCT</span><br><span class="line">TAAAAAAAAAAAACACAATAGATATGATTGTATATAATCAAAAAATGCTATTGAATGTAAATAATTTTTT</span><br><span class="line">AACAATTTCAAATTTTTAAAAACGTTCTCGCTTAGTTATATCAAATTATTCGGATTTTGGTTTTTTTATT</span><br><span class="line">ATTAATTATTATTATATTAATAATAAATTATTATTCTACTCAATTATTAAAATTTGAATAATTTTGTGGC</span><br><span class="line">CTCAAAGGGCCATTTGTTTTATAATGACAATTTTATTGAAGGAATACCAAAAGAATTGAAATAAACATAC</span><br><span class="line">GATTGATATATTAAAGTATTAGCGTGTTTTATTTCTTTCTTTTAGGTGAAAATGCTGCCTTCCTGGGCTT</span><br><span class="line">TAGGTGAAGTATGGATGTTTTTGGCCAGTTTTATGGTTTATTTGGTGCCTTTGGTTTTTAGTTTGGACCT</span><br><span class="line">TTTAGCTGATTTTTTGGCCTTTAATCGGTGAATTTTGCAGTTTGTAGTCGTACGAATGTAGTTTTAGGCA</span><br><span class="line">GCAGATGGTGGTTTTTTCTTCTCGTCTTTTTTTCGGTTTAGCAGCTTTAATTGGTGATTTTTTTTGCTTT</span><br><span class="line">GATTGATTTCTTACAGCCGGTTTTATGTTTTAATTGTCCAGTTTGCTAAAGTAGAATTCGTTTTACAGTT</span><br><span class="line">GCAAGCTTTTTTGGGTTTTGTACCAGAAGCCCTTCTTTCTTTTTCTTTTAGAACTTTTTTTTCTTAGAAC</span><br><span class="line">TTTTTTTTTTTAGAACATTTTTTTTTAGGAACTTTTTTTTTAGAACTTTTTTTAGAAACTTTTGTTATCT</span><br><span class="line">TTTCTGTTTGTTCTTGTTATCTTTTTGTTTGTTTTGTTATCATTTTTTTGTTTCTTTTGGTTCTCTTTTT</span><br><span class="line">GTTTGTTTTTGTTATCATTTTTTGTTTGTTTTTGGGTATCTCTTTTTTTGTTTGTTTTGTTATCTTTTTG</span><br><span class="line">TTGTTTGGTTTATCTTTTCCTTTGTTTTGTTTATTTTTTTGTTTGTTTTGGTTATATTTTTGTTTGTTTG</span><br><span class="line">TTATCTTTTTTGTTTTTTTTGTTATCTTTTTGTTTGTTTGTATCTTTTTTTGTTTGTTTTGTTATCTTTT</span><br><span class="line">TTTGTTTGTTTTGTTATCTTTTTGTTTGTTTTTGTTATCTTTTTGTTTGTTTTGTTTATCTTTATTGTTT</span><br><span class="line">GTTTTGTTATCTTTTTGTTTGTTTTGTTATCTTCTTTTGTTTGTTTTGTTATCTTTTTGCTTTGTTTTGT</span><br><span class="line">TATCTTTTTGTTTTGTTTTGTTATCTTTTTGTTGTTCTTTATTCTGGTTAATCATTTTTTTGGTTGTTTT</span><br><span class="line">GTTCATCTTTTTGTGTTTGTTTTTGTTATCTTTTTGTTTGTTTTGTTAATCTTTTTGTTTGCTTTGTTAT</span><br><span class="line">CATTTTTTGTTTGCTTTGTTATTTTTTGTTTGCTTTGTTATATTTTTGTTGCTTTGTTATCATTTTTGTT</span><br><span class="line">TGCTTTGGTTATCTTTGTTTGCTTTGTTATCTTTTTTTGTTGGCTTTGTTTATCGTTTTTGTATTTGCTT</span><br><span class="line">TGTTATCGTTTTTGTTTGCTTTGTTATCTTTTTTGTTTTGCGTTTTTTTAGC</span><br></pre></td></tr></table></figure>
<p>The differences between <code>minimap2</code> and <code>blastn</code> are explainable, they used different algorithm and different databases. And generally, I think <code>minimap2</code> is reliable.</p>
<h2 id="in-summary"><a class="markdownIt-Anchor" href="#in-summary"></a> In summary</h2>
<p>In summary, I removed contaminants in reads by the following steps:</p>
<ul>
<li>Prepare contamination library (bacteria, viral, fungi, protozoa, and archaea from Refseq and mtDNAs).</li>
<li>Use <code>BBDuk</code> to remove contaminants from illumina short reads.</li>
<li>Use <code>minimap2</code> to remove contaminants from PacBio long reads.</li>
</ul>
<p>But there are some concerns existing:</p>
<ul>
<li>Some real sequences may also be removed along with the contaminants.</li>
<li>Many repeat sequences were removed, and I don’t know where they were from.</li>
<li>For the ‘ambivalent’ reads, I kept them for downstream analysis, but I didn’t know whether they should be removed, say, throw away reads which the primary aligment were contaminant.</li>
</ul>
<p>There are some other long-reads mapper, and people also try to tune the parameters of short-read aligners to work with long-reads. There are some threads/posts talking about this:</p>
<ul>
<li><a href="https://www.biostars.org/p/116696/" target="_blank" rel="noopener">Question: Long read alignment</a></li>
<li><a href="https://www.biostars.org/p/130978/" target="_blank" rel="noopener">Question: Long read-alignment + variant calling</a></li>
<li><a href="https://www.biostars.org/p/231120/" target="_blank" rel="noopener">Question: Alternative to BLASR ?</a></li>
<li><a href="https://www.biostars.org/p/231559/" target="_blank" rel="noopener">Question: BWA-MEM using long PacBio reads</a></li>
<li><a href="https://davetang.org/muse/2012/10/12/mapping-long-reads-with-bowtie/" target="_blank" rel="noopener">Mapping long reads with Bowtie</a></li>
<li><a href="https://groups.google.com/forum/#!searchin/rna-star/starlong/rna-star/-2mBTPWRCJY/PKSPCNcade8J" target="_blank" rel="noopener">STAR: segmentation fault when using long reads</a></li>
</ul>
<p>I don’t really understand the “mapping” things now, but I expect that there will be several dominant tools for long-reads mapping, just as short-reads mapping.</p>
<h2 id="useful-links"><a class="markdownIt-Anchor" href="#useful-links"></a> Useful links</h2>
<ul>
<li><a href="https://www.biostars.org/p/244912/" target="_blank" rel="noopener">Question: Removing contaminations from PacBio reads</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/genome/doc/ftpfaq/#allcomplete" target="_blank" rel="noopener">How can I download RefSeq data for all complete bacterial genomes?</a></li>
<li><a href="http://seqanswers.com/forums/showthread.php?t=41288" target="_blank" rel="noopener">Introducing BBSplit: Read Binning Tool for Metagenomes and Contaminated Libraries</a></li>
<li><a href="http://seqanswers.com/forums/showthread.php?t=58221" target="_blank" rel="noopener">Yes … BBMap can do that!</a></li>
<li><a href="https://github.com/HRGV/DGHM2017_assembly/wiki/From-raw-reads-to-assembly---STEP-by-STEP" target="_blank" rel="noopener">From raw reads to assembly STEP by STEP</a></li>
<li><a href="http://seqanswers.com/forums/showthread.php?t=78397" target="_blank" rel="noopener"> Removing contamination with BBDUK</a></li>
<li><a href="https://www.biostars.org/p/143019/" target="_blank" rel="noopener">Question: Tool to separate human and mouse rna seq reads</a></li>
</ul>
<h2 id="change-log"><a class="markdownIt-Anchor" href="#change-log"></a> Change log</h2>
<ul>
<li>20180628: create the note.</li>
<li>20180725: complete the note.</li>
<li>20180807: update the ‘minimap-arm’ part, add the results of <code>kraken</code>.</li>
<li>20180815: add the part of ‘Minimap2-2.12 (r827)’</li>
</ul>
<hr class="footnotes-sep">
<section class="footnotes">
<ol class="footnotes-list">
<li id="fn1" class="footnote-item"><p>Chaisson MJ, Tesler G. 2012. Mapping single molecule sequencing reads using basic local alignment with successive refinement (BLASR): application and theory. BMC Bioinformatics. 13:238. doi:10.1186/1471-2105-13-238. <a href="#fnref1" class="footnote-backref">↩︎</a></p>
</li>
<li id="fn2" class="footnote-item"><p>Jain C, Dilthey A, Koren S, Aluru S, Phillippy AM. 2018 Apr 30. A Fast Approximate Algorithm for Mapping Long Reads to Large Reference Databases. Journal of Computational Biology. doi:10.1089/cmb.2018.0036. [accessed 2018 Jul 2]. <a href="https://www.liebertpub.com/doi/10.1089/cmb.2018.0036" target="_blank" rel="noopener">https://www.liebertpub.com/doi/10.1089/cmb.2018.0036</a>. <a href="#fnref2" class="footnote-backref">↩︎</a></p>
</li>
<li id="fn3" class="footnote-item"><p>Li H. 2017 Aug 4. Minimap2: versatile pairwise alignment for nucleotide sequences. arXiv:170801492 [q-bio]. [accessed 2018 Jan 10]. <a href="http://arxiv.org/abs/1708.01492" target="_blank" rel="noopener">http://arxiv.org/abs/1708.01492</a>. <a href="#fnref3" class="footnote-backref">↩︎</a></p>
</li>
</ol>
</section>

      
    </div>

    

    
    
    

    

    

    
      <div>
        <ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>Post author:  </strong>Yiwei Niu</li>
  <li class="post-copyright-link">
    <strong>Post link: </strong>
    <a href="https://yiweiniu.github.io/blog/2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/" title="Remove Microbial Contamination in Reads">https://yiweiniu.github.io/blog/2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/</a>
  </li>
  <li class="post-copyright-license">
    <strong>Copyright Notice:  </strong>All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="external nofollow" target="_blank">CC BY-NC-SA 4.0</a> unless stating additionally.</li>
</ul>

      </div>
    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/blog/tags/bio-tools/" rel="tag"># bio-tools</a>
          
            <a href="/blog/tags/contamination/" rel="tag"># contamination</a>
          
            <a href="/blog/tags/reads/" rel="tag"># reads</a>
          
            <a href="/blog/tags/BBDuk/" rel="tag"># BBDuk</a>
          
            <a href="/blog/tags/MashMap/" rel="tag"># MashMap</a>
          
            <a href="/blog/tags/long-reads/" rel="tag"># long-reads</a>
          
            <a href="/blog/tags/long-reads-alignment/" rel="tag"># long-reads alignment</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/blog/2018/06/Genome-Assembly-Pipeline-wtdbg/" rel="next" title="Genome Assembly Pipeline: wtdbg">
                <i class="fa fa-chevron-left"></i> Genome Assembly Pipeline: wtdbg
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>


  </div>


          </div>
          

  
    <div class="comments" id="comments">
      <div id="disqus_thread">
        <noscript>
          Please enable JavaScript to view the
          <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a>
        </noscript>
      </div>
    </div>

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            Overview
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image"
                src="/blog/images/portrait.jpg"
                alt="Yiwei Niu" />
            
              <p class="site-author-name" itemprop="name">Yiwei Niu</p>
              <p class="site-description motion-element" itemprop="description">a beginner, a learner</p>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/blog/archives">
                
                    <span class="site-state-item-count">19</span>
                    <span class="site-state-item-name">posts</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-categories">
                  <a href="/blog/categories/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">15</span>
                    <span class="site-state-item-name">categories</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-tags">
                  <a href="/blog/tags/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">29</span>
                    <span class="site-state-item-name">tags</span>
                  </a>
                </div>
              
            </nav>
          

          
            <div class="feed-link motion-element">
              <a href="/blog/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          
            <div class="links-of-author motion-element">
              
                <span class="links-of-author-item">
                  <a href="mailto:xiaohuwangwang@qq.com" target="_blank" title="E-Mail"><i class="fa fa-fw fa-envelope"></i>E-Mail</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="https://github.com/YiweiNiu" target="_blank" title="GitHub"><i class="fa fa-fw fa-github"></i>GitHub</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="http://weibo.com/u/3049858680" target="_blank" title="Weibo"><i class="fa fa-fw fa-weibo"></i>Weibo</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="https://twitter.com/ywniu" target="_blank" title="Twitter"><i class="fa fa-fw fa-twitter"></i>Twitter</a>
                  
                </span>
              
            </div>
          

          
          

          
          
            <div class="links-of-blogroll motion-element links-of-blogroll-block">
              <div class="links-of-blogroll-title">
                <i class="fa  fa-fw fa-link"></i>
                Links
              </div>
              <ul class="links-of-blogroll-list">
                
                  <li class="links-of-blogroll-item">
                    <a href="https://kaigedong.github.io/" title="Kaige Dong" target="_blank">Kaige Dong</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://homolog.us/blogs/" title="Homolog.us" target="_blank">Homolog.us</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="https://cgatoxford.wordpress.com/" title="CGAT" target="_blank">CGAT</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://www.gettinggeneticsdone.com/" title="Getting Genetics Done" target="_blank">Getting Genetics Done</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://www.acgt.me/" title="ACGT" target="_blank">ACGT</a>
                  </li>
                
              </ul>
            </div>
          

          
            
          
          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#prepare-the-contamination-library"><span class="nav-number">1.</span> <span class="nav-text"> Prepare the contamination library</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#bbduk-for-short-reads"><span class="nav-number">2.</span> <span class="nav-text"> BBDuk for short reads</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#long-reads"><span class="nav-number">3.</span> <span class="nav-text"> Long reads</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#mappacbio"><span class="nav-number">3.1.</span> <span class="nav-text"> mapPacBio</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#blasr"><span class="nav-number">3.2.</span> <span class="nav-text"> blasr</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#mashmap"><span class="nav-number">3.3.</span> <span class="nav-text"> MashMap</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#minimap2"><span class="nav-number">3.4.</span> <span class="nav-text"> minimap2</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#mimimap2"><span class="nav-number">3.4.1.</span> <span class="nav-text"> mimimap2</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#minimap2-arm"><span class="nav-number">3.4.2.</span> <span class="nav-text"> minimap2-arm</span></a></li></ol></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#in-summary"><span class="nav-number">4.</span> <span class="nav-text"> In summary</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#useful-links"><span class="nav-number">5.</span> <span class="nav-text"> Useful links</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#change-log"><span class="nav-number">6.</span> <span class="nav-text"> Change log</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; 2017 &mdash; <span itemprop="copyrightYear">2018</span>
  <span class="with-love" id="animate">
    <i class="fa fa-mars"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Yiwei Niu</span>

  

  
</div>




  <div class="powered-by">Powered by <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a></div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">Theme &mdash; <a class="theme-link" target="_blank" href="https://theme-next.org">NexT.Gemini</a></div>




        
<div class="busuanzi-count">
  <script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script>

  
    <span class="site-uv" title="Total Visitors">
      <i class="fa fa-user"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_uv"></span>
    </span>
  

  
    <span class="site-pv" title="Total Views">
      <i class="fa fa-eye"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_pv"></span>
    </span>
  
</div>









        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
          <span id="scrollpercent"><span>0</span>%</span>
        
      </div>
    

    
	
    

    
      <div>
        <!-- Go to www.addthis.com/dashboard to customize your tools -->
<div class="addthis_inline_share_toolbox">
  <script type = "text/javascript" src = "//s7.addthis.com/js/300/addthis_widget.js#pubid=ra-5a7ad07b7823d03b" async = "async" ></script>
</div>

      </div>
    
  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>


























  
  
    <script type="text/javascript" src="/blog/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  


  


  <script type="text/javascript" src="/blog/js/src/utils.js?v=6.3.0"></script>

  <script type="text/javascript" src="/blog/js/src/motion.js?v=6.3.0"></script>



  
  


  <script type="text/javascript" src="/blog/js/src/affix.js?v=6.3.0"></script>

  <script type="text/javascript" src="/blog/js/src/schemes/pisces.js?v=6.3.0"></script>



  
  <script type="text/javascript" src="/blog/js/src/scrollspy.js?v=6.3.0"></script>
<script type="text/javascript" src="/blog/js/src/post-details.js?v=6.3.0"></script>



  


  <script type="text/javascript" src="/blog/js/src/bootstrap.js?v=6.3.0"></script>



  

  
    <script id="dsq-count-scr" src="https://yiwei-nius-note.disqus.com/count.js" async></script>
  

  
    <script type="text/javascript">
      var disqus_config = function () {
        this.page.url = 'https://yiweiniu.github.io/blog/2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/';
        this.page.identifier = '2018/07/Remove-Contamination-of-Pokaryotic-Organisms-in-Reads/';
        this.page.title = 'Remove Microbial Contamination in Reads';
        };
      function loadComments () {
        var d = document, s = d.createElement('script');
        s.src = 'https://yiwei-nius-note.disqus.com/embed.js';
        s.setAttribute('data-timestamp', '' + +new Date());
        (d.head || d.body).appendChild(s);
      }
      
        loadComments();
      
    </script>
  












  





  

  

  

  

  
  

  

  

  

  

  
  <style>
    .copy-btn {
      display: inline-block;
      padding: 6px 12px;
      font-size: 13px;
      font-weight: 700;
      line-height: 20px;
      color: #333;
      white-space: nowrap;
      vertical-align: middle;
      cursor: pointer;
      background-color: #eee;
      background-image: linear-gradient(#fcfcfc, #eee);
      border: 1px solid #d5d5d5;
      border-radius: 3px;
      user-select: none;
      outline: 0;
    }

    .highlight-wrap .copy-btn {
      transition: opacity .3s ease-in-out;
      opacity: 0;
      padding: 2px 6px;
      position: absolute;
      right: 4px;
      top: 8px;
    }

    .highlight-wrap:hover .copy-btn,
    .highlight-wrap .copy-btn:focus {
      opacity: 1
    }

    .highlight-wrap {
      position: relative;
    }
  </style>
  <script>
    $('.highlight').each(function (i, e) {
      var $wrap = $('<div>').addClass('highlight-wrap')
      $(e).after($wrap)
      $wrap.append($('<button>').addClass('copy-btn').append('Copy').on('click', function (e) {
        var code = $(this).parent().find('.code').find('.line').map(function (i, e) {
          return $(e).text()
        }).toArray().join('\n')
        var ta = document.createElement('textarea')
        document.body.appendChild(ta)
        ta.style.position = 'absolute'
        ta.style.top = '0px'
        ta.style.left = '0px'
        ta.value = code
        ta.select()
        ta.focus()
        var result = document.execCommand('copy')
        document.body.removeChild(ta)
        
          if(result)$(this).text('Copied')
          else $(this).text('Copy failed')
        
        $(this).blur()
      })).on('mouseleave', function (e) {
        var $b = $(this).find('.copy-btn')
        setTimeout(function () {
          $b.text('Copy')
        }, 300)
      }).append(e)
    })
  </script>


</body>
</html>
