<!DOCTYPE html>



  


<html class="theme-next gemini use-motion" lang="_en">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2"/>
<meta name="theme-color" content="#222">












<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />



  <meta name="google-site-verification" content="3dnqtuxhGQK5nK7aqRtU2c3DzZlaVtBMN6M39UoGcgY" />




















<link href="/blog/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/blog/css/main.css?v=6.1.0" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/blog/images/apple-touch-icon-next.png?v=6.1.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/blog/images/favicon-32x32-next.png?v=6.1.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/blog/images/favicon-16x16-next.png?v=6.1.0">


  <link rel="mask-icon" href="/blog/images/logo.svg?v=6.1.0" color="#222">









<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/blog/',
    scheme: 'Gemini',
    version: '6.1.0',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":true,"onmobile":false},
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="I want to find out some circRNAs from RNA-seq data (total RNA-seq, not poly-A enriched). There are many tools for this mission. Here is a good review paper [1] talking about computational methods for">
<meta name="keywords" content="bio-tools,RNA-seq,circRNA">
<meta property="og:type" content="article">
<meta property="og:title" content="Identify circRNAs and Fusions from RNA-seq Using STARChip">
<meta property="og:url" content="https://yiweiniu.github.io/blog/2018/04/Identify-circRNAs-and-Fusions-from-RNA-seq-Using-STARChip/index.html">
<meta property="og:site_name" content="Yiwei Niu&#39;s Note">
<meta property="og:description" content="I want to find out some circRNAs from RNA-seq data (total RNA-seq, not poly-A enriched). There are many tools for this mission. Here is a good review paper [1] talking about computational methods for">
<meta property="og:locale" content="_en">
<meta property="og:updated_time" content="2018-04-23T09:02:24.205Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Identify circRNAs and Fusions from RNA-seq Using STARChip">
<meta name="twitter:description" content="I want to find out some circRNAs from RNA-seq data (total RNA-seq, not poly-A enriched). There are many tools for this mission. Here is a good review paper [1] talking about computational methods for">



  <link rel="alternate" href="/blog/atom.xml" title="Yiwei Niu's Note" type="application/atom+xml" />




  <link rel="canonical" href="https://yiweiniu.github.io/blog/2018/04/Identify-circRNAs-and-Fusions-from-RNA-seq-Using-STARChip/"/>



<script type="text/javascript" id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>Identify circRNAs and Fusions from RNA-seq Using STARChip | Yiwei Niu's Note</title>
  









  <noscript>
  <style type="text/css">
    .use-motion .motion-element,
    .use-motion .brand,
    .use-motion .menu-item,
    .sidebar-inner,
    .use-motion .post-block,
    .use-motion .pagination,
    .use-motion .comments,
    .use-motion .post-header,
    .use-motion .post-body,
    .use-motion .collection-title { opacity: initial; }

    .use-motion .logo,
    .use-motion .site-title,
    .use-motion .site-subtitle {
      opacity: initial;
      top: initial;
    }

    .use-motion {
      .logo-line-before i { left: initial; }
      .logo-line-after i { right: initial; }
    }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="_en">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"> 

<div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/blog/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Yiwei Niu's Note</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle">to share, to learn</p>
      
  </div>

  <div class="site-nav-toggle">
    <button aria-label="Toggle navigation bar">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        
          
  
  <li class="menu-item menu-item-home menu-item-active">
    <a href="/blog/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-home"></i> <br />Home</a>
</li>

      
        
        
          
  
  <li class="menu-item menu-item-categories menu-item-active">
    <a href="/blog/categories/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-th"></i> <br />Categories</a>
</li>

      
        
        
          
  
  <li class="menu-item menu-item-tags">
    <a href="/blog/tags/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />Tags</a>
</li>

      
        
        
          
  
  <li class="menu-item menu-item-archives">
    <a href="/blog/archives/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />Archives</a>
</li>

      
        
        
          
  
  <li class="menu-item menu-item-about">
    <a href="https:/yiweiniu.github.io/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-user"></i> <br />About</a>
</li>

      

      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br />Search</a>
        </li>
      
    </ul>
  

  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off"
             placeholder="Searching..." spellcheck="false"
             type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>


  



 </div>
    </header>

    


    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
            

          
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="https://yiweiniu.github.io/blog/blog/2018/04/Identify-circRNAs-and-Fusions-from-RNA-seq-Using-STARChip/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Yiwei Niu">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/blog/images/portrait.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Yiwei Niu's Note">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">Identify circRNAs and Fusions from RNA-seq Using STARChip</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              
              <time title="Post created" itemprop="dateCreated datePublished" datetime="2018-04-23T16:56:25+08:00">2018-04-23</time>
            

            
            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/blog/categories/RNA-seq/" itemprop="url" rel="index"><span itemprop="name">RNA-seq</span></a></span>

                
                
                  , 
                
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/blog/categories/RNA-seq/circRNA/" itemprop="url" rel="index"><span itemprop="name">circRNA</span></a></span>

                
                
              
            </span>
          

          
            
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
                <a href="/blog/2018/04/Identify-circRNAs-and-Fusions-from-RNA-seq-Using-STARChip/#comments" itemprop="discussionUrl">
                  <span class="post-comments-count disqus-comment-count"
                        data-disqus-identifier="2018/04/Identify-circRNAs-and-Fusions-from-RNA-seq-Using-STARChip/" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          

          
            <span class="post-meta-divider">|</span>
            <span class="post-meta-item-icon"
            >
            <i class="fa fa-eye"></i>
             Views: 
            <span class="busuanzi-value" id="busuanzi_value_page_pv" ></span>
            </span>
          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                <span title="Symbols count in article">16k</span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                <span title="Reading time">0:16</span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <p>I want to find out some <a href="https://en.wikipedia.org/wiki/Circular_RNA" target="_blank" rel="noopener">circRNAs</a> from RNA-seq data (total RNA-seq, not poly-A enriched).</p>
<p>There are many tools for this mission. Here is a good review paper <sup class="footnote-ref"><a href="#fn1" id="fnref1">[1]</a></sup> talking about computational methods for analyzing circRNAs, both identification and downstream analysis. Also another review paper about identifying circRNAs <sup class="footnote-ref"><a href="#fn2" id="fnref2">[2]</a></sup>. There are also two evaluation papers for the identification tools <sup class="footnote-ref"><a href="#fn3" id="fnref3">[3]</a></sup><sup class="footnote-ref"><a href="#fn4" id="fnref4">[4]</a></sup>.</p>
<p>From all the tools I know, <code>CIRCexplorer2</code> <sup class="footnote-ref"><a href="#fn5" id="fnref5">[5]</a></sup> and <code>CIRI</code> <sup class="footnote-ref"><a href="#fn6" id="fnref6">[6]</a></sup> are well matained. But I want to try something new: <code>STARChip</code> <sup class="footnote-ref"><a href="#fn7" id="fnref7">[7]</a></sup>.</p>
<blockquote>
<p>STARChip is short for Star Chimeric Post, written by Dr. Nicholas Kipp Akers as part of his work in Bojan Losic’s group at the Icahn Institute of Genomics and Multiscale Biology at Mount Sinai School of Medicine<br><br>
This software is designed to take the chimeric output from the STAR alignment tool and discover high confidence <strong>fusions</strong> and <strong>circular RNA</strong> in the data. Before running, you must have used a recent version of STAR with chimeric output turned on, to align your RNA-Seq data.</p>
</blockquote>
<p>So, it can <strong>identify fusions and circRNAs at the same time</strong>. According to its paper, for circRNA detection, “STARChip achieves the best precision of all tools tested and nearly the best sensitivity. This does not appear to come at an increased resource cost. Both CIRI and CIRCexplorer had competitive precision and sensitivity values; STARChip required 43 and 179% of the runtimes of these packages, respectively, and ∼72% of the memory requirements.”; for fusions, “With STARChip, we have attempted to emphasize precision at the expense of sensitivity in these particular gold-standard studies, reasoning that such hyper-tuning inflates type I error in mining novel datasets.”</p>
<p>I’ve discussed with the author Kipp Akers about the precision: <a href="https://github.com/LosicLab/starchip/issues/9#issuecomment-381181507" target="_blank" rel="noopener">https://github.com/LosicLab/starchip/issues/9#issuecomment-381181507</a>.  He said:</p>
<blockquote>
<p>To your final question, my goal with STARChip was to develop a tool that focused on precision. There are a dozen fusion finders out there that sacrifice everything to get the highest sensitivity. For my projects, this was not too helpful. However, STARChip’s read requirement settings can be set manually and because it runs so quickly, it’s easy to play with the settings to turn up sensitivity and turn down precision and see what you get. Feel free to do so, and let me know what you find!</p>
</blockquote>
<p>I agree with the designing purpose of <code>STARChip</code>, so I decide to give it a shot.</p>
<p>There are two main modules in <code>STARChip</code>:</p>
<ul>
<li>starchip-fusions is for fusion detection. It runs on individual samples.
<ul>
<li><code>/path/to/starchip/starchip-fusions.pl output_seed Chimeric.out.junction Paramters.txt</code></li>
</ul>
</li>
<li>starchip-circles is for circRNA detection. It runs on groups of samples.
<ul>
<li><code>/path/to/starchip/starchip-circles.pl STARdirs.txt Parameters.txt</code></li>
<li><code>/path/to/starchip/starchip-circles.pl fastq_files.txt parameters.txt</code></li>
</ul>
</li>
</ul>
<p>Notes below are more for my own convenience. See its <a href="https://github.com/LosicLab/starchip" target="_blank" rel="noopener">git repo</a> for full usage.</p>
<h2 id="prepare"><a class="markdownIt-Anchor" href="#prepare"></a> prepare</h2>
<blockquote>
<p>STARChip is written to be an extension of the STAR read aligner. It is optional for STARChip to run STAR on your samples. In most instances to run STARChip you must first run star on each of your samples. See the STAR documentation for installation, as well as building or downloading a STAR genome index. It is absolutely critical however, that you follow the STAR manual’s instructions and build a genome using all chromosomes plus unplaced contigs. Not doing so will strongly inflate your false positives rate, because reads that map perfectly to an unplaced contig will instead find the next best alignment, often a chimeric alignment. Run STAR with the following parameters required for chimeric output: –chimSegmentMin X –chimJunctionOverhangMin X (where X is an integer). Your project will have it’s own requirements, but a good starting point for your star alignments might look like:<br><br>
<code>STAR --genomeDir /path/to/starIndex/ --readFilesIn file1_1.fastq.gz file1_2.fastq.gz --runThreadN 11 --outReadsUnmapped Fastx --quantMode GeneCounts --chimSegmentMin 15 --chimJunctionOverhangMin 15 --outSAMstrandField intronMotif --readFilesCommand zcat --outSAMtype BAM Unsorted</code></p>
</blockquote>
<h3 id="referencebed-files"><a class="markdownIt-Anchor" href="#referencebed-files"></a> reference/BED files</h3>
<blockquote>
<p>STARChip makes use of gtf files for annotating fusions and circRNA with gene names.</p>
</blockquote>
<p>First, download the package and prepare annotation files:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">$</span><span class="bash"> git <span class="built_in">clone</span> https://github.com/LosicLab/starchip.git &amp;&amp; <span class="built_in">cd</span> starchip</span></span><br><span class="line"></span><br><span class="line"><span class="meta">$</span><span class="bash"> mkdir starchip_ref &amp;&amp; ./setup.sh ~/RefData/Homo_sapiens/GENCODE_v27/gencode.v27.annotation.gtf ~/RefData/Homo_sapiens/GRCh38_no_alt/genome.fa ./starchip_ref</span></span><br></pre></td></tr></table></figure>
<h3 id="additional-files-for-fusions"><a class="markdownIt-Anchor" href="#additional-files-for-fusions"></a> additional files for Fusions</h3>
<p><code>starchip-fusions</code> filters using the location of known repeats in bed format as well. Following the instructions in the picture to download repeats from UCSC genome browser.</p>
<ol>
<li>Go to <a href="http://genome.ucsc.edu/cgi-bin/hgTables" target="_blank" rel="noopener">http://genome.ucsc.edu/cgi-bin/hgTables</a></li>
<li>Change ‘genome’ to your desired genome</li>
<li>Change the following settings:
<ol>
<li>group: Repeat</li>
<li>track: RepeatMasker</li>
<li>region: genome</li>
<li>output format: BED</li>
<li>output file: some reasonable name.bed</li>
</ol>
</li>
<li>Click ‘get output’ to download your bed file.</li>
<li>On your local machine sort the bed file: <code>sort -k1,1 -k2,2n repeats.bed &gt; repeats.sorted.bed</code></li>
</ol>
<p><strong>If you’re working on <code>hg19</code> or <code>hg38</code>, you don’t have to do the following things</strong>. The files needed are already included in the directory of <code>STARChip</code>.</p>
<p><code>starchip-fusions</code> can also make use of known antibody parts, and copy number variants. These files come with starchip for human hg19 and hg38 in the reference directory. For other species you can create your own in the simple format: <code>Chromosome StartPosition EndPosition</code></p>
<p>Finally, starchip-fusions uses known gene families and known/common false-positive pairs to filter out fusions which are likely mapping errors or PCR artifacts. Family data can be downloaded from ensembl biomart:</p>
<ol>
<li>Go to <a href="http://www.ensembl.org/biomart/martview" target="_blank" rel="noopener">http://www.ensembl.org/biomart/martview</a></li>
<li>Database: Ensembl Genes</li>
<li>Dataset: Your species</li>
<li>Click Attributes on the left hand side.
<ol>
<li>Under GENE dropdown, select only “Gene Name”</li>
<li>Under PROTEIN FAMILIES AND DOMAINS dropdown select Ensembl Protein Family ID.</li>
</ol>
</li>
<li>Click Results at the top.</li>
<li>Export the file. It should have two columns, Family ID and Gene ID.</li>
</ol>
<p>Known false positives are stored within data/pseudogenes.txt. In practice, we’ve found that pseudogenes and tissue specific highly expressed genes are commonly “fused” via PCR template switching errors. Feel free to put add any additional lines that result from your data to this file in the format: <code>Gene1Name Gene2Name</code></p>
<h2 id="run-starchip"><a class="markdownIt-Anchor" href="#run-starchip"></a> run STARChip</h2>
<p>Since my previous run of <code>STAR</code> didn’t use parameters <code>--chimSegmentMin</code> and <code>--chimJunctionOverhangMin</code>, I have to start with <code>Fastq</code> files.</p>
<p><code>starchip-circles</code> can run from <code>Fastq</code> files, but <code>starchip-fusions</code> starts from <code>Chimeric.out.junction</code>. I’ll first run <code>starchip-circles</code> then run <code>starchip-fusions</code>.</p>
<p>First of all, I prepare dirs for <code>STARChip</code> under my WORKDIR like this:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">STARChip/</span><br><span class="line">├── STARChip-circRNA</span><br><span class="line">│   ├── starchip-circles.fastqfiles        # the fastq files</span><br><span class="line">│   └── starchip-circles.params            # starchip-circles parameters</span><br><span class="line">└── STARChip-fusions</span><br><span class="line">    └── starchip-fusions.param              # starchip-fusions parameters</span><br></pre></td></tr></table></figure>
<h3 id="run-starchip-circles"><a class="markdownIt-Anchor" href="#run-starchip-circles"></a> run starchip-circles</h3>
<p>The parameter file and <code>Fastq</code> file:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">$</span><span class="bash"> cat starchip-circles.params </span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#Parameters for starchimp-circles</span></span></span><br><span class="line">readsCutoff = 5</span><br><span class="line">minSubjectLimit = 10</span><br><span class="line">cpus = 20</span><br><span class="line">do_splice = True</span><br><span class="line">cpmCutoff = 0</span><br><span class="line">subjectCPMcutoff = 0 </span><br><span class="line">annotate = true</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash">Reference Files</span></span><br><span class="line">refbed = /home/niuyw/software/starchip/starchip_ref/gencode.v27.annotation.gtf.bed  #use setup.sh to create this from a gtf. </span><br><span class="line">refFasta = /home/niuyw/RefData/Homo_sapiens/GRCh38_no_alt/genome.fa</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash">STAR Parameters</span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Do you use a prefix for your STAR output?</span></span></span><br><span class="line">starprefix =  </span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Are you starting from fastq and need to run STAR alignment? </span></span></span><br><span class="line">runSTAR = True</span><br><span class="line">STARgenome = /home/niuyw/RefData/Homo_sapiens/GRCh38_no_alt/STARgenomes  #not necassary if runSTAR != True</span><br><span class="line">STARreadcommand = zcat   #cat for fastq, zcat for fastq.gz etc. not necassary	if runSTAR != True</span><br><span class="line">IDstepsback = 1 ## this is the position from the right of your path of the name of your files.  </span><br><span class="line"><span class="meta">		#</span><span class="bash"><span class="comment">#for example: /path/to/sample1/star/2.4.2/output/Chimeric.out.junction </span></span></span><br><span class="line"><span class="meta">		#</span><span class="bash"><span class="comment">#sample1 is 4 steps back.</span></span></span><br><span class="line"><span class="meta">		#</span><span class="bash"><span class="comment">#or /path/to/star/2.4.2/sample1/Chimeric.out.junction</span></span></span><br><span class="line"><span class="meta">		#</span><span class="bash">sample1 is 1 step back.</span></span><br><span class="line"></span><br><span class="line"><span class="meta">$</span><span class="bash"> head -3 starchip-circles.fastqfiles </span></span><br><span class="line">/home/wuw/20160823_Xijing/clean_fastq/XJ-3-1-25_R1.fastq.gz  /home/wuw/20160823_Xijing/clean_fastq/XJ-3-1-25_R2.fastq.gz</span><br><span class="line">/home/wuw/20160823_Xijing/clean_fastq/XJ-2-1-25_R1.fastq.gz  /home/wuw/20160823_Xijing/clean_fastq/XJ-2-1-25_R2.fastq.gz</span><br><span class="line">/home/wuw/20160823_Xijing/clean_fastq/XJ-7-1-25_R1.fastq.gz  /home/wuw/20160823_Xijing/clean_fastq/XJ-7-1-25_R2.fastq.gz</span><br></pre></td></tr></table></figure>
<p>Then go into the <code>$WORKDIR/STARChip/STARChip-circRNA</code> and run <code>starchip-circles</code> to generate scripts:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">$</span><span class="bash"> <span class="variable">$path2circles</span> starchip-circles.fastqfiles starchip-circles.params</span></span><br><span class="line">Using the following parameters: Circular RNA must have at least 5 reads in at least 10 subjects/output files.  Using 20 CPUs.</span><br><span class="line">Rscript must be callable.  Requiring 0 subjects/outputs with 0 Counts per million circular reads to count a given circular RNA</span><br><span class="line">Other requirements are bedtools (&gt;= 2.24.0)</span><br><span class="line">You have indicated you would like STARChip to perform STAR alignments.  starchip-circles.fastqfiles should contain a list of fastq files; 1 sample per line, multiple files separated by a comma, and paired end files separated by a space.</span><br><span class="line">STARChip run scripts generated, please run ./Step1.sh through Step4.sh to detect and quantify circRNA</span><br></pre></td></tr></table></figure>
<p>There will be four scripts:</p>
<ul>
<li><code>Step1.sh</code>: align</li>
<li><code>Step2.sh</code>: discover circRNA</li>
<li><code>Step3.sh</code>: re-align</li>
<li><code>Step4.sh</code>: quantify/annotate</li>
</ul>
<p><code>Step2.sh</code> and <code>Step3.sh</code> use <code>STAR</code> in the system <code>PATH</code>, but I want to use another one:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sed -i '3,$s|^|/home/niuyw/software/STAR-2.5.3a/bin/Linux_x86_64_static/|g' Step1.sh</span><br><span class="line">sed -i '3,$s|^|/home/niuyw/software/STAR-2.5.3a/bin/Linux_x86_64_static/|g' Step3.sh</span><br></pre></td></tr></table></figure>
<p>I’m working on a PBS grid system, then I create a script to submit these scripts:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash">!/bin/bash</span></span><br><span class="line"><span class="meta">#</span><span class="bash">PBS -V</span></span><br><span class="line"><span class="meta">#</span><span class="bash">PBS -j eo</span></span><br><span class="line"><span class="meta">#</span><span class="bash">PBS -N STARChip-circles</span></span><br><span class="line"><span class="meta">#</span><span class="bash">PBS -q Blade</span></span><br><span class="line"><span class="meta">#</span><span class="bash">PBS -l nodes=1:ppn=20</span></span><br><span class="line"><span class="meta">#</span><span class="bash">PBS -d /home/wuw/20160823_Xijing/STARChip/STARChip-circRNA</span></span><br><span class="line"></span><br><span class="line">echo Start time is `date +%Y/%m/%d--%H:%M`</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> work dir</span></span><br><span class="line">WORKDIR=/home/wuw/20160823_Xijing/STARChip/STARChip-circRNA</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> starchip-circles</span></span><br><span class="line">sh Step1.sh</span><br><span class="line">sh Step2.sh</span><br><span class="line">sh Step3.sh</span><br><span class="line">sh Step4.sh</span><br><span class="line"></span><br><span class="line">echo Finish time is `date +%Y/%m/%d--%H:%M`</span><br></pre></td></tr></table></figure>
<p>In my samples, only four circRNAs were identified by <code>STARChip-circles</code>.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ cat circRNA.5reads.10ind.countmatrix</span><br><span class="line">8_4-10_R1.fastq.gz	8_4-11_R1.fastq.gz	8_4-3_R1.fastq.gz	8_4-4_R1.fastq.gz	8_4-5_R1.fastq.gz	8_4-6_R1.fastq.gz	8_4-7_R1.fastq.gz	8_4-8_R1.fastq.gz   8_4-9_R1.fastq.gz	XJ-10-1-25_R1.fastq.gz	XJ-11-1-25_R1.fastq.gz	XJ-1-1-25_R1.fastq.gz	XJ-12-1-25_R1.fastq.gz	XJ-13-1-25_R1.fastq.gz	XJ-2-1-25_R1.fastq.gz	XJ-3-1-25_R1.fastq.gXJ-4-1-25_R1.fastq.gz	XJ-5-1-25_R1.fastq.gz	XJ-6-1-25_R1.fastq.gz	XJ-7-1-25_R1.fastq.gz	XJ-8-1-25_R1.fastq.gz	XJ-9-1-25_R1.fastq.gz</span><br><span class="line">chr1:117402186-117442325	0	0	0	2	0	0	4	0	6	3	8	0	0	4	3	2	2	2	0	0   3</span><br><span class="line">chr15:101213315-101216678	0	0	38	20	8	0	0	0	0	0	0	26	0	0	58	60	18	12	25	23  0</span><br><span class="line">chr15:90217439-90219891	0	4	0	1	0	0	0	0	1	0	1	0	5	2	0	1	0	2	0	1	9   4</span><br><span class="line">chr9:111786793-111787947	0	0	33	186	23	0	0	0	0	1	0	10	0	0	5	29	42	14	15	15  0</span><br></pre></td></tr></table></figure>
<h3 id="run-starchip-fusions"><a class="markdownIt-Anchor" href="#run-starchip-fusions"></a> run starchip-fusions</h3>
<p>The parameter file:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">## Parameters for fusions-from-star.pl</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Describing Your Data:</span></span></span><br><span class="line"></span><br><span class="line">pairedend = TRUE  #TRUE means paired end data.  any other value means single end. $spancutoff should be 0 if data is single end.</span><br><span class="line">consensus = TRUE  # anything but TRUE will make this skip the consensus sequence generation for each fusion.</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Filters, Cutoffs </span></span></span><br><span class="line"></span><br><span class="line">splitReads = auto # number of minimum read support at jxn.  Minimum 2.  Greatly impacts running time. Other options are: "auto" , "highsensitivity" , and "highprecision"   </span><br><span class="line">uniqueReads = 2 # number of unique read support values (higher indicates more likely to be real. lower is more likely amplification artifact).</span><br><span class="line">spancutoff = 1 #minimum number of non-split reads support.  If single end data, this must be 0 or auto.  Other options are: "auto" , "highsensitivity" , and "highprecision"</span><br><span class="line">wiggle = 500  #number of base-pairs of 'wiggle-room' when determining the location of a fusion (for spanning read counts)</span><br><span class="line">overlapLimit = 5 #wiggle room for joining very closely called fusion sites.</span><br><span class="line">samechrom_wiggle = 20000 #this is the distance that fusions have to be from each other if on the same chromosome.  Set to 0 if you want no filtering of same-chromosome pr</span><br><span class="line">lopsidedupper = 10 # (topsidereads + 0.1) / (bottomsidereads + 0.1) must be below this value. set very high to disable.  Reccomended setting 5</span><br><span class="line">lopsidedlower = 0.1 # (topsidereads + 0.1) / (bottomsidereads + 0.1) must be above this value. set to 0 to disable. Reccomended setting 0.2</span><br><span class="line">cnvwiggle = 1000 #we skip fusions that can be explained by known cnvs.  how close to the edges of the cnv must our fusion be?</span><br><span class="line">circlesize = 100000 #we skip fusions that look more like circular rna/backsplices.  how big (bp) could a circle be?    </span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Local Reference Files:</span></span></span><br><span class="line"><span class="meta">	#</span><span class="bash">refbed is a bed format version of a gtf.  This should probably be derived from the same GTF that STAR aligned with using setup.sh.   </span></span><br><span class="line">refbed=/home/niuyw/software/starchip/starchip_ref/gencode.v27.annotation.gtf.bed</span><br><span class="line">        # a bed format list of known repeats</span><br><span class="line">repeatbed=/home/niuyw/RefData/RepeatMasker/hg38.ucsc.180414.rmsk.sorted.bed</span><br><span class="line">        # fasta reference.  should be indexed (run 'samtools faidx file.fa')</span><br><span class="line">refFasta = /home/niuyw/RefData/Homo_sapiens/GRCh38_no_alt/genome.fa</span><br><span class="line">abparts = reference/hg38.abparts</span><br><span class="line">cnvs = reference/conrad_hg38.cnvs</span><br><span class="line">familyfile = reference/ensfams.txt</span><br><span class="line">falsepositives = reference/knownFP.txt </span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash">Scoring Parameters (feel free to tweak).</span></span><br><span class="line">splitscoremod = 10</span><br><span class="line">spanscoremod = 20</span><br><span class="line">skewpenalty = 4</span><br><span class="line">repeatpenalty = 0.5  # score = score*(repeatpenalty^repeats)  --&gt; a fusion can have 0,1,or 2 sites fall into repeat regions.</span><br></pre></td></tr></table></figure>
<p>Based on the output of previous <code>STAR</code> running for <code>starchip-circles</code>, the script to run <code>starchip-fusions</code> contains:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash">!/bin/bash</span></span><br><span class="line"><span class="meta">#</span><span class="bash">PBS -V</span></span><br><span class="line"><span class="meta">#</span><span class="bash">PBS -j eo</span></span><br><span class="line"><span class="meta">#</span><span class="bash">PBS -N STARChip-fusions</span></span><br><span class="line"><span class="meta">#</span><span class="bash">PBS -q Blade</span></span><br><span class="line"><span class="meta">#</span><span class="bash">PBS -l nodes=1:ppn=2</span></span><br><span class="line"><span class="meta">#</span><span class="bash">PBS -d /home/wuw/20160823_Xijing/STARChip/STARChip-fusions</span></span><br><span class="line"></span><br><span class="line">echo Start time is `date +%Y/%m/%d--%H:%M`</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> work dir</span></span><br><span class="line">WORKDIR=/home/wuw/20160823_Xijing/STARChip</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> starchip-fusions</span></span><br><span class="line">TOOLDIR=/home/niuyw/software</span><br><span class="line">path2fusions=$TOOLDIR/starchip/starchip-fusions.pl</span><br><span class="line"></span><br><span class="line">for sample in for sample in XJ-3-1-25 XJ-2-1-25 XJ-7-1-25 XJ-4-1-25 XJ-5-1-25 XJ-6-1-25 XJ-1-1-25 XJ-10-1-25 XJ-9-1-25 XJ-13-1-25 XJ-11-1-25 XJ-12-1-25 XJ-8-1-25 8_4-3 8_4-4 8_4-5 8_4-6 8_4-7 8_4-8 8_4-9 8_4-10 8_4-11</span><br><span class="line">do</span><br><span class="line">	/usr/bin/perl $path2fusions $sample $WORKDIR/STARChip-circRNA/STARout/$&#123;sample&#125;_R1.fastq.gz/Chimeric.out.junction starchip-fusions.param</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line">echo Finish time is `date +%Y/%m/%d--%H:%M`</span><br></pre></td></tr></table></figure>
<p>In my samples, no fusions were found by <code>STARChip-fusions</code>, and I don’t want to tweak parameters to improve sensitivity.</p>
<h2 id="change-notes"><a class="markdownIt-Anchor" href="#change-notes"></a> Change notes</h2>
<ul>
<li>20180413: create the note.</li>
</ul>
<hr class="footnotes-sep">
<section class="footnotes">
<ol class="footnotes-list">
<li id="fn1" class="footnote-item"><p>Gao Y, Zhao F. 2018 Jan 12. Computational Strategies for Exploring Circular RNAs. Trends in Genetics. doi:10.1016/j.tig.2017.12.016. [accessed 2018 Jan 15]. <a href="https://www.sciencedirect.com/science/article/pii/S0168952517302366" target="_blank" rel="noopener">https://www.sciencedirect.com/science/article/pii/S0168952517302366</a>. <a href="#fnref1" class="footnote-backref">↩︎</a></p>
</li>
<li id="fn2" class="footnote-item"><p>Szabo L, Salzman J. 2016. Detecting circular RNAs: bioinformatic and experimental challenges. Nat Rev Genet 17:679–692. doi:10.1038/nrg.2016.114. <a href="#fnref2" class="footnote-backref">↩︎</a></p>
</li>
<li id="fn3" class="footnote-item"><p>Hansen TB, Ven? MT, Damgaard CK, Kjems J. 2016. Comparison of circular RNA prediction tools. Nucleic Acids Research 44:e58–e58. doi:10.1093/nar/gkv1458. <a href="#fnref3" class="footnote-backref">↩︎</a></p>
</li>
<li id="fn4" class="footnote-item"><p>Zeng X, Lin W, Guo M, Zou Q. 2017. A comprehensive overview and evaluation of circular RNA detection tools. PLOS Computational Biology 13:e1005420. doi:10.1371/journal.pcbi.1005420. <a href="#fnref4" class="footnote-backref">↩︎</a></p>
</li>
<li id="fn5" class="footnote-item"><p>Zhang X-O, Dong R, Zhang Y, Zhang J-L, Luo Z, Zhang J, Chen L-L, Yang L. 2016. Diverse alternative back-splicing and alternative splicing landscape of circular RNAs. Genome Res. 26:1277–1287. doi:10.1101/gr.202895.115. <a href="#fnref5" class="footnote-backref">↩︎</a></p>
</li>
<li id="fn6" class="footnote-item"><p>Gao Y, Wang J, Zhao F. 2015. CIRI: an efficient and unbiased algorithm for de novo circular RNA identification. Genome Biology 16:4. doi:10.1186/s13059-014-0571-3. <a href="#fnref6" class="footnote-backref">↩︎</a></p>
</li>
<li id="fn7" class="footnote-item"><p>Akers NK, Schadt EE, Losic B. 2018 Feb 20. STAR Chimeric Post for rapid detection of circular RNA and fusion transcripts. Bioinformatics:bty091–bty091. doi:10.1093/bioinformatics/bty091. <a href="#fnref7" class="footnote-backref">↩︎</a></p>
</li>
</ol>
</section>

      
    </div>

    

    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/blog/tags/bio-tools/" rel="tag"># bio-tools</a>
          
            <a href="/blog/tags/RNA-seq/" rel="tag"># RNA-seq</a>
          
            <a href="/blog/tags/circRNA/" rel="tag"># circRNA</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/blog/2018/04/How-to-install-perl-modules/" rel="next" title="How to Install Perl Modules">
                <i class="fa fa-chevron-left"></i> How to Install Perl Modules
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/blog/2018/04/Genome-Assembly-Pipeline-SMARTdenovo/" rel="prev" title="Genome Assembly Pipeline: SMARTdenovo">
                Genome Assembly Pipeline: SMARTdenovo <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
        <!-- Go to www.addthis.com/dashboard to customize your tools -->
<div class="addthis_inline_share_toolbox">
  <script type = "text/javascript" src = "//s7.addthis.com/js/300/addthis_widget.js#pubid=ra-5a7ad07b7823d03b" async = "async" ></script>
</div>

      
    </div>
  </div>


          </div>
          

  
    <div class="comments" id="comments">
      <div id="disqus_thread">
        <noscript>
          Please enable JavaScript to view the
          <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a>
        </noscript>
      </div>
    </div>

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            Overview
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image"
                src="/blog/images/portrait.jpg"
                alt="Yiwei Niu" />
            
              <p class="site-author-name" itemprop="name">Yiwei Niu</p>
              <p class="site-description motion-element" itemprop="description">a beginner, a learner</p>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/blog/archives/">
                
                    <span class="site-state-item-count">16</span>
                    <span class="site-state-item-name">posts</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-categories">
                  <a href="/blog/categories/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">12</span>
                    <span class="site-state-item-name">categories</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-tags">
                  <a href="/blog/tags/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">23</span>
                    <span class="site-state-item-name">tags</span>
                  </a>
                </div>
              
            </nav>
          

          
            <div class="feed-link motion-element">
              <a href="/blog/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          
            <div class="links-of-author motion-element">
              
                <span class="links-of-author-item">
                  <a href="mailto:xiaohuwangwang@qq.com" target="_blank" title="E-Mail"><i class="fa fa-fw fa-envelope"></i>E-Mail</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="https://github.com/YiweiNiu" target="_blank" title="GitHub"><i class="fa fa-fw fa-github"></i>GitHub</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="http://weibo.com/u/3049858680" target="_blank" title="Weibo"><i class="fa fa-fw fa-weibo"></i>Weibo</a>
                  
                </span>
              
                <span class="links-of-author-item">
                  <a href="https://twitter.com/ywniu" target="_blank" title="Twitter"><i class="fa fa-fw fa-twitter"></i>Twitter</a>
                  
                </span>
              
            </div>
          

          
          

          
          
            <div class="links-of-blogroll motion-element links-of-blogroll-block">
              <div class="links-of-blogroll-title">
                <i class="fa  fa-fw fa-link"></i>
                Links
              </div>
              <ul class="links-of-blogroll-list">
                
                  <li class="links-of-blogroll-item">
                    <a href="https://kaigedong.github.io/" title="Kaige Dong" target="_blank">Kaige Dong</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://homolog.us/blogs/" title="Homolog.us" target="_blank">Homolog.us</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="https://cgatoxford.wordpress.com/" title="CGAT" target="_blank">CGAT</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://www.gettinggeneticsdone.com/" title="Getting Genetics Done" target="_blank">Getting Genetics Done</a>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <a href="http://www.acgt.me/" title="ACGT" target="_blank">ACGT</a>
                  </li>
                
              </ul>
            </div>
          

          
            
          
          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#prepare"><span class="nav-number">1.</span> <span class="nav-text"> prepare</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#referencebed-files"><span class="nav-number">1.1.</span> <span class="nav-text"> reference/BED files</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#additional-files-for-fusions"><span class="nav-number">1.2.</span> <span class="nav-text"> additional files for Fusions</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#run-starchip"><span class="nav-number">2.</span> <span class="nav-text"> run STARChip</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#run-starchip-circles"><span class="nav-number">2.1.</span> <span class="nav-text"> run starchip-circles</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#run-starchip-fusions"><span class="nav-number">2.2.</span> <span class="nav-text"> run starchip-fusions</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#change-notes"><span class="nav-number">3.</span> <span class="nav-text"> Change notes</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; 2017 &mdash; <span itemprop="copyrightYear">2018</span>
  <span class="with-love" id="animate">
    <i class="fa fa-mars"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Yiwei Niu</span>

  

  
</div>




  <div class="powered-by">Powered by <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a></div>








        
<div class="busuanzi-count">
  <script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script>

  
    <span class="site-uv" title="Total Visitors">
      <i class="fa fa-user"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_uv"></span>
    </span>
  

  
    <span class="site-pv" title="Total Views">
      <i class="fa fa-eye"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_pv"></span>
    </span>
  
</div>









        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
          <span id="scrollpercent"><span>0</span>%</span>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>


























  
  
    <script type="text/javascript" src="/blog/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  


  


  <script type="text/javascript" src="/blog/js/src/utils.js?v=6.1.0"></script>

  <script type="text/javascript" src="/blog/js/src/motion.js?v=6.1.0"></script>



  
  


  <script type="text/javascript" src="/blog/js/src/affix.js?v=6.1.0"></script>

  <script type="text/javascript" src="/blog/js/src/schemes/pisces.js?v=6.1.0"></script>



  
  <script type="text/javascript" src="/blog/js/src/scrollspy.js?v=6.1.0"></script>
<script type="text/javascript" src="/blog/js/src/post-details.js?v=6.1.0"></script>



  


  <script type="text/javascript" src="/blog/js/src/bootstrap.js?v=6.1.0"></script>



  

  
    <script id="dsq-count-scr" src="https://yiwei-nius-note.disqus.com/count.js" async></script>
  

  
    <script type="text/javascript">
      var disqus_config = function () {
        this.page.url = 'https://yiweiniu.github.io/blog/2018/04/Identify-circRNAs-and-Fusions-from-RNA-seq-Using-STARChip/';
        this.page.identifier = '2018/04/Identify-circRNAs-and-Fusions-from-RNA-seq-Using-STARChip/';
        this.page.title = 'Identify circRNAs and Fusions from RNA-seq Using STARChip';
      };
      function loadComments () {
        var d = document, s = d.createElement('script');
        s.src = 'https://yiwei-nius-note.disqus.com/embed.js';
        s.setAttribute('data-timestamp', '' + +new Date());
        (d.head || d.body).appendChild(s);
      }
      
        loadComments();
      
    </script>
  





	





  












  

  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/blog/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  

  

  
  

  

  

  

  

</body>
</html>
